{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "I4W06DiqLvTB",
        "outputId": "d1fe1681-8ca7-43be-e707-968570e2c534"
      },
      "outputs": [],
      "source": [
        "# !pip show openai"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 68,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "utC8KMT_EKFl",
        "outputId": "e97b5442-85cf-43e7-fa28-2207884f79f6"
      },
      "outputs": [],
      "source": [
        "# # Make sure the OpenAI library is installed\n",
        "# %pip install openai==0.28.1\n",
        "\n",
        "# # We'll need to install the Chroma client\n",
        "# %pip install chromadb\n",
        "\n",
        "# # Install numpy for data manipulation\n",
        "# %pip install numpy\n",
        "\n",
        "# %pip install tiktoken\n",
        "\n",
        "#%pip install matplotlib"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 35,
      "metadata": {
        "id": "xEtyB14PET0k"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import openai\n",
        "import chromadb\n",
        "import numpy as np\n",
        "import requests"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "PdF6NPTWHsJC"
      },
      "outputs": [],
      "source": [
        "df = pd.read_csv('tableExport-2.csv')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 44,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Processed 100 rows.\n",
            "Processed 200 rows.\n",
            "Processed 300 rows.\n",
            "Processed 400 rows.\n",
            "Processed 500 rows.\n",
            "Processed 600 rows.\n",
            "Processed 700 rows.\n",
            "Processed 800 rows.\n",
            "Processed 900 rows.\n",
            "Processed 1000 rows.\n",
            "Processed 1100 rows.\n",
            "Processed 1200 rows.\n",
            "Processed 1300 rows.\n",
            "Processed 1400 rows.\n",
            "Processed 1500 rows.\n",
            "Processed 1600 rows.\n",
            "Total rows processed: 1635\n"
          ]
        }
      ],
      "source": [
        "import requests\n",
        "import pandas as pd\n",
        "\n",
        "# Assuming df is your existing DataFrame\n",
        "\n",
        "# Define a function to fetch README content\n",
        "def get_readme(url):\n",
        "    if pd.isna(url) or url == \"\":\n",
        "        return \"\"\n",
        "    try:\n",
        "        # Construct URL to the raw README file\n",
        "        readme_url = url.replace('github.com', 'raw.githubusercontent.com').rstrip('/') + '/master/README.md'\n",
        "        response = requests.get(readme_url)\n",
        "        if response.status_code == 200:\n",
        "            return response.text\n",
        "        else:\n",
        "            return \"\"\n",
        "    except Exception:\n",
        "        # Handle exceptions\n",
        "        return \"\"\n",
        "\n",
        "# Counter for printing status\n",
        "counter = 0\n",
        "\n",
        "# Apply the function to each row in the DataFrame\n",
        "for index, row in df.iterrows():\n",
        "    df.at[index, 'Readme'] = get_readme(row['Code'])\n",
        "    \n",
        "    # Increment the counter\n",
        "    counter += 1\n",
        "    if counter % 100 == 0:\n",
        "        print(f\"Processed {counter} rows.\")\n",
        "\n",
        "# Optionally, if you want to see the final count:\n",
        "print(f\"Total rows processed: {counter}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 284
        },
        "id": "u6NEzIE_IM62",
        "outputId": "52c2eed9-9687-4be4-98e1-b15755cdf840"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Platform</th>\n",
              "      <th>DOIs</th>\n",
              "      <th>Pub Dates</th>\n",
              "      <th>Citations</th>\n",
              "      <th>Code</th>\n",
              "      <th>Description</th>\n",
              "      <th>License</th>\n",
              "      <th>Categories</th>\n",
              "      <th>Assembly</th>\n",
              "      <th>...</th>\n",
              "      <th>Immune</th>\n",
              "      <th>Variants</th>\n",
              "      <th>Haplotypes</th>\n",
              "      <th>Allele Specific</th>\n",
              "      <th>Perturbations</th>\n",
              "      <th>Visualisation</th>\n",
              "      <th>Interactive</th>\n",
              "      <th>Simulation</th>\n",
              "      <th>Added</th>\n",
              "      <th>Updated</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>acorde</td>\n",
              "      <td>R</td>\n",
              "      <td>10.1038/s41467-022-29497-w, 10.1101/2021.05.07...</td>\n",
              "      <td>2022-04-05, NA</td>\n",
              "      <td>6</td>\n",
              "      <td>https://github.com/ConesaLab/acorde</td>\n",
              "      <td>Isoform co-usage networks from single-cell RNA...</td>\n",
              "      <td>GPL-3.0</td>\n",
              "      <td>Alternative Splicing, Differential Expression,...</td>\n",
              "      <td>False</td>\n",
              "      <td>...</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>True</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>2021-05-14</td>\n",
              "      <td>2022-04-30</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ACTINN</td>\n",
              "      <td>Python</td>\n",
              "      <td>10.1093/bioinformatics/btz592, 10.1101/532093</td>\n",
              "      <td>2019-07-29, NA</td>\n",
              "      <td>109</td>\n",
              "      <td>https://github.com/mafeiyang/ACTINN</td>\n",
              "      <td>ACTINN (Automated Cell Type Identification usi...</td>\n",
              "      <td>GPL-3.0</td>\n",
              "      <td>Classification</td>\n",
              "      <td>False</td>\n",
              "      <td>...</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>2019-02-01</td>\n",
              "      <td>2021-06-28</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>2 rows Ã— 44 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     Name Platform                                               DOIs  \\\n",
              "0  acorde        R  10.1038/s41467-022-29497-w, 10.1101/2021.05.07...   \n",
              "1  ACTINN   Python      10.1093/bioinformatics/btz592, 10.1101/532093   \n",
              "\n",
              "        Pub Dates Citations                                 Code  \\\n",
              "0  2022-04-05, NA         6  https://github.com/ConesaLab/acorde   \n",
              "1  2019-07-29, NA       109  https://github.com/mafeiyang/ACTINN   \n",
              "\n",
              "                                         Description  License  \\\n",
              "0  Isoform co-usage networks from single-cell RNA...  GPL-3.0   \n",
              "1  ACTINN (Automated Cell Type Identification usi...  GPL-3.0   \n",
              "\n",
              "                                          Categories  Assembly  ...  Immune  \\\n",
              "0  Alternative Splicing, Differential Expression,...     False  ...   False   \n",
              "1                                     Classification     False  ...   False   \n",
              "\n",
              "   Variants  Haplotypes  Allele Specific  Perturbations  Visualisation  \\\n",
              "0     False       False            False          False           True   \n",
              "1     False       False            False          False          False   \n",
              "\n",
              "   Interactive  Simulation       Added     Updated  \n",
              "0        False       False  2021-05-14  2022-04-30  \n",
              "1        False       False  2019-02-01  2021-06-28  \n",
              "\n",
              "[2 rows x 44 columns]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "df.head(2)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 41,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "1635"
            ]
          },
          "execution_count": 41,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "len(df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 32,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "['Name',\n",
              " 'Platform',\n",
              " 'DOIs',\n",
              " 'Pub Dates',\n",
              " 'Citations',\n",
              " 'Code',\n",
              " 'Description',\n",
              " 'License',\n",
              " 'Categories',\n",
              " 'Assembly',\n",
              " 'Alignment',\n",
              " 'UMIs',\n",
              " 'Quantification',\n",
              " 'Quality Control',\n",
              " 'Normalisation',\n",
              " 'Imputation',\n",
              " 'Integration',\n",
              " 'Gene Filtering',\n",
              " 'Clustering',\n",
              " 'Classification',\n",
              " 'Ordering',\n",
              " 'Differential Expression',\n",
              " 'Marker Genes',\n",
              " 'Expression Patterns',\n",
              " 'Variable Genes',\n",
              " 'Gene Sets',\n",
              " 'Gene Networks',\n",
              " 'Cell Cycle',\n",
              " 'Dimensionality Reduction',\n",
              " 'Transformation',\n",
              " 'Modality',\n",
              " 'Alternative Splicing',\n",
              " 'Rare Cells',\n",
              " 'Stem Cells',\n",
              " 'Immune',\n",
              " 'Variants',\n",
              " 'Haplotypes',\n",
              " 'Allele Specific',\n",
              " 'Perturbations',\n",
              " 'Visualisation',\n",
              " 'Interactive',\n",
              " 'Simulation',\n",
              " 'Added',\n",
              " 'Updated',\n",
              " 'extented_desc',\n",
              " 'tokens_in_ext_desc']"
            ]
          },
          "execution_count": 32,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "list(df.columns)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "id": "JFatrRiWINet"
      },
      "outputs": [],
      "source": [
        "chroma_client = chromadb.EphemeralClient() # Equivalent to chromadb.Client(), ephemeral.\n",
        "# Uncomment for persistent client\n",
        "# chroma_client = chromadb.PersistentClient()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "9w79gbU-J79t"
      },
      "outputs": [],
      "source": [
        "EMBEDDING_MODEL = \"text-embedding-ada-002\"\n",
        "# change this to biotech specialised model later"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ds9NAdBiJo6b",
        "outputId": "b96c357b-3304-4796-a242-21404ceeb178"
      },
      "outputs": [],
      "source": [
        "from chromadb.utils.embedding_functions import OpenAIEmbeddingFunction\n",
        "\n",
        "# from google.colab import userdata\n",
        "\n",
        "# if userdata.get('OPENAI_API_KEY') is not None:\n",
        "#     openai.api_key = userdata.get('OPENAI_API_KEY')\n",
        "#     print (\"OPENAI_API_KEY is ready\")\n",
        "# else:\n",
        "#     print (\"OPENAI_API_KEY environment variable not found\")\n",
        "\n",
        "\n",
        "embedding_function = OpenAIEmbeddingFunction(api_key='', model_name=EMBEDDING_MODEL)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "metadata": {
        "id": "hSKEDPv5J8_D"
      },
      "outputs": [
        {
          "ename": "UniqueConstraintError",
          "evalue": "Collection scRNA_Tools already exists",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mUniqueConstraintError\u001b[0m                     Traceback (most recent call last)",
            "\u001b[1;32m/home/deek/Documents/Faunabio/MachineLearning/GPT/hackathon/scMagic2/sc_tools_retrieval.ipynb Cell 11\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell:/home/deek/Documents/Faunabio/MachineLearning/GPT/hackathon/scMagic2/sc_tools_retrieval.ipynb#X12sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m scrnatools_description_collection \u001b[39m=\u001b[39m chroma_client\u001b[39m.\u001b[39;49mcreate_collection(name\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mscRNA_Tools\u001b[39;49m\u001b[39m'\u001b[39;49m, embedding_function\u001b[39m=\u001b[39;49membedding_function)\n",
            "File \u001b[0;32m~/.local/lib/python3.10/site-packages/chromadb/api/client.py:187\u001b[0m, in \u001b[0;36mClient.create_collection\u001b[0;34m(self, name, metadata, embedding_function, data_loader, get_or_create)\u001b[0m\n\u001b[1;32m    176\u001b[0m \u001b[39m@override\u001b[39m\n\u001b[1;32m    177\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcreate_collection\u001b[39m(\n\u001b[1;32m    178\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    185\u001b[0m     get_or_create: \u001b[39mbool\u001b[39m \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m,\n\u001b[1;32m    186\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Collection:\n\u001b[0;32m--> 187\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_server\u001b[39m.\u001b[39;49mcreate_collection(\n\u001b[1;32m    188\u001b[0m         name\u001b[39m=\u001b[39;49mname,\n\u001b[1;32m    189\u001b[0m         metadata\u001b[39m=\u001b[39;49mmetadata,\n\u001b[1;32m    190\u001b[0m         embedding_function\u001b[39m=\u001b[39;49membedding_function,\n\u001b[1;32m    191\u001b[0m         data_loader\u001b[39m=\u001b[39;49mdata_loader,\n\u001b[1;32m    192\u001b[0m         tenant\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mtenant,\n\u001b[1;32m    193\u001b[0m         database\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mdatabase,\n\u001b[1;32m    194\u001b[0m         get_or_create\u001b[39m=\u001b[39;49mget_or_create,\n\u001b[1;32m    195\u001b[0m     )\n",
            "File \u001b[0;32m~/.local/lib/python3.10/site-packages/chromadb/telemetry/opentelemetry/__init__.py:127\u001b[0m, in \u001b[0;36mtrace_method.<locals>.decorator.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    125\u001b[0m \u001b[39mglobal\u001b[39;00m tracer, granularity\n\u001b[1;32m    126\u001b[0m \u001b[39mif\u001b[39;00m trace_granularity \u001b[39m<\u001b[39m granularity:\n\u001b[0;32m--> 127\u001b[0m     \u001b[39mreturn\u001b[39;00m f(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    128\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m tracer:\n\u001b[1;32m    129\u001b[0m     \u001b[39mreturn\u001b[39;00m f(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n",
            "File \u001b[0;32m~/.local/lib/python3.10/site-packages/chromadb/api/segment.py:166\u001b[0m, in \u001b[0;36mSegmentAPI.create_collection\u001b[0;34m(self, name, metadata, embedding_function, data_loader, get_or_create, tenant, database)\u001b[0m\n\u001b[1;32m    162\u001b[0m check_index_name(name)\n\u001b[1;32m    164\u001b[0m \u001b[39mid\u001b[39m \u001b[39m=\u001b[39m uuid4()\n\u001b[0;32m--> 166\u001b[0m coll, created \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sysdb\u001b[39m.\u001b[39;49mcreate_collection(\n\u001b[1;32m    167\u001b[0m     \u001b[39mid\u001b[39;49m\u001b[39m=\u001b[39;49m\u001b[39mid\u001b[39;49m,\n\u001b[1;32m    168\u001b[0m     name\u001b[39m=\u001b[39;49mname,\n\u001b[1;32m    169\u001b[0m     metadata\u001b[39m=\u001b[39;49mmetadata,\n\u001b[1;32m    170\u001b[0m     dimension\u001b[39m=\u001b[39;49m\u001b[39mNone\u001b[39;49;00m,\n\u001b[1;32m    171\u001b[0m     get_or_create\u001b[39m=\u001b[39;49mget_or_create,\n\u001b[1;32m    172\u001b[0m     tenant\u001b[39m=\u001b[39;49mtenant,\n\u001b[1;32m    173\u001b[0m     database\u001b[39m=\u001b[39;49mdatabase,\n\u001b[1;32m    174\u001b[0m )\n\u001b[1;32m    176\u001b[0m \u001b[39mif\u001b[39;00m created:\n\u001b[1;32m    177\u001b[0m     segments \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_manager\u001b[39m.\u001b[39mcreate_segments(coll)\n",
            "File \u001b[0;32m~/.local/lib/python3.10/site-packages/chromadb/telemetry/opentelemetry/__init__.py:127\u001b[0m, in \u001b[0;36mtrace_method.<locals>.decorator.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    125\u001b[0m \u001b[39mglobal\u001b[39;00m tracer, granularity\n\u001b[1;32m    126\u001b[0m \u001b[39mif\u001b[39;00m trace_granularity \u001b[39m<\u001b[39m granularity:\n\u001b[0;32m--> 127\u001b[0m     \u001b[39mreturn\u001b[39;00m f(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    128\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m tracer:\n\u001b[1;32m    129\u001b[0m     \u001b[39mreturn\u001b[39;00m f(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n",
            "File \u001b[0;32m~/.local/lib/python3.10/site-packages/chromadb/db/mixins/sysdb.py:225\u001b[0m, in \u001b[0;36mSqlSysDB.create_collection\u001b[0;34m(self, id, name, metadata, dimension, get_or_create, tenant, database)\u001b[0m\n\u001b[1;32m    218\u001b[0m         \u001b[39mreturn\u001b[39;00m (\n\u001b[1;32m    219\u001b[0m             \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mget_collections(\n\u001b[1;32m    220\u001b[0m                 \u001b[39mid\u001b[39m\u001b[39m=\u001b[39mcollection[\u001b[39m\"\u001b[39m\u001b[39mid\u001b[39m\u001b[39m\"\u001b[39m], tenant\u001b[39m=\u001b[39mtenant, database\u001b[39m=\u001b[39mdatabase\n\u001b[1;32m    221\u001b[0m             )[\u001b[39m0\u001b[39m],\n\u001b[1;32m    222\u001b[0m             \u001b[39mFalse\u001b[39;00m,\n\u001b[1;32m    223\u001b[0m         )\n\u001b[1;32m    224\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 225\u001b[0m         \u001b[39mraise\u001b[39;00m UniqueConstraintError(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCollection \u001b[39m\u001b[39m{\u001b[39;00mname\u001b[39m}\u001b[39;00m\u001b[39m already exists\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    227\u001b[0m topic \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_assignment_policy\u001b[39m.\u001b[39massign_collection(\u001b[39mid\u001b[39m)\n\u001b[1;32m    228\u001b[0m collection \u001b[39m=\u001b[39m Collection(\n\u001b[1;32m    229\u001b[0m     \u001b[39mid\u001b[39m\u001b[39m=\u001b[39m\u001b[39mid\u001b[39m,\n\u001b[1;32m    230\u001b[0m     topic\u001b[39m=\u001b[39mtopic,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    235\u001b[0m     database\u001b[39m=\u001b[39mdatabase,\n\u001b[1;32m    236\u001b[0m )\n",
            "\u001b[0;31mUniqueConstraintError\u001b[0m: Collection scRNA_Tools already exists"
          ]
        }
      ],
      "source": [
        "scrnatools_description_collection = chroma_client.create_collection(name='scRNA_Tools', embedding_function=embedding_function)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 69,
      "metadata": {
        "id": "QpI43znWhHz5"
      },
      "outputs": [],
      "source": [
        "import tiktoken\n",
        "def num_tokens_from_string(string: str, model_name: str = \"gpt-3.5-turbo\") -> int:\n",
        "    \"\"\"Returns the number of tokens in a text string.\"\"\"\n",
        "    encoding = tiktoken.encoding_for_model(model_name)\n",
        "    num_tokens = len(encoding.encode(string))\n",
        "    return num_tokens"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 65,
      "metadata": {
        "id": "Qi618y5Thg_p"
      },
      "outputs": [],
      "source": [
        "def make_desc(row):\n",
        "    # Example function using row 'A' and 'B'\n",
        "    return 'Platform: ' + row['Platform'] + \\\n",
        "            '\\n Description: ' + row['Description'] + \\\n",
        "            '\\n Categories: ' + row['Categories']\n",
        "\n",
        "df['extented_desc'] = df.apply(make_desc, axis=1)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 66,
      "metadata": {},
      "outputs": [],
      "source": [
        "def make_desc(row):\n",
        "    # Example function using row 'A' and 'B'\n",
        "    return 'Platform: ' + row['Platform'] + \\\n",
        "            '\\n Description: ' + row['Description'] + \\\n",
        "            '\\n Categories: ' + row['Categories'] + \\\n",
        "            '\\n Readme: ' + row['Readme']\n",
        "\n",
        "df['extented_desc_readme'] = df.apply(make_desc, axis=1)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "id": "pJCwFeDsiZyG"
      },
      "outputs": [],
      "source": [
        "assert df.isna().sum()['extented_desc'] == 0 # check for nans in desc to make embedding collection"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "tqmoJhxuj65P"
      },
      "outputs": [],
      "source": [
        "df['tokens_in_ext_desc'] = df['extented_desc'].apply(num_tokens_from_string)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 447
        },
        "id": "LsrivA0craDn",
        "outputId": "d1973fa8-6c3d-4dc4-80d1-aeb922af8f87"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "<Axes: >"
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        },
        {
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAigAAAGdCAYAAAA44ojeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/SrBM8AAAACXBIWXMAAA9hAAAPYQGoP6dpAAAt+ElEQVR4nO3df3RU5YH/8U9+TCYEmMRQMyFCMNu6hQgKJUqmuNsuhkSaWpUcW/ymmLYcPaUBhWwR0wXkhxrNtmKxEWoPC/Yo65ZdsSUiZEAN6xIgBNnyw410yxorTLIrDQFSJkPmfv/oydQhIZOBJPOQvF/n5OA895k7z/30Ej6dmTsTZVmWJQAAAINER3oBAAAAl6KgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMExvpBVwJv9+vkydPavjw4YqKior0cgAAQA9YlqWzZ88qLS1N0dHdP0dyTRaUkydPavTo0ZFeBgAAuAIff/yxRo0a1e2ca7KgDB8+XNKfD9DhcERsHT6fT1VVVcrNzZXNZovYOkxFPqGRUWhk1D3yCY2MQuuvjFpaWjR69OjAv+PduSYLSsfLOg6HI+IFJSEhQQ6Hg5O+C+QTGhmFRkbdI5/QyCi0/s6oJ2/P4E2yAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4FBQAAGCcsArKjTfeqKioqE4/xcXFkqQLFy6ouLhYI0aM0LBhw1RQUKDGxsagfTQ0NCg/P18JCQlKSUnRokWLdPHixd47IgAAcM0Lq6DU1tbq1KlTgR+32y1Juv/++yVJCxcu1NatW7V582ZVV1fr5MmTmjlzZuD+7e3tys/PV1tbm/bs2aOXX35ZGzdu1LJly3rxkAAAwLUurIJy/fXXKzU1NfBTWVmpz3/+8/rKV76iM2fOaP369Xruuec0bdo0TZ48WRs2bNCePXu0d+9eSVJVVZWOHTumV155RRMnTtSMGTO0atUqVVRUqK2trU8OEAAAXHuu+NuM29ra9Morr6ikpERRUVGqq6uTz+dTTk5OYM7YsWOVnp6umpoaZWdnq6amRhMmTJDT6QzMycvL09y5c3X06FFNmjSpy8fyer3yer2B2y0tLZL+/O2LPp/vSg/hqnU8diTXYDLyCY2MQiOj7pFPaGQUWn9lFM7+r7igvPHGG2pubtZ3vvMdSZLH41FcXJySkpKC5jmdTnk8nsCcz5aTju0d2y6nrKxMK1as6DReVVWlhISEKz2EXtPxUhe6Rj6hkVFoZNQ98gmNjELr64xaW1t7PPeKC8r69es1Y8YMpaWlXekueqy0tFQlJSWB2y0tLRo9erRyc3PlcDj6/PEvx+fzye12a/r06bLZbBFbhySNX74joo/fFXu0pVVZfi09EC2vP6rT9iPL8yKwKrOYdA6Zioy6Rz6hkVFo/ZVRxysgPXFFBeWjjz7Szp079frrrwfGUlNT1dbWpubm5qBnURobG5WamhqYs3///qB9dVzl0zGnK3a7XXa7vdO4zWYz4mQzYR3e9s4FwBRef1SX64t0ZiYx4RwyHRl1j3xCI6PQ+jqjcPZ9RZ+DsmHDBqWkpCg/Pz8wNnnyZNlsNu3atSswVl9fr4aGBrlcLkmSy+XS4cOH1dTUFJjjdrvlcDiUmZl5JUsBAAADUNjPoPj9fm3YsEFFRUWKjf3L3RMTEzVnzhyVlJQoOTlZDodD8+fPl8vlUnZ2tiQpNzdXmZmZmj17tsrLy+XxeLRkyRIVFxd3+QwJAAAYnMIuKDt37lRDQ4O+973vddq2evVqRUdHq6CgQF6vV3l5eXrxxRcD22NiYlRZWam5c+fK5XJp6NChKioq0sqVK6/uKAAAwIASdkHJzc2VZVldbouPj1dFRYUqKioue/8xY8Zo27Zt4T4sAAAYRPguHgAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4FBQAAGAcCgoAADAOBQUAABiHggIAAIxDQQEAAMahoAAAAONQUAAAgHEoKAAAwDgUFAAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxwi4on3zyib797W9rxIgRGjJkiCZMmKADBw4EtluWpWXLlmnkyJEaMmSIcnJydPz48aB9nD59WoWFhXI4HEpKStKcOXN07ty5qz8aAAAwIIRVUP74xz9q6tSpstlseuutt3Ts2DH95Cc/0XXXXReYU15erjVr1mjdunXat2+fhg4dqry8PF24cCEwp7CwUEePHpXb7VZlZaV2796thx9+uPeOCgAAXNNiw5n87LPPavTo0dqwYUNgLCMjI/DflmXp+eef15IlS3TPPfdIkn75y1/K6XTqjTfe0KxZs/TBBx9o+/btqq2tVVZWliTphRde0Ne+9jX9+Mc/VlpaWm8cFwAAuIaFVVB+85vfKC8vT/fff7+qq6t1ww036Ac/+IEeeughSdKJEyfk8XiUk5MTuE9iYqKmTJmimpoazZo1SzU1NUpKSgqUE0nKyclRdHS09u3bp/vuu6/T43q9Xnm93sDtlpYWSZLP55PP5wvviHtRx2NHcg0d7DFWpJfQiT3aCvrzUibkFmkmnUOmIqPukU9oZBRaf2UUzv7DKii///3vtXbtWpWUlOhHP/qRamtr9cgjjyguLk5FRUXyeDySJKfTGXQ/p9MZ2ObxeJSSkhK8iNhYJScnB+ZcqqysTCtWrOg0XlVVpYSEhHAOoU+43e5IL0Hlt0d6BZe3Ksvf5fi2bdv6eSXmMuEcMh0ZdY98QiOj0Po6o9bW1h7PDaug+P1+ZWVl6emnn5YkTZo0SUeOHNG6detUVFQU3irDUFpaqpKSksDtlpYWjR49Wrm5uXI4HH32uKH4fD653W5Nnz5dNpstYuuQpPHLd0T08btij7a0KsuvpQei5fVHddp+ZHleBFZlFpPOIVORUffIJzQyCq2/Mup4BaQnwiooI0eOVGZmZtDYuHHj9G//9m+SpNTUVElSY2OjRo4cGZjT2NioiRMnBuY0NTUF7ePixYs6ffp04P6XstvtstvtncZtNpsRJ5sJ6/C2dy4ApvD6o7pcX6QzM4kJ55DpyKh75BMaGYXW1xmFs++wruKZOnWq6uvrg8Y+/PBDjRkzRtKf3zCbmpqqXbt2Bba3tLRo3759crlckiSXy6Xm5mbV1dUF5rz99tvy+/2aMmVKOMsBAAADVFjPoCxcuFBf/vKX9fTTT+ub3/ym9u/fr5deekkvvfSSJCkqKkoLFizQk08+qZtuukkZGRlaunSp0tLSdO+990r68zMud911lx566CGtW7dOPp9P8+bN06xZs7iCBwAASAqzoNx2223asmWLSktLtXLlSmVkZOj5559XYWFhYM5jjz2m8+fP6+GHH1Zzc7PuuOMObd++XfHx8YE5r776qubNm6c777xT0dHRKigo0Jo1a3rvqAAAwDUtrIIiSV//+tf19a9//bLbo6KitHLlSq1cufKyc5KTk7Vp06ZwHxoAAAwSfBcPAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4FBQAAGAcCgoAADAOBQUAABiHggIAAIxDQQEAAMahoAAAAONQUAAAgHEoKAAAwDgUFAAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4FBQAAGAcCgoAADAOBQUAABiHggIAAIxDQQEAAMahoAAAAONQUAAAgHEoKAAAwDhhFZTly5crKioq6Gfs2LGB7RcuXFBxcbFGjBihYcOGqaCgQI2NjUH7aGhoUH5+vhISEpSSkqJFixbp4sWLvXM0AABgQIgN9w4333yzdu7c+ZcdxP5lFwsXLtSbb76pzZs3KzExUfPmzdPMmTP1H//xH5Kk9vZ25efnKzU1VXv27NGpU6f04IMPymaz6emnn+6FwwEAAANB2AUlNjZWqampncbPnDmj9evXa9OmTZo2bZokacOGDRo3bpz27t2r7OxsVVVV6dixY9q5c6ecTqcmTpyoVatWafHixVq+fLni4uKu/ogAAMA1L+yCcvz4caWlpSk+Pl4ul0tlZWVKT09XXV2dfD6fcnJyAnPHjh2r9PR01dTUKDs7WzU1NZowYYKcTmdgTl5enubOnaujR49q0qRJXT6m1+uV1+sN3G5paZEk+Xw++Xy+cA+h13Q8diTX0MEeY0V6CZ3Yo62gPy9lQm6RZtI5ZCoy6h75hEZGofVXRuHsP6yCMmXKFG3cuFFf/OIXderUKa1YsUJ/8zd/oyNHjsjj8SguLk5JSUlB93E6nfJ4PJIkj8cTVE46tndsu5yysjKtWLGi03hVVZUSEhLCOYQ+4Xa7I70Eld8e6RVc3qosf5fj27Zt6+eVmMuEc8h0ZNQ98gmNjELr64xaW1t7PDesgjJjxozAf99yyy2aMmWKxowZo1/96lcaMmRIOLsKS2lpqUpKSgK3W1paNHr0aOXm5srhcPTZ44bi8/nkdrs1ffp02Wy2iK1DksYv3xHRx++KPdrSqiy/lh6Iltcf1Wn7keV5EViVWUw6h0xFRt0jn9DIKLT+yqjjFZCeCPslns9KSkrSX//1X+t3v/udpk+frra2NjU3Nwc9i9LY2Bh4z0pqaqr2798ftI+Oq3y6el9LB7vdLrvd3mncZrMZcbKZsA5ve+cCYAqvP6rL9UU6M5OYcA6Zjoy6Rz6hkVFofZ1ROPu+qs9BOXfunP77v/9bI0eO1OTJk2Wz2bRr167A9vr6ejU0NMjlckmSXC6XDh8+rKampsAct9sth8OhzMzMq1kKAAAYQMJ6BuWHP/yh7r77bo0ZM0YnT57UE088oZiYGD3wwANKTEzUnDlzVFJSouTkZDkcDs2fP18ul0vZ2dmSpNzcXGVmZmr27NkqLy+Xx+PRkiVLVFxc3OUzJAAAYHAKq6D84Q9/0AMPPKBPP/1U119/ve644w7t3btX119/vSRp9erVio6OVkFBgbxer/Ly8vTiiy8G7h8TE6PKykrNnTtXLpdLQ4cOVVFRkVauXNm7RwUAAK5pYRWU1157rdvt8fHxqqioUEVFxWXnjBkzhqs3AABAt/guHgAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4FBQAAGAcCgoAADAOBQUAABiHggIAAIxDQQEAAMahoAAAAONQUAAAgHEoKAAAwDgUFAAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjxEZ6ASa68fE3ezTPHmOp/HZp/PId8rZH9fGqAAAYPHgGBQAAGOeqCsozzzyjqKgoLViwIDB24cIFFRcXa8SIERo2bJgKCgrU2NgYdL+Ghgbl5+crISFBKSkpWrRokS5evHg1SwEAAAPIFReU2tpa/fznP9ctt9wSNL5w4UJt3bpVmzdvVnV1tU6ePKmZM2cGtre3tys/P19tbW3as2ePXn75ZW3cuFHLli278qMAAAADyhUVlHPnzqmwsFC/+MUvdN111wXGz5w5o/Xr1+u5557TtGnTNHnyZG3YsEF79uzR3r17JUlVVVU6duyYXnnlFU2cOFEzZszQqlWrVFFRoba2tt45KgAAcE27ojfJFhcXKz8/Xzk5OXryyScD43V1dfL5fMrJyQmMjR07Vunp6aqpqVF2drZqamo0YcIEOZ3OwJy8vDzNnTtXR48e1aRJkzo9ntfrldfrDdxuaWmRJPl8Pvl8vis5hG7ZY6yezYu2gv5EsFD59MX/dteajgzI4vLIqHvkExoZhdZfGYWz/7ALymuvvaaDBw+qtra20zaPx6O4uDglJSUFjTudTnk8nsCcz5aTju0d27pSVlamFStWdBqvqqpSQkJCuIcQUvnt4c1fleXv9TUMJJfLZ9u2bf28EnO53e5IL8F4ZNQ98gmNjELr64xaW1t7PDesgvLxxx/r0UcfldvtVnx8fNgLu1KlpaUqKSkJ3G5padHo0aOVm5srh8PR6483fvmOHs2zR1taleXX0gPR8vq5zPhSofI5sjwvAqsyi8/nk9vt1vTp02Wz2SK9HCORUffIJzQyCq2/Mup4BaQnwioodXV1ampq0pe+9KXAWHt7u3bv3q2f/exn2rFjh9ra2tTc3Bz0LEpjY6NSU1MlSampqdq/f3/Qfjuu8umYcym73S673d5p3Gaz9UmQ4X6midcfxeegdONy+fCL4i/66lweSMioe+QTGhmF1tcZhbPvsN4ke+edd+rw4cM6dOhQ4CcrK0uFhYWB/7bZbNq1a1fgPvX19WpoaJDL5ZIkuVwuHT58WE1NTYE5brdbDodDmZmZ4SwHAAAMUGE9gzJ8+HCNHz8+aGzo0KEaMWJEYHzOnDkqKSlRcnKyHA6H5s+fL5fLpezsbElSbm6uMjMzNXv2bJWXl8vj8WjJkiUqLi7u8lkSAAAw+PT6R92vXr1a0dHRKigokNfrVV5enl588cXA9piYGFVWVmru3LlyuVwaOnSoioqKtHLlyt5eCgAAuEZddUF59913g27Hx8eroqJCFRUVl73PmDFjuIIDAABcFt/FAwAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4FBQAAGCc2EgvAIPXjY+/GeklhO1/nsmP9BIAYFDgGRQAAGAcCgoAADAOBQUAABiHggIAAIxDQQEAAMahoAAAAONQUAAAgHEoKAAAwDgUFAAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4YRWUtWvX6pZbbpHD4ZDD4ZDL5dJbb70V2H7hwgUVFxdrxIgRGjZsmAoKCtTY2Bi0j4aGBuXn5yshIUEpKSlatGiRLl682DtHAwAABoSwCsqoUaP0zDPPqK6uTgcOHNC0adN0zz336OjRo5KkhQsXauvWrdq8ebOqq6t18uRJzZw5M3D/9vZ25efnq62tTXv27NHLL7+sjRs3atmyZb17VAAA4JoWG87ku+++O+j2U089pbVr12rv3r0aNWqU1q9fr02bNmnatGmSpA0bNmjcuHHau3evsrOzVVVVpWPHjmnnzp1yOp2aOHGiVq1apcWLF2v58uWKi4vrvSMDAADXrLAKyme1t7dr8+bNOn/+vFwul+rq6uTz+ZSTkxOYM3bsWKWnp6umpkbZ2dmqqanRhAkT5HQ6A3Py8vI0d+5cHT16VJMmTerysbxer7xeb+B2S0uLJMnn88nn813pIVyWPcbq2bxoK+hPBBuI+fT2+daxv744jwcKMuoe+YRGRqH1V0bh7D/sgnL48GG5XC5duHBBw4YN05YtW5SZmalDhw4pLi5OSUlJQfOdTqc8Ho8kyePxBJWTju0d2y6nrKxMK1as6DReVVWlhISEcA8hpPLbw5u/Ksvf62sYSAZSPtu2beuT/brd7j7Z70BCRt0jn9DIKLS+zqi1tbXHc8MuKF/84hd16NAhnTlzRv/6r/+qoqIiVVdXh7ubsJSWlqqkpCRwu6WlRaNHj1Zubq4cDkevP9745Tt6NM8ebWlVll9LD0TL64/q9XVc6wZiPkeW5/Xq/nw+n9xut6ZPny6bzdar+x4oyKh75BMaGYXWXxl1vALSE2EXlLi4OH3hC1+QJE2ePFm1tbX66U9/qm9961tqa2tTc3Nz0LMojY2NSk1NlSSlpqZq//79QfvruMqnY05X7Ha77HZ7p3GbzdYnQXrbw/vH1OuPCvs+g8lAyqev/uL21bk8kJBR98gnNDIKra8zCmffV/05KH6/X16vV5MnT5bNZtOuXbsC2+rr69XQ0CCXyyVJcrlcOnz4sJqamgJz3G63HA6HMjMzr3YpAABggAjrGZTS0lLNmDFD6enpOnv2rDZt2qR3331XO3bsUGJioubMmaOSkhIlJyfL4XBo/vz5crlcys7OliTl5uYqMzNTs2fPVnl5uTwej5YsWaLi4uIunyEBAACDU1gFpampSQ8++KBOnTqlxMRE3XLLLdqxY4emT58uSVq9erWio6NVUFAgr9ervLw8vfjii4H7x8TEqLKyUnPnzpXL5dLQoUNVVFSklStX9u5RAQCAa1pYBWX9+vXdbo+Pj1dFRYUqKiouO2fMmDF9diUEAAAYGPguHgAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4FBQAAGAcCgoAADAOBQUAABiHggIAAIxDQQEAAMahoAAAAONQUAAAgHEoKAAAwDgUFAAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxwiooZWVluu222zR8+HClpKTo3nvvVX19fdCcCxcuqLi4WCNGjNCwYcNUUFCgxsbGoDkNDQ3Kz89XQkKCUlJStGjRIl28ePHqjwYAAAwIYRWU6upqFRcXa+/evXK73fL5fMrNzdX58+cDcxYuXKitW7dq8+bNqq6u1smTJzVz5szA9vb2duXn56utrU179uzRyy+/rI0bN2rZsmW9d1QAAOCaFhvO5O3btwfd3rhxo1JSUlRXV6e//du/1ZkzZ7R+/Xpt2rRJ06ZNkyRt2LBB48aN0969e5Wdna2qqiodO3ZMO3fulNPp1MSJE7Vq1SotXrxYy5cvV1xcXO8dHQAAuCZd1XtQzpw5I0lKTk6WJNXV1cnn8yknJycwZ+zYsUpPT1dNTY0kqaamRhMmTJDT6QzMycvLU0tLi44ePXo1ywEAAANEWM+gfJbf79eCBQs0depUjR8/XpLk8XgUFxenpKSkoLlOp1Mejycw57PlpGN7x7aueL1eeb3ewO2WlhZJks/nk8/nu9JDuCx7jNWzedFW0J8INhDz6e3zrWN/fXEeDxRk1D3yCY2MQuuvjMLZ/xUXlOLiYh05ckTvvffele6ix8rKyrRixYpO41VVVUpISOj1xyu/Pbz5q7L8vb6GgWQg5bNt27Y+2a/b7e6T/Q4kZNQ98gmNjELr64xaW1t7PPeKCsq8efNUWVmp3bt3a9SoUYHx1NRUtbW1qbm5OehZlMbGRqWmpgbm7N+/P2h/HVf5dMy5VGlpqUpKSgK3W1paNHr0aOXm5srhcFzJIXRr/PIdPZpnj7a0KsuvpQei5fVH9fo6rnUDMZ8jy/N6dX8+n09ut1vTp0+XzWbr1X0PFGTUPfIJjYxC66+MOl4B6YmwCoplWZo/f762bNmid999VxkZGUHbJ0+eLJvNpl27dqmgoECSVF9fr4aGBrlcLkmSy+XSU089paamJqWkpEj6c2NzOBzKzMzs8nHtdrvsdnuncZvN1idBetvD+8fU648K+z6DyUDKp6/+4vbVuTyQkFH3yCc0MgqtrzMKZ99hFZTi4mJt2rRJv/71rzV8+PDAe0YSExM1ZMgQJSYmas6cOSopKVFycrIcDofmz58vl8ul7OxsSVJubq4yMzM1e/ZslZeXy+PxaMmSJSouLu6yhAAAgMEnrIKydu1aSdJXv/rVoPENGzboO9/5jiRp9erVio6OVkFBgbxer/Ly8vTiiy8G5sbExKiyslJz586Vy+XS0KFDVVRUpJUrV17dkQAAgAEj7Jd4QomPj1dFRYUqKiouO2fMmDF99mZDAABw7eO7eAAAgHEoKAAAwDgUFAAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4FBQAAGAcCgoAADAOBQUAABiHggIAAIxDQQEAAMahoAAAAONQUAAAgHEoKAAAwDgUFAAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjhF1Qdu/erbvvvltpaWmKiorSG2+8EbTdsiwtW7ZMI0eO1JAhQ5STk6Pjx48HzTl9+rQKCwvlcDiUlJSkOXPm6Ny5c1d1IAAAYOAIu6CcP39et956qyoqKrrcXl5erjVr1mjdunXat2+fhg4dqry8PF24cCEwp7CwUEePHpXb7VZlZaV2796thx9++MqPAgAADCix4d5hxowZmjFjRpfbLMvS888/ryVLluiee+6RJP3yl7+U0+nUG2+8oVmzZumDDz7Q9u3bVVtbq6ysLEnSCy+8oK997Wv68Y9/rLS0tKs4HKBv3fj4m726P3uMpfLbpfHLd8jbHtWr++7wP8/k98l+AaAvhV1QunPixAl5PB7l5OQExhITEzVlyhTV1NRo1qxZqqmpUVJSUqCcSFJOTo6io6O1b98+3XfffZ326/V65fV6A7dbWlokST6fTz6frzcPQdKf/9Ho0bxoK+hPBCOf0Pojo774O9KfOtZ/rR9HXyGf0MgotP7KKJz992pB8Xg8kiSn0xk07nQ6A9s8Ho9SUlKCFxEbq+Tk5MCcS5WVlWnFihWdxquqqpSQkNAbSw9Sfnt481dl+Xt9DQMJ+YTWlxlt27atz/bdn9xud6SXYDTyCY2MQuvrjFpbW3s8t1cLSl8pLS1VSUlJ4HZLS4tGjx6t3NxcORyOXn+88ct39GiePdrSqiy/lh6IltffN0/PX8vIJ7T+yOjI8rw+2W9/8fl8crvdmj59umw2W6SXYxzyCY2MQuuvjDpeAemJXi0oqampkqTGxkaNHDkyMN7Y2KiJEycG5jQ1NQXd7+LFizp9+nTg/pey2+2y2+2dxm02W58EGe57Abz+qD57/8BAQD6h9WVGA+UXcl/9fR8oyCc0MgqtrzMKZ9+9+jkoGRkZSk1N1a5duwJjLS0t2rdvn1wulyTJ5XKpublZdXV1gTlvv/22/H6/pkyZ0pvLAQAA16iwn0E5d+6cfve73wVunzhxQocOHVJycrLS09O1YMECPfnkk7rpppuUkZGhpUuXKi0tTffee68kady4cbrrrrv00EMPad26dfL5fJo3b55mzZrFFTwAAEDSFRSUAwcO6O/+7u8CtzveG1JUVKSNGzfqscce0/nz5/Xwww+rublZd9xxh7Zv3674+PjAfV599VXNmzdPd955p6Kjo1VQUKA1a9b0wuEAAICBIOyC8tWvflWWdflLIqOiorRy5UqtXLnysnOSk5O1adOmcB8aAAAMEnwXDwAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDgUFAAAYh4ICAACMQ0EBAADGoaAAAADjUFAAAIBxKCgAAMA4FBQAAGAcCgoAADAOBQUAABiHggIAAIxDQQEAAMahoAAAAONQUAAAgHEoKAAAwDgUFAAAYJzYSC8AQN+68fE3I72EsP3PM/mRXgKACOMZFAAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwDh91D8A4n/14fnuMpfLbpfHLd8jbHhXBVXWPj+cHehfPoAAAAONQUAAAgHEoKAAAwDgRLSgVFRW68cYbFR8frylTpmj//v2RXA4AADBExArKv/zLv6ikpERPPPGEDh48qFtvvVV5eXlqamqK1JIAAIAhInYVz3PPPaeHHnpI3/3udyVJ69at05tvvql/+qd/0uOPPx6pZQHAFfnslUf96WqucuLKI5gsIgWlra1NdXV1Ki0tDYxFR0crJydHNTU1neZ7vV55vd7A7TNnzkiSTp8+LZ/P1+vri714vmfz/JZaW/2K9UWr3W/u5Y+RQj6hkVFoZNS9q8nnCz/8VR+tqu/sK70z7Pv4fD61trbq008/lc1m64NVXfv6K6OzZ89KkizLCjk3IgXl//7v/9Te3i6n0xk07nQ69V//9V+d5peVlWnFihWdxjMyMvpsjT31/yK9AMORT2hkFBoZdW8w5fO5n0R6BegNZ8+eVWJiYrdzrokPaistLVVJSUngtt/v1+nTpzVixAhFRUXu/1G1tLRo9OjR+vjjj+VwOCK2DlORT2hkFBoZdY98QiOj0PorI8uydPbsWaWlpYWcG5GC8rnPfU4xMTFqbGwMGm9sbFRqamqn+Xa7XXa7PWgsKSmpL5cYFofDwUnfDfIJjYxCI6PukU9oZBRaf2QU6pmTDhG5iicuLk6TJ0/Wrl27AmN+v1+7du2Sy+WKxJIAAIBBIvYST0lJiYqKipSVlaXbb79dzz//vM6fPx+4qgcAAAxeESso3/rWt/S///u/WrZsmTwejyZOnKjt27d3euOsyex2u5544olOLz/hz8gnNDIKjYy6Rz6hkVFoJmYUZfXkWh8AAIB+xHfxAAAA41BQAACAcSgoAADAOBQUAABgHApKCGVlZbrttts0fPhwpaSk6N5771V9fX3QnAsXLqi4uFgjRozQsGHDVFBQ0OlD6AaLZ555RlFRUVqwYEFgjHykTz75RN/+9rc1YsQIDRkyRBMmTNCBAwcC2y3L0rJlyzRy5EgNGTJEOTk5On78eARX3L/a29u1dOlSZWRkaMiQIfr85z+vVatWBX1fx2DLaPfu3br77ruVlpamqKgovfHGG0Hbe5LH6dOnVVhYKIfDoaSkJM2ZM0fnzp3rx6PoW91l5PP5tHjxYk2YMEFDhw5VWlqaHnzwQZ08eTJoHwM5o1Dn0Gd9//vfV1RUlJ5//vmg8UjmQ0EJobq6WsXFxdq7d6/cbrd8Pp9yc3N1/vxfvlBw4cKF2rp1qzZv3qzq6mqdPHlSM2fOjOCqI6O2tlY///nPdcsttwSND/Z8/vjHP2rq1Kmy2Wx66623dOzYMf3kJz/RddddF5hTXl6uNWvWaN26ddq3b5+GDh2qvLw8XbhwIYIr7z/PPvus1q5dq5/97Gf64IMP9Oyzz6q8vFwvvPBCYM5gy+j8+fO69dZbVVFR0eX2nuRRWFioo0ePyu12q7KyUrt379bDDz/cX4fQ57rLqLW1VQcPHtTSpUt18OBBvf7666qvr9c3vvGNoHkDOaNQ51CHLVu2aO/evV1+/HxE87EQlqamJkuSVV1dbVmWZTU3N1s2m83avHlzYM4HH3xgSbJqamoitcx+d/bsWeumm26y3G639ZWvfMV69NFHLcsiH8uyrMWLF1t33HHHZbf7/X4rNTXV+sd//MfAWHNzs2W3261//ud/7o8lRlx+fr71ve99L2hs5syZVmFhoWVZZCTJ2rJlS+B2T/I4duyYJcmqra0NzHnrrbesqKgo65NPPum3tfeXSzPqyv79+y1J1kcffWRZ1uDK6HL5/OEPf7BuuOEG68iRI9aYMWOs1atXB7ZFOh+eQQnTmTNnJEnJycmSpLq6Ovl8PuXk5ATmjB07Vunp6aqpqYnIGiOhuLhY+fn5QTlI5CNJv/nNb5SVlaX7779fKSkpmjRpkn7xi18Etp84cUIejycoo8TERE2ZMmXQZPTlL39Zu3bt0ocffihJ+s///E+99957mjFjhiQyulRP8qipqVFSUpKysrICc3JychQdHa19+/b1+5pNcObMGUVFRQW+y22wZ+T3+zV79mwtWrRIN998c6ftkc7nmvg2Y1P4/X4tWLBAU6dO1fjx4yVJHo9HcXFxnb680Ol0yuPxRGCV/e+1117TwYMHVVtb22kb+Ui///3vtXbtWpWUlOhHP/qRamtr9cgjjyguLk5FRUWBHC79FOXBlNHjjz+ulpYWjR07VjExMWpvb9dTTz2lwsJCSSKjS/QkD4/Ho5SUlKDtsbGxSk5OHpSZXbhwQYsXL9YDDzwQ+DK8wZ7Rs88+q9jYWD3yyCNdbo90PhSUMBQXF+vIkSN67733Ir0UY3z88cd69NFH5Xa7FR8fH+nlGMnv9ysrK0tPP/20JGnSpEk6cuSI1q1bp6Kiogivzgy/+tWv9Oqrr2rTpk26+eabdejQIS1YsEBpaWlkhKvm8/n0zW9+U5Zlae3atZFejhHq6ur005/+VAcPHlRUVFSkl9MlXuLpoXnz5qmyslLvvPOORo0aFRhPTU1VW1ubmpubg+Y3NjYqNTW1n1fZ/+rq6tTU1KQvfelLio2NVWxsrKqrq7VmzRrFxsbK6XQO6nwkaeTIkcrMzAwaGzdunBoaGiQpkMOlVzYNpowWLVqkxx9/XLNmzdKECRM0e/ZsLVy4UGVlZZLI6FI9ySM1NVVNTU1B2y9evKjTp08Pqsw6yslHH30kt9sdePZEGtwZ/fu//7uampqUnp4e+N390Ucf6e///u914403Sop8PhSUECzL0rx587Rlyxa9/fbbysjICNo+efJk2Ww27dq1KzBWX1+vhoYGuVyu/l5uv7vzzjt1+PBhHTp0KPCTlZWlwsLCwH8P5nwkaerUqZ0uTf/www81ZswYSVJGRoZSU1ODMmppadG+ffsGTUatra2Kjg7+dRQTEyO/3y+JjC7VkzxcLpeam5tVV1cXmPP222/L7/drypQp/b7mSOgoJ8ePH9fOnTs1YsSIoO2DOaPZs2frt7/9bdDv7rS0NC1atEg7duyQZEA+ff423Gvc3LlzrcTEROvdd9+1Tp06FfhpbW0NzPn+979vpaenW2+//bZ14MABy+VyWS6XK4KrjqzPXsVjWeSzf/9+KzY21nrqqaes48ePW6+++qqVkJBgvfLKK4E5zzzzjJWUlGT9+te/tn77299a99xzj5WRkWH96U9/iuDK+09RUZF1ww03WJWVldaJEyes119/3frc5z5nPfbYY4E5gy2js2fPWu+//771/vvvW5Ks5557znr//fcDV6D0JI+77rrLmjRpkrVv3z7rvffes2666SbrgQceiNQh9bruMmpra7O+8Y1vWKNGjbIOHToU9Pvb6/UG9jGQMwp1Dl3q0qt4LCuy+VBQQpDU5c+GDRsCc/70pz9ZP/jBD6zrrrvOSkhIsO677z7r1KlTkVt0hF1aUMjHsrZu3WqNHz/estvt1tixY62XXnopaLvf77eWLl1qOZ1Oy263W3feeadVX18fodX2v5aWFuvRRx+10tPTrfj4eOuv/uqvrH/4h38I+odksGX0zjvvdPm7p6ioyLKsnuXx6aefWg888IA1bNgwy+FwWN/97nets2fPRuBo+kZ3GZ04ceKyv7/feeedwD4GckahzqFLdVVQIplPlGV95qMaAQAADMB7UAAAgHEoKAAAwDgUFAAAYBwKCgAAMA4FBQAAGIeCAgAAjENBAQAAxqGgAAAA41BQAACAcSgoAADAOBQUAABgHAoKAAAwzv8HUAviDnCkUMIAAAAASUVORK5CYII=",
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "df['tokens_in_ext_desc'].hist()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 85,
      "metadata": {},
      "outputs": [],
      "source": [
        "df['extented_desc_readme_trim'] = df['extented_desc_readme'].apply(lambda x: x[:22000] if pd.notna(x) else x)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 86,
      "metadata": {},
      "outputs": [],
      "source": [
        "df['tokens_in_ext_desc_readme_trim'] = df['extented_desc_readme_trim'].apply(num_tokens_from_string)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 87,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "<Axes: >"
            ]
          },
          "execution_count": 87,
          "metadata": {},
          "output_type": "execute_result"
        },
        {
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjQAAAGdCAYAAAAFcOm4AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/SrBM8AAAACXBIWXMAAA9hAAAPYQGoP6dpAAApIUlEQVR4nO3dfXRU5YHH8d8kmQwJMISAedMEEF8A5SUFCdNSFyUkQA6WytkjmlpEFo5s6IqxCFiFoFUo7bFWDsK6a8E9EqnuESyIQASBsgaUrAgBTwREsUJCC5sEiIaBPPuHm7uMCUJg8vLMfD/nzNG598m9zy8Tk5/3zr3jMsYYAQAAWCyitScAAABwtSg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrRbX2BK5EXV2djh49qo4dO8rlcrX2dAAAwGUwxujUqVNKSUlRRERwj6lYWWiOHj2q1NTU1p4GAAC4Al9++aWuu+66oG7TykLTsWNHSd9+Q7xeb1C26ff7tXHjRmVlZcntdgdlm20VWUNPuOSUyBqqwiVruOSUGs9aXV2t1NRU5+94MFlZaOpPM3m93qAWmtjYWHm93rD4ISNraAmXnBJZQ1W4ZA2XnNL3Z22Ot4vwpmAAAGA9Cg0AALAehQYAAFivSYVm/vz5uu2229SxY0clJCRo7NixKisrCxgzbNgwuVyugMdDDz0UMObIkSPKyclRbGysEhISNGPGDJ07d+7q0wAAgLDUpDcFb926VXl5ebrtttt07tw5Pf7448rKytL+/fvVvn17Z9zkyZP11FNPOc9jY2Odfz9//rxycnKUlJSk999/X8eOHdPPf/5zud1uPfvss0GIBAAAwk2TCs369esDni9fvlwJCQkqKSnR7bff7iyPjY1VUlJSo9vYuHGj9u/fr3fffVeJiYkaMGCAnn76ac2cOVMFBQWKjo6+ghgAACCcXdVl21VVVZKk+Pj4gOUrVqzQq6++qqSkJI0ZM0ZPPvmkc5SmuLhYffv2VWJiojM+OztbU6dO1b59+5Sent5gP7W1taqtrXWeV1dXS/r2kjC/3381ERz12wnW9toysoaecMkpkTVUhUvWcMkpNZ61OXO7jDHmSr6wrq5Od911lyorK7V9+3Zn+UsvvaRu3bopJSVFe/bs0cyZMzV48GC9+eabkqQpU6boiy++0IYNG5yvqampUfv27bVu3TqNGjWqwb4KCgo0b968BssLCwsDTmcBAIC2q6amRvfdd5+qqqqCdh+5eld8hCYvL0+lpaUBZUb6trDU69u3r5KTkzV8+HAdOnRIPXv2vKJ9zZ49W/n5+c7z+jsNZmVlBfXGekVFRRoxYkRY3OyIrKElXHJKZA1V4ZI1XHJKjWetP8PSHK6o0EybNk1r167Vtm3bLvlZDBkZGZKkgwcPqmfPnkpKStIHH3wQMKaiokKSLvq+G4/HI4/H02C52+0O+g9Ec2yzrSJr6AmXnBJZQ1W4ZA2XnFJg1ubM3KTLto0xmjZtmlatWqXNmzerR48el/ya3bt3S5KSk5MlST6fT3v37tXx48edMUVFRfJ6verTp09TpgMAACCpiUdo8vLyVFhYqLfeeksdO3ZUeXm5JKlTp06KiYnRoUOHVFhYqNGjR6tLly7as2ePHnnkEd1+++3q16+fJCkrK0t9+vTR/fffr4ULF6q8vFxPPPGE8vLyGj0KAwAAcClNOkKzZMkSVVVVadiwYUpOTnYef/rTnyRJ0dHRevfdd5WVlaVevXrp0Ucf1bhx47RmzRpnG5GRkVq7dq0iIyPl8/n0s5/9TD//+c8D7lsDAADQFE06QnOpC6JSU1O1devWS26nW7duWrduXVN2DQAAcFFXdR+aUNV91tuXHPP5gpwWmAkAALgcfDglAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrUWgAAID1KDQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArEehAQAA1qPQAAAA61FoAACA9Sg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrUWgAAID1KDQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArEehAQAA1qPQAAAA61FoAACA9Sg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrUWgAAID1KDQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANZrUqGZP3++brvtNnXs2FEJCQkaO3asysrKAsZ88803ysvLU5cuXdShQweNGzdOFRUVAWOOHDminJwcxcbGKiEhQTNmzNC5c+euPg0AAAhLTSo0W7duVV5ennbs2KGioiL5/X5lZWXpzJkzzphHHnlEa9as0RtvvKGtW7fq6NGjuvvuu53158+fV05Ojs6ePav3339fr7zyipYvX645c+YELxUAAAgrUU0ZvH79+oDny5cvV0JCgkpKSnT77berqqpKL7/8sgoLC3XnnXdKkpYtW6bevXtrx44dGjJkiDZu3Kj9+/fr3XffVWJiogYMGKCnn35aM2fOVEFBgaKjo4OXDgAAhIUmFZrvqqqqkiTFx8dLkkpKSuT3+5WZmemM6dWrl9LS0lRcXKwhQ4aouLhYffv2VWJiojMmOztbU6dO1b59+5Sent5gP7W1taqtrXWeV1dXS5L8fr/8fv/VRHDUb8fv98sTaS57vI0uzBrqwiVruOSUyBqqwiVruOSUGs/anLldxphL//VuRF1dne666y5VVlZq+/btkqTCwkJNnDgxoHxI0uDBg3XHHXfoN7/5jaZMmaIvvvhCGzZscNbX1NSoffv2WrdunUaNGtVgXwUFBZo3b16D5YWFhYqNjb2S6QMAgBZWU1Oj++67T1VVVfJ6vUHd9hUfocnLy1NpaalTZprT7NmzlZ+f7zyvrq5WamqqsrKygvYN8fv9Kioq0ogRI5T+zOZLji8tyA7KflvDhVndbndrT6dZhUvWcMkpkTVUhUvWcMkpNZ61/gxLc7iiQjNt2jStXbtW27Zt03XXXecsT0pK0tmzZ1VZWam4uDhneUVFhZKSkpwxH3zwQcD26q+Cqh/zXR6PRx6Pp8Fyt9sd9B8It9ut2vOuyxpnu+b4/rVV4ZI1XHJKZA1V4ZI1XHJKgVmbM3OTrnIyxmjatGlatWqVNm/erB49egSsHzhwoNxutzZt2uQsKysr05EjR+Tz+SRJPp9Pe/fu1fHjx50xRUVF8nq96tOnz9VkAQAAYapJR2jy8vJUWFiot956Sx07dlR5ebkkqVOnToqJiVGnTp00adIk5efnKz4+Xl6vV7/4xS/k8/k0ZMgQSVJWVpb69Omj+++/XwsXLlR5ebmeeOIJ5eXlNXoUBgAA4FKaVGiWLFkiSRo2bFjA8mXLlumBBx6QJP3+979XRESExo0bp9raWmVnZ+vFF190xkZGRmrt2rWaOnWqfD6f2rdvrwkTJuipp566uiQAACBsNanQXM4FUe3atdPixYu1ePHii47p1q2b1q1b15RdAwAAXBSf5QQAAKxHoQEAANaj0AAAAOtRaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArEehAQAA1qPQAAAA61FoAACA9Sg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrUWgAAID1KDQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArEehAQAA1qPQAAAA61FoAACA9Sg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrUWgAAID1KDQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArEehAQAA1otq7QmEu+6z3r7kmM8X5LTATAAAsBdHaAAAgPUoNAAAwHpNLjTbtm3TmDFjlJKSIpfLpdWrVwesf+CBB+RyuQIeI0eODBhz8uRJ5ebmyuv1Ki4uTpMmTdLp06evKggAAAhfTS40Z86cUf/+/bV48eKLjhk5cqSOHTvmPF577bWA9bm5udq3b5+Kioq0du1abdu2TVOmTGn67AEAAHQFbwoeNWqURo0a9b1jPB6PkpKSGl33ySefaP369frwww81aNAgSdKiRYs0evRo/e53v1NKSkpTpwQAAMJcs1zltGXLFiUkJKhz586688479etf/1pdunSRJBUXFysuLs4pM5KUmZmpiIgI7dy5Uz/96U8bbK+2tla1tbXO8+rqakmS3++X3+8Pypzrt+P3++WJNJc9/mq15L6+u71gb7ctCpes4ZJTImuoCpes4ZJTajxrc+Z2GWMu/Rf1Yl/scmnVqlUaO3ass2zlypWKjY1Vjx49dOjQIT3++OPq0KGDiouLFRkZqWeffVavvPKKysrKAraVkJCgefPmaerUqQ32U1BQoHnz5jVYXlhYqNjY2CudPgAAaEE1NTW67777VFVVJa/XG9RtB/0Izfjx451/79u3r/r166eePXtqy5YtGj58+BVtc/bs2crPz3eeV1dXKzU1VVlZWUH7hvj9fhUVFWnEiBFKf2bzJceXFmQHZb+3FmxosX3VuzCr2+0O6rbbmnDJGi45JbKGqnDJGi45pcaz1p9haQ7NfmO966+/Xl27dtXBgwc1fPhwJSUl6fjx4wFjzp07p5MnT170fTcej0cej6fBcrfbHfQfCLfbrdrzrssaFwwtua/Gthvq/0HVC5es4ZJTImuoCpes4ZJTCszanJmbvdD89a9/1YkTJ5ScnCxJ8vl8qqysVElJiQYOHChJ2rx5s+rq6pSRkdHc02lRl3MXYAAAcPWaXGhOnz6tgwcPOs8PHz6s3bt3Kz4+XvHx8Zo3b57GjRunpKQkHTp0SI899phuuOEGZWd/e9qkd+/eGjlypCZPnqylS5fK7/dr2rRpGj9+PFc4AQCAK9Lk+9Ds2rVL6enpSk9PlyTl5+crPT1dc+bMUWRkpPbs2aO77rpLN910kyZNmqSBAwfqL3/5S8ApoxUrVqhXr14aPny4Ro8eraFDh+qll14KXioAABBWmnyEZtiwYfq+C6M2bLj0m1zj4+NVWFjY1F0DAAA0is9yAgAA1qPQAAAA61FoAACA9Sg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrUWgAAID1KDQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWC+qtSeAS+s+6+1Ljvl8QU4LzAQAgLaJIzQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPUoNAAAwHrcKfgKXc7dewEAQMvgCA0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArEehAQAA1qPQAAAA6/Fp2yHicj79+/MFOS0wEwAAWh5HaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArEehAQAA1qPQAAAA61FoAACA9Sg0AADAehQaAABgPQoNAACwXpMLzbZt2zRmzBilpKTI5XJp9erVAeuNMZozZ46Sk5MVExOjzMxMHThwIGDMyZMnlZubK6/Xq7i4OE2aNEmnT5++qiAAACB8NbnQnDlzRv3799fixYsbXb9w4UK98MILWrp0qXbu3Kn27dsrOztb33zzjTMmNzdX+/btU1FRkdauXatt27ZpypQpV54CAACEtaimfsGoUaM0atSoRtcZY/T888/riSee0E9+8hNJ0n/8x38oMTFRq1ev1vjx4/XJJ59o/fr1+vDDDzVo0CBJ0qJFizR69Gj97ne/U0pKylXEAQAA4Sio76E5fPiwysvLlZmZ6Szr1KmTMjIyVFxcLEkqLi5WXFycU2YkKTMzUxEREdq5c2cwpwMAAMJEk4/QfJ/y8nJJUmJiYsDyxMREZ115ebkSEhICJxEVpfj4eGfMd9XW1qq2ttZ5Xl1dLUny+/3y+/1BmXv9dvx+vzyRJijbbGsuzHjhP0NZuGQNl5wSWUNVuGQNl5xS41mbM3dQC01zmT9/vubNm9dg+caNGxUbGxvUfRUVFWnh4KBuss1Yt25dwPOioqJWmknLC5es4ZJTImuoCpes4ZJTCsxaU1PTbPsJaqFJSkqSJFVUVCg5OdlZXlFRoQEDBjhjjh8/HvB1586d08mTJ52v/67Zs2crPz/feV5dXa3U1FRlZWXJ6/UGZe5+v19FRUUaMWKE0p/ZHJRttjWlBdmSArO63e5WnlXzCpes4ZJTImuoCpes4ZJTajxr/RmW5hDUQtOjRw8lJSVp06ZNToGprq7Wzp07NXXqVEmSz+dTZWWlSkpKNHDgQEnS5s2bVVdXp4yMjEa36/F45PF4Gix3u91B/4Fwu92qPe8K6jbbiu9+r5rj+9dWhUvWcMkpkTVUhUvWcMkpBWZtzsxNLjSnT5/WwYMHneeHDx/W7t27FR8fr7S0NE2fPl2//vWvdeONN6pHjx568sknlZKSorFjx0qSevfurZEjR2ry5MlaunSp/H6/pk2bpvHjx3OFEwAAuCJNLjS7du3SHXfc4TyvPxU0YcIELV++XI899pjOnDmjKVOmqLKyUkOHDtX69evVrl0752tWrFihadOmafjw4YqIiNC4ceP0wgsvBCEOAAAIR00uNMOGDZMxF78KyOVy6amnntJTTz110THx8fEqLCxs6q4BAAAaxWc5AQAA61FoAACA9Sg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrBfXDKdG2dZ/1tiTJE2m0cLB0a8GGBh/E+fmCnNaYGgAAV4UjNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArEehAQAA1qPQAAAA61FoAACA9Sg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrUWgAAID1KDQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPWiWnsCaFu6z3r7kmM+X5DTAjMBAODycYQGAABYj0IDAACsR6EBAADWo9AAAADrUWgAAID1KDQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPX4tG00GZ/IDQBoazhCAwAArEehAQAA1qPQAAAA61FoAACA9Sg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsx52C0Sy4mzAAoCVxhAYAAFiPQgMAAKwX9EJTUFAgl8sV8OjVq5ez/ptvvlFeXp66dOmiDh06aNy4caqoqAj2NAAAQBhpliM0t9xyi44dO+Y8tm/f7qx75JFHtGbNGr3xxhvaunWrjh49qrvvvrs5pgEAAMJEs7wpOCoqSklJSQ2WV1VV6eWXX1ZhYaHuvPNOSdKyZcvUu3dv7dixQ0OGDGmO6QAAgBDXLIXmwIEDSklJUbt27eTz+TR//nylpaWppKREfr9fmZmZzthevXopLS1NxcXFFy00tbW1qq2tdZ5XV1dLkvx+v/x+f1DmXL8dv98vT6QJyjbbKk+ECfhnawnWa3c5+2iJfbWmcMkpkTVUhUvWcMkpNZ61OXO7jDFB/av2zjvv6PTp07r55pt17NgxzZs3T1999ZVKS0u1Zs0aTZw4MaCcSNLgwYN1xx136De/+U2j2ywoKNC8efMaLC8sLFRsbGwwpw8AAJpJTU2N7rvvPlVVVcnr9QZ120EvNN9VWVmpbt266bnnnlNMTMwVFZrGjtCkpqbq73//e9C+IX6/X0VFRRoxYoTSn9kclG22VZ4Io6cH1enJXRGqrXO12jxKC7KbfR8Xvq5ut7vZ99dawiWnRNZQFS5ZwyWn1HjW6upqde3atVkKTbPfWC8uLk433XSTDh48qBEjRujs2bOqrKxUXFycM6aioqLR99zU83g88ng8DZa73e6g/0C43W7Vnm+9P/ItqbbO1apZb3xy4yXHBOvme83xs9IWhUtOiayhKlyyhktOKTBrc2Zu9vvQnD59WocOHVJycrIGDhwot9utTZs2OevLysp05MgR+Xy+5p4KAAAIUUE/QvPLX/5SY8aMUbdu3XT06FHNnTtXkZGRuvfee9WpUydNmjRJ+fn5io+Pl9fr1S9+8Qv5fD6ucAIAAFcs6IXmr3/9q+69916dOHFC11xzjYYOHaodO3bommuukST9/ve/V0REhMaNG6fa2lplZ2frxRdfDPY0AABAGAl6oVm5cuX3rm/Xrp0WL16sxYsXB3vXAAAgTPFZTgAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArNfsdwoGrkb3WW9f1dd7Io0WDg7SZAAAbRZHaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI+rnBAWbi3YoNrzru8d8/mCnBaaDQAg2DhCAwAArEehAQAA1qPQAAAA61FoAACA9Sg0AADAehQaAABgPQoNAACwHoUGAABYj0IDAACsR6EBAADWo9AAAADrUWgAAID1KDQAAMB6FBoAAGA9Cg0AALAehQYAAFiPQgMAAKxHoQEAANaj0AAAAOtFtfYEAJt0n/X2Jcd8viCnBWYCALgQhQYIspYsPfX78kQaLRws3VqwQbXnXc2yLwBoyzjlBAAArEehAQAA1qPQAAAA61FoAACA9XhTMPB/LufNvACAtokjNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI83BQMIGj4aAkBr4QgNAACwHkdogBDHURMA4YAjNAAAwHoUGgAAYD1OOQHgtBQA63GEBgAAWI8jNEAb1dY+W6qtzefWgg2qPe/63jGhelSJI2pAQxyhAQAA1uMIDQCEII7iINxQaIBW0NZO36DtuPBnwxNptHDw5Z1eA8Idp5wAAID1OEIDALgoTl3BFhQaAC3qak+31Z+GAYALccoJAABYjyM0ABAEnJoBWheFBgAugavSvl+olrlQzRWqOOUEAACsxxEaAGEtnI++tMXsl7rnDkdEcDGtWmgWL16s3/72tyovL1f//v21aNEiDR7M5QsAgCvHqaLw1GqF5k9/+pPy8/O1dOlSZWRk6Pnnn1d2drbKysqUkJDQWtMCEELa2hGItjYftB2UsKvXau+hee655zR58mRNnDhRffr00dKlSxUbG6s//vGPrTUlAABgqVY5QnP27FmVlJRo9uzZzrKIiAhlZmaquLi4wfja2lrV1tY6z6uqqiRJJ0+elN/vD8qc/H6/ampqdOLECUWdOxOUbbZVUXVGNTV1ivJH6HxdaH8+TLhkDZecElltdcMvX//e9Z4IoyfSL531UtuRLu8PW0tuZ+fs4c6/X/i3xu12//++LuPvzokTJy5jRpcnY/6mS465cN5XorGsp06dkiQZY65q240yreCrr74yksz7778fsHzGjBlm8ODBDcbPnTvXSOLBgwcPHjx4hMDjyy+/DHq3sOIqp9mzZys/P995XldXp5MnT6pLly5yuYLzfy3V1dVKTU3Vl19+Ka/XG5RttlVkDT3hklMia6gKl6zhklNqPKsxRqdOnVJKSkrQ99cqhaZr166KjIxURUVFwPKKigolJSU1GO/xeOTxeAKWxcXFNcvcvF5vyP+Q1SNr6AmXnBJZQ1W4ZA2XnFLDrJ06dWqW/bTKm4Kjo6M1cOBAbdr0/+fw6urqtGnTJvl8vtaYEgAAsFirnXLKz8/XhAkTNGjQIA0ePFjPP/+8zpw5o4kTJ7bWlAAAgKVardDcc889+tvf/qY5c+aovLxcAwYM0Pr165WYmNgq8/F4PJo7d26DU1uhiKyhJ1xySmQNVeGSNVxySi2f1WVMc1w7BQAA0HL4cEoAAGA9Cg0AALAehQYAAFiPQgMAAKxHofk/ixcvVvfu3dWuXTtlZGTogw8+aO0pfa9t27ZpzJgxSklJkcvl0urVqwPWG2M0Z84cJScnKyYmRpmZmTpw4EDAmJMnTyo3N1der1dxcXGaNGmSTp8+HTBmz549+vGPf6x27dopNTVVCxcubO5oAebPn6/bbrtNHTt2VEJCgsaOHauysrKAMd98843y8vLUpUsXdejQQePGjWtw08YjR44oJydHsbGxSkhI0IwZM3Tu3LmAMVu2bNEPfvADeTwe3XDDDVq+fHlzxwuwZMkS9evXz7kJlc/n0zvvvOOsD5Wc37VgwQK5XC5Nnz7dWRYqWQsKCuRyuQIevXr1ctaHSs56X331lX72s5+pS5cuiomJUd++fbVr1y5nfaj8XurevXuD19XlcikvL09S6Lyu58+f15NPPqkePXooJiZGPXv21NNPPx3wOUxt6jUN+ocpWGjlypUmOjra/PGPfzT79u0zkydPNnFxcaaioqK1p3ZR69atM7/61a/Mm2++aSSZVatWBaxfsGCB6dSpk1m9erX5+OOPzV133WV69Ohhvv76a2fMyJEjTf/+/c2OHTvMX/7yF3PDDTeYe++911lfVVVlEhMTTW5uriktLTWvvfaaiYmJMf/6r//aUjFNdna2WbZsmSktLTW7d+82o0ePNmlpaeb06dPOmIceesikpqaaTZs2mV27dpkhQ4aYH/7wh876c+fOmVtvvdVkZmaajz76yKxbt8507drVzJ492xnz2WefmdjYWJOfn2/2799vFi1aZCIjI8369etbLOuf//xn8/bbb5tPP/3UlJWVmccff9y43W5TWloaUjkv9MEHH5ju3bubfv36mYcffthZHipZ586da2655RZz7Ngx5/G3v/0t5HIaY8zJkydNt27dzAMPPGB27txpPvvsM7NhwwZz8OBBZ0yo/F46fvx4wGtaVFRkJJn33nvPGBM6r+szzzxjunTpYtauXWsOHz5s3njjDdOhQwfzhz/8wRnTll5TCo0xZvDgwSYvL895fv78eZOSkmLmz5/firO6fN8tNHV1dSYpKcn89re/dZZVVlYaj8djXnvtNWOMMfv37zeSzIcffuiMeeedd4zL5TJfffWVMcaYF1980XTu3NnU1tY6Y2bOnGluvvnmZk50ccePHzeSzNatW40x3+Zyu93mjTfecMZ88sknRpIpLi42xnxb/iIiIkx5ebkzZsmSJcbr9TrZHnvsMXPLLbcE7Ouee+4x2dnZzR3pe3Xu3Nn8+7//e0jmPHXqlLnxxhtNUVGR+Yd/+Aen0IRS1rlz55r+/fs3ui6Uchrz7e+GoUOHXnR9KP9eevjhh03Pnj1NXV1dSL2uOTk55sEHHwxYdvfdd5vc3FxjTNt7TcP+lNPZs2dVUlKizMxMZ1lERIQyMzNVXFzcijO7cocPH1Z5eXlApk6dOikjI8PJVFxcrLi4OA0aNMgZk5mZqYiICO3cudMZc/vttys6OtoZk52drbKyMv3P//xPC6UJVFVVJUmKj4+XJJWUlMjv9wdk7dWrl9LS0gKy9u3bN+CmjdnZ2aqurta+ffucMRduo35Ma/0MnD9/XitXrtSZM2fk8/lCMmdeXp5ycnIazCfUsh44cEApKSm6/vrrlZubqyNHjkgKvZx//vOfNWjQIP3jP/6jEhISlJ6ern/7t39z1ofq76WzZ8/q1Vdf1YMPPiiXyxVSr+sPf/hDbdq0SZ9++qkk6eOPP9b27ds1atQoSW3vNQ37QvP3v/9d58+fb3CH4sTERJWXl7fSrK5O/by/L1N5ebkSEhIC1kdFRSk+Pj5gTGPbuHAfLamurk7Tp0/Xj370I916663OPKKjoxt8WOl3s14qx8XGVFdX6+uvv26OOI3au3evOnToII/Ho4ceekirVq1Snz59Qi7nypUr9d///d+aP39+g3WhlDUjI0PLly/X+vXrtWTJEh0+fFg//vGPderUqZDKKUmfffaZlixZohtvvFEbNmzQ1KlT9S//8i965ZVXAuYbar+XVq9ercrKSj3wwAPOHELldZ01a5bGjx+vXr16ye12Kz09XdOnT1dubm7AXNvKa9pqH30ANFVeXp5KS0u1ffv21p5Ks7n55pu1e/duVVVV6T//8z81YcIEbd26tbWnFVRffvmlHn74YRUVFaldu3atPZ1mVf9/spLUr18/ZWRkqFu3bnr99dcVExPTijMLvrq6Og0aNEjPPvusJCk9PV2lpaVaunSpJkyY0Mqzaz4vv/yyRo0apZSUlNaeStC9/vrrWrFihQoLC3XLLbdo9+7dmj59ulJSUtrkaxr2R2i6du2qyMjIBu9Ar6ioUFJSUivN6urUz/v7MiUlJen48eMB68+dO6eTJ08GjGlsGxfuo6VMmzZNa9eu1XvvvafrrrvOWZ6UlKSzZ8+qsrIyYPx3s14qx8XGeL3eFv3DEx0drRtuuEEDBw7U/Pnz1b9/f/3hD38IqZwlJSU6fvy4fvCDHygqKkpRUVHaunWrXnjhBUVFRSkxMTFksn5XXFycbrrpJh08eDCkXlNJSk5OVp8+fQKW9e7d2znFFoq/l7744gu9++67+qd/+idnWSi9rjNmzHCO0vTt21f333+/HnnkEefIalt7TcO+0ERHR2vgwIHatGmTs6yurk6bNm2Sz+drxZlduR49eigpKSkgU3V1tXbu3Olk8vl8qqysVElJiTNm8+bNqqurU0ZGhjNm27Zt8vv9zpiioiLdfPPN6ty5c4tkMcZo2rRpWrVqlTZv3qwePXoErB84cKDcbndA1rKyMh05ciQg6969ewP+oyoqKpLX63V+Aft8voBt1I9p7Z+Buro61dbWhlTO4cOHa+/evdq9e7fzGDRokHJzc51/D5Ws33X69GkdOnRIycnJIfWaStKPfvSjBrdU+PTTT9WtWzdJofV7qd6yZcuUkJCgnJwcZ1kova41NTWKiAisCZGRkaqrq5PUBl/TJr2FOEStXLnSeDwes3z5crN//34zZcoUExcXF/AO9Lbm1KlT5qOPPjIfffSRkWSee+4589FHH5kvvvjCGPPtpXRxcXHmrbfeMnv27DE/+clPGr2ULj093ezcudNs377d3HjjjQGX0lVWVprExERz//33m9LSUrNy5UoTGxvbopdHTp061XTq1Mls2bIl4DLJmpoaZ8xDDz1k0tLSzObNm82uXbuMz+czPp/PWV9/iWRWVpbZvXu3Wb9+vbnmmmsavURyxowZ5pNPPjGLFy9u8UskZ82aZbZu3WoOHz5s9uzZY2bNmmVcLpfZuHFjSOVszIVXORkTOlkfffRRs2XLFnP48GHzX//1XyYzM9N07drVHD9+PKRyGvPtJfhRUVHmmWeeMQcOHDArVqwwsbGx5tVXX3XGhMrvJWO+vRo2LS3NzJw5s8G6UHldJ0yYYK699lrnsu0333zTdO3a1Tz22GPOmLb0mlJo/s+iRYtMWlqaiY6ONoMHDzY7duxo7Sl9r/fee89IavCYMGGCMebby+mefPJJk5iYaDwejxk+fLgpKysL2MaJEyfMvffeazp06GC8Xq+ZOHGiOXXqVMCYjz/+2AwdOtR4PB5z7bXXmgULFrRURGOMaTSjJLNs2TJnzNdff23++Z//2XTu3NnExsaan/70p+bYsWMB2/n888/NqFGjTExMjOnatat59NFHjd/vDxjz3nvvmQEDBpjo6Ghz/fXXB+yjJTz44IOmW7duJjo62lxzzTVm+PDhTpkxJnRyNua7hSZUst5zzz0mOTnZREdHm2uvvdbcc889AfdlCZWc9dasWWNuvfVW4/F4TK9evcxLL70UsD5Ufi8ZY8yGDRuMpAbzNyZ0Xtfq6mrz8MMPm7S0NNOuXTtz/fXXm1/96lcBl1e3pdfUZcwFt/wDAACwUNi/hwYAANiPQgMAAKxHoQEAANaj0AAAAOtRaAAAgPUoNAAAwHoUGgAAYD0KDQAAsB6FBgAAWI9CAwAArEehAQAA1qPQAAAA6/0v+o9FesFV+9wAAAAASUVORK5CYII=",
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "df['tokens_in_ext_desc_readme_trim'].hist(bins = 50)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {
        "id": "U-1htAXoj62n"
      },
      "outputs": [],
      "source": [
        "assert df['tokens_in_ext_desc'].sum() < 100000\n",
        "# as if Nov 7 2023 allshort  descriptions of tools is jsut 68K tokens, so\n",
        "# no need for embed db truly speaking, can also just make claude calls every time need to pick a tool"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {
        "id": "DC1s8Np6mGwo"
      },
      "outputs": [],
      "source": [
        "assert df.Name.nunique() == df.shape[0]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 33,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "0"
            ]
          },
          "execution_count": 33,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "df['extented_desc'].isna().sum()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 34,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "13"
            ]
          },
          "execution_count": 34,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "df['Code'].isna().sum()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 45,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "0"
            ]
          },
          "execution_count": 45,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "df['Readme'].isna().sum()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 47,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "0     # acorde: isoform co-usage networks from singl...\n",
              "1     # Automated Cell Type Identification using Neu...\n",
              "2                                                      \n",
              "3     ## ACTIONet R Package\\nSee [ACTIONet-R](https:...\n",
              "4     # ACTIVA: Realistic scRNAseq Generation with A...\n",
              "5     # activeSVC\\nActiveSVC selects features for la...\n",
              "6     # **ADC**\\r\\n\\r\\nADC is an efficient method to...\n",
              "7     \\n\\n A Parameter-free Deep Embedded Clustering...\n",
              "8     \\n<!-- badges: start -->\\n[![Build Status](htt...\n",
              "9     # Description\\nadobo is a Python-framework for...\n",
              "10                                                     \n",
              "11    # AE-TPGG\\n\\n\\nThis is a autoencoder based on ...\n",
              "12    \\n<!-- README.md is generated from README.Rmd....\n",
              "13    # airpart\\nDifferential allelic imbalance part...\n",
              "14    # aLDG\\n\\nWelcome to ``aLDG``!  ``aLDG`` is an...\n",
              "15    <img alt=\"logo\" src=\"https://github.com/COMBIN...\n",
              "16    ## alevinQC\\n\\n[![Codecov.io coverage status](...\n",
              "17    # Adaptively-thresholded Low Rank Approximatio...\n",
              "18    # AltAnalyze # \\n\\nAn automated cross-platform...\n",
              "19    \\n# AlzScPred :- A Tool for Identification of ...\n",
              "Name: Readme, dtype: object"
            ]
          },
          "execution_count": 47,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "df['Readme'].head(20)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 60,
      "metadata": {},
      "outputs": [],
      "source": [
        "df['Readme'] = df['Readme'].apply(lambda x: x.replace('\\n', ' ').replace('\\r', ' ').replace(\"'\", \"\\\\'\") if pd.notna(x) else x)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Embedd Vectors "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 57,
      "metadata": {
        "id": "vo7Qyj60e1Ao"
      },
      "outputs": [],
      "source": [
        "# Add the content vectors\n",
        "scrnatools_description_collection.add(\n",
        "    documents = list(df['extented_desc']),\n",
        "    metadatas = df.drop(['extented_desc'], axis = 1).to_dict(orient='records'),\n",
        "    ids = list(df.Name)\n",
        ")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 88,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Add of existing embedding ID: acorde\n",
            "Add of existing embedding ID: ACTINN\n",
            "Add of existing embedding ID: ACTION\n",
            "Add of existing embedding ID: ACTIONet\n",
            "Add of existing embedding ID: ACTIVA\n",
            "Add of existing embedding ID: activeSVC\n",
            "Add of existing embedding ID: ADC\n",
            "Add of existing embedding ID: ADClust\n",
            "Add of existing embedding ID: ADImpute\n",
            "Add of existing embedding ID: adobo\n",
            "Add of existing embedding ID: adverSCarial\n",
            "Add of existing embedding ID: AE-TPGG\n",
            "Add of existing embedding ID: aggregateBioVar\n",
            "Add of existing embedding ID: airpart\n",
            "Add of existing embedding ID: aLDG\n",
            "Add of existing embedding ID: alevin-fry\n",
            "Add of existing embedding ID: alevinQC\n",
            "Add of existing embedding ID: ALRA\n",
            "Add of existing embedding ID: AltAnalyze\n",
            "Add of existing embedding ID: AlzScPred\n",
            "Add of existing embedding ID: AmbientContaminationMetrics\n",
            "Add of existing embedding ID: anchor\n",
            "Add of existing embedding ID: ANS\n",
            "Add of existing embedding ID: anticor_features\n",
            "Add of existing embedding ID: APL\n",
            "Add of existing embedding ID: ASAP\n",
            "Add of existing embedding ID: Asc-Seurat\n",
            "Add of existing embedding ID: ascend\n",
            "Add of existing embedding ID: ASTK\n",
            "Add of existing embedding ID: ASURAT\n",
            "Add of existing embedding ID: ATAXIC\n",
            "Add of existing embedding ID: AUCell\n",
            "Add of existing embedding ID: Augur\n",
            "Add of existing embedding ID: AutoClass\n",
            "Add of existing embedding ID: AutoImpute\n",
            "Add of existing embedding ID: AutomaticCellTypeIdentification\n",
            "Add of existing embedding ID: AWGAN\n",
            "Add of existing embedding ID: BABEL\n",
            "Add of existing embedding ID: BackSPIN\n",
            "Add of existing embedding ID: badger\n",
            "Add of existing embedding ID: BALDR\n",
            "Add of existing embedding ID: BAMMSC\n",
            "Add of existing embedding ID: BarCounter\n",
            "Add of existing embedding ID: baredSC\n",
            "Add of existing embedding ID: BarMixer\n",
            "Add of existing embedding ID: BASIC\n",
            "Add of existing embedding ID: BASiCS\n",
            "Add of existing embedding ID: BASiCStan\n",
            "Add of existing embedding ID: BasisVAE\n",
            "Add of existing embedding ID: BatchBench\n",
            "Add of existing embedding ID: batchelor\n",
            "Add of existing embedding ID: BATMAN\n",
            "Add of existing embedding ID: BayesEATS\n",
            "Add of existing embedding ID: BayesPrism\n",
            "Add of existing embedding ID: bayNorm\n",
            "Add of existing embedding ID: bbknn\n",
            "Add of existing embedding ID: BC_tSNE\n",
            "Add of existing embedding ID: bcbio\n",
            "Add of existing embedding ID: bcbioSingleCell\n",
            "Add of existing embedding ID: BCseq\n",
            "Add of existing embedding ID: BDA\n",
            "Add of existing embedding ID: BEANIE\n",
            "Add of existing embedding ID: BEARscc\n",
            "Add of existing embedding ID: BEELINE\n",
            "Add of existing embedding ID: BEENE\n",
            "Add of existing embedding ID: BEER\n",
            "Add of existing embedding ID: BENISSE\n",
            "Add of existing embedding ID: BERMUDA\n",
            "Add of existing embedding ID: BESCA\n",
            "Add of existing embedding ID: Beyondcell\n",
            "Add of existing embedding ID: Bfimpute\n",
            "Add of existing embedding ID: bigSCale\n",
            "Add of existing embedding ID: bigSCale2\n",
            "Add of existing embedding ID: bindSC\n",
            "Add of existing embedding ID: BingleSeq\n",
            "Add of existing embedding ID: biolord\n",
            "Add of existing embedding ID: Biomodelling\n",
            "Add of existing embedding ID: BioTIP\n",
            "Add of existing embedding ID: BIRD\n",
            "Add of existing embedding ID: BISCUIT\n",
            "Add of existing embedding ID: Bisque\n",
            "Add of existing embedding ID: BITFAM\n",
            "Add of existing embedding ID: BLGGM\n",
            "Add of existing embedding ID: bollito\n",
            "Add of existing embedding ID: bonvoyage\n",
            "Add of existing embedding ID: BPSC\n",
            "Add of existing embedding ID: BraCeR\n",
            "Add of existing embedding ID: BranchedGP\n",
            "Add of existing embedding ID: BRAPeS\n",
            "Add of existing embedding ID: BREM-SC\n",
            "Add of existing embedding ID: BRIE\n",
            "Add of existing embedding ID: BSDE\n",
            "Add of existing embedding ID: bseqsc\n",
            "Add of existing embedding ID: BTR\n",
            "Add of existing embedding ID: BUSseq\n",
            "Add of existing embedding ID: bustools\n",
            "Add of existing embedding ID: BZINB\n",
            "Add of existing embedding ID: CAbiNet\n",
            "Add of existing embedding ID: CACTUS\n",
            "Add of existing embedding ID: CAISC\n",
            "Add of existing embedding ID: CALISTA\n",
            "Add of existing embedding ID: CAMLU\n",
            "Add of existing embedding ID: CancerInSilico\n",
            "Add of existing embedding ID: Canek\n",
            "Add of existing embedding ID: CanSig\n",
            "Add of existing embedding ID: CAPITAL\n",
            "Add of existing embedding ID: Capybara\n",
            "Add of existing embedding ID: CARDAMOM\n",
            "Add of existing embedding ID: CarDEC\n",
            "Add of existing embedding ID: cardelino\n",
            "Add of existing embedding ID: CaSee\n",
            "Add of existing embedding ID: CASi\n",
            "Add of existing embedding ID: CaSpER\n",
            "Add of existing embedding ID: CaSTLe\n",
            "Add of existing embedding ID: CausalDAG\n",
            "Add of existing embedding ID: cb_sniffer\n",
            "Add of existing embedding ID: ccdf\n",
            "Add of existing embedding ID: ccfindR\n",
            "Add of existing embedding ID: ccImpute\n",
            "Add of existing embedding ID: ccNet\n",
            "Add of existing embedding ID: CCNMF\n",
            "Add of existing embedding ID: CCPE\n",
            "Add of existing embedding ID: CCPlotR\n",
            "Add of existing embedding ID: ccRemover\n",
            "Add of existing embedding ID: CCSF\n",
            "Add of existing embedding ID: CCSN\n",
            "Add of existing embedding ID: CDI\n",
            "Add of existing embedding ID: CDR-g\n",
            "Add of existing embedding ID: celaref\n",
            "Add of existing embedding ID: celda\n",
            "Add of existing embedding ID: CeLEry\n",
            "Add of existing embedding ID: Cell-BLAST\n",
            "Add of existing embedding ID: cell2cell\n",
            "Add of existing embedding ID: cell2fate\n",
            "Add of existing embedding ID: cell2location\n",
            "Add of existing embedding ID: cell2sentence\n",
            "Add of existing embedding ID: cell2sentence-ft\n",
            "Add of existing embedding ID: cellAlign\n",
            "Add of existing embedding ID: CellANOVA\n",
            "Add of existing embedding ID: Cellar\n",
            "Add of existing embedding ID: CellaRepertorium\n",
            "Add of existing embedding ID: cellassign\n",
            "Add of existing embedding ID: cellatlas\n",
            "Add of existing embedding ID: CellBench\n",
            "Add of existing embedding ID: CellBender\n",
            "Add of existing embedding ID: CellBIC\n",
            "Add of existing embedding ID: cellBrowser\n",
            "Add of existing embedding ID: CellCall\n",
            "Add of existing embedding ID: CellChat\n",
            "Add of existing embedding ID: CellContrast\n",
            "Add of existing embedding ID: cellDancer\n",
            "Add of existing embedding ID: CellDrift\n",
            "Add of existing embedding ID: cellenium\n",
            "Add of existing embedding ID: CellexalVR\n",
            "Add of existing embedding ID: CellFishing\n",
            "Add of existing embedding ID: CellGEP\n",
            "Add of existing embedding ID: cellgrid\n",
            "Add of existing embedding ID: CellHint\n",
            "Add of existing embedding ID: CellID\n",
            "Add of existing embedding ID: Cellity\n",
            "Add of existing embedding ID: CellLayers\n",
            "Add of existing embedding ID: CellMap\n",
            "Add of existing embedding ID: CellMeSH\n",
            "Add of existing embedding ID: CellMixS\n",
            "Add of existing embedding ID: CellO\n",
            "Add of existing embedding ID: cellograph\n",
            "Add of existing embedding ID: CellOracle\n",
            "Add of existing embedding ID: CellOT\n",
            "Add of existing embedding ID: CellPath\n",
            "Add of existing embedding ID: CellPhoneDB\n",
            "Add of existing embedding ID: CellPLM\n",
            "Add of existing embedding ID: cellpypes\n",
            "Add of existing embedding ID: CellRanger\n",
            "Add of existing embedding ID: CellRank\n",
            "Add of existing embedding ID: CellRegMap\n",
            "Add of existing embedding ID: CellRouter\n",
            "Add of existing embedding ID: CellSIUS\n",
            "Add of existing embedding ID: cellsnake\n",
            "Add of existing embedding ID: cellSNP\n",
            "Add of existing embedding ID: cellsnp-lite\n",
            "Add of existing embedding ID: CellST\n",
            "Add of existing embedding ID: CellTagR\n",
            "Add of existing embedding ID: celltalker\n",
            "Add of existing embedding ID: CellTools\n",
            "Add of existing embedding ID: CellTrails\n",
            "Add of existing embedding ID: CellTree\n",
            "Add of existing embedding ID: CellTrek\n",
            "Add of existing embedding ID: CellTypist\n",
            "Add of existing embedding ID: CellView\n",
            "Add of existing embedding ID: cellxgene\n",
            "Add of existing embedding ID: cellxgene_VIP\n",
            "Add of existing embedding ID: cellXY\n",
            "Add of existing embedding ID: Census\n",
            "Add of existing embedding ID: Cepo\n",
            "Add of existing embedding ID: Cerebro\n",
            "Add of existing embedding ID: CeSpGRN\n",
            "Add of existing embedding ID: CeTF\n",
            "Add of existing embedding ID: cFIT\n",
            "Add of existing embedding ID: CHETAH\n",
            "Add of existing embedding ID: Chloris\n",
            "Add of existing embedding ID: chooseR\n",
            "Add of existing embedding ID: Chord\n",
            "Add of existing embedding ID: CIARA\n",
            "Add of existing embedding ID: CIDER\n",
            "Add of existing embedding ID: CIDR\n",
            "Add of existing embedding ID: CIForm\n",
            "Add of existing embedding ID: CINEMA-OT\n",
            "Add of existing embedding ID: CIPR\n",
            "Add of existing embedding ID: Cirrocumulus\n",
            "Add of existing embedding ID: CiteFuse\n",
            "Add of existing embedding ID: CITEseqCount\n",
            "Add of existing embedding ID: Citrus\n",
            "Add of existing embedding ID: CL-Impute\n",
            "Add of existing embedding ID: CLEAR\n",
            "Add of existing embedding ID: clonealign\n",
            "Add of existing embedding ID: ClonoCluster\n",
            "Add of existing embedding ID: ClustAssess\n",
            "Add of existing embedding ID: cluster_diffex\n",
            "Add of existing embedding ID: ClusterDE\n",
            "Add of existing embedding ID: clusterExperiment\n",
            "Add of existing embedding ID: ClusterFoldSimilarity\n",
            "Add of existing embedding ID: Clustergrammer2\n",
            "Add of existing embedding ID: ClusterMap\n",
            "Add of existing embedding ID: ClusterMine\n",
            "Add of existing embedding ID: clustermole\n",
            "Add of existing embedding ID: clusterpval\n",
            "Add of existing embedding ID: clustifyr\n",
            "Add of existing embedding ID: CMFImpute\n",
            "Add of existing embedding ID: cna\n",
            "Add of existing embedding ID: cNMF\n",
            "Add of existing embedding ID: CNNC\n",
            "Add of existing embedding ID: COAC\n",
            "Add of existing embedding ID: Cobolt\n",
            "Add of existing embedding ID: cocoa\n",
            "Add of existing embedding ID: CoGAPs\n",
            "Add of existing embedding ID: cola\n",
            "Add of existing embedding ID: CombiROC\n",
            "Add of existing embedding ID: COMET\n",
            "Add of existing embedding ID: CommPath\n",
            "Add of existing embedding ID: Compartmap\n",
            "Add of existing embedding ID: compass\n",
            "Add of existing embedding ID: CONCLUS\n",
            "Add of existing embedding ID: condiments\n",
            "Add of existing embedding ID: CoNGA\n",
            "Add of existing embedding ID: CONGAS\n",
            "Add of existing embedding ID: CONGASp\n",
            "Add of existing embedding ID: CONICS\n",
            "Add of existing embedding ID: Connectome\n",
            "Add of existing embedding ID: Conos\n",
            "Add of existing embedding ID: constclust\n",
            "Add of existing embedding ID: contrastiveVI\n",
            "Add of existing embedding ID: COPILOT\n",
            "Add of existing embedding ID: CopulaGCN\n",
            "Add of existing embedding ID: CopyKAT\n",
            "Add of existing embedding ID: corral\n",
            "Add of existing embedding ID: Corvo\n",
            "Add of existing embedding ID: COSG\n",
            "Add of existing embedding ID: COSGR\n",
            "Add of existing embedding ID: cosin\n",
            "Add of existing embedding ID: COSLIR\n",
            "Add of existing embedding ID: CoSpar\n",
            "Add of existing embedding ID: CosTaL\n",
            "Add of existing embedding ID: COT\n",
            "Add of existing embedding ID: COTAN\n",
            "Add of existing embedding ID: countClust\n",
            "Add of existing embedding ID: countland\n",
            "Add of existing embedding ID: countsplit\n",
            "Add of existing embedding ID: coupleCoC\n",
            "Add of existing embedding ID: coupleCoC_plus\n",
            "Add of existing embedding ID: coupledAE-patchseq\n",
            "Add of existing embedding ID: CPA\n",
            "Add of existing embedding ID: CPLVM\n",
            "Add of existing embedding ID: CRMetrics\n",
            "Add of existing embedding ID: CrossTalkeR\n",
            "Add of existing embedding ID: CS-CORE\n",
            "Add of existing embedding ID: CSHMM\n",
            "Add of existing embedding ID: CSHMM-TF\n",
            "Add of existing embedding ID: CSMF\n",
            "Add of existing embedding ID: CSN\n",
            "Add of existing embedding ID: CSOmap\n",
            "Add of existing embedding ID: CStreet\n",
            "Add of existing embedding ID: ctgGEM\n",
            "Add of existing embedding ID: CTMM\n",
            "Add of existing embedding ID: CTpathway\n",
            "Add of existing embedding ID: cTPnet\n",
            "Add of existing embedding ID: cvCovEst\n",
            "Add of existing embedding ID: Cyclum\n",
            "Add of existing embedding ID: cytocipher\n",
            "Add of existing embedding ID: Cytopath\n",
            "Add of existing embedding ID: CytoSPACE\n",
            "Add of existing embedding ID: CytoTalk\n",
            "Add of existing embedding ID: D-EE\n",
            "Add of existing embedding ID: D-SPIN\n",
            "Add of existing embedding ID: D3E\n",
            "Add of existing embedding ID: DALI\n",
            "Add of existing embedding ID: DANCE\n",
            "Add of existing embedding ID: dandelion\n",
            "Add of existing embedding ID: DAPCA\n",
            "Add of existing embedding ID: DAZZLE\n",
            "Add of existing embedding ID: dbcti\n",
            "Add of existing embedding ID: DC3\n",
            "Add of existing embedding ID: DCA\n",
            "Add of existing embedding ID: DCA_R\n",
            "Add of existing embedding ID: DCATS\n",
            "Add of existing embedding ID: DCLEAR\n",
            "Add of existing embedding ID: ddqc\n",
            "Add of existing embedding ID: ddSeeker\n",
            "Add of existing embedding ID: de-goulash\n",
            "Add of existing embedding ID: DECENT\n",
            "Add of existing embedding ID: DECODE\n",
            "Add of existing embedding ID: decontX\n",
            "Add of existing embedding ID: decoupler\n",
            "Add of existing embedding ID: deCS\n",
            "Add of existing embedding ID: DEEF\n",
            "Add of existing embedding ID: DeepCCI\n",
            "Add of existing embedding ID: DeepCycle\n",
            "Add of existing embedding ID: DeepDRIM\n",
            "Add of existing embedding ID: DeepImpute\n",
            "Add of existing embedding ID: DeepRIG\n",
            "Add of existing embedding ID: DeepSEM\n",
            "Add of existing embedding ID: DeepSort\n",
            "Add of existing embedding ID: DeepVelo\n",
            "Add of existing embedding ID: DEGAS\n",
            "Add of existing embedding ID: DElegate\n",
            "Add of existing embedding ID: DeLorean\n",
            "Add of existing embedding ID: DELVE\n",
            "Add of existing embedding ID: deMULTIplex2\n",
            "Add of existing embedding ID: Demuxafy\n",
            "Add of existing embedding ID: Demuxalot\n",
            "Add of existing embedding ID: demuxlet\n",
            "Add of existing embedding ID: demuxmix\n",
            "Add of existing embedding ID: DEN\n",
            "Add of existing embedding ID: DENDRO\n",
            "Add of existing embedding ID: DendroSplit\n",
            "Add of existing embedding ID: densityCut\n",
            "Add of existing embedding ID: DensityPath\n",
            "Add of existing embedding ID: densvis\n",
            "Add of existing embedding ID: densVis\n",
            "Add of existing embedding ID: DepecheR\n",
            "Add of existing embedding ID: DESC\n",
            "Add of existing embedding ID: DESCEND\n",
            "Add of existing embedding ID: DEsingle\n",
            "Add of existing embedding ID: destiny\n",
            "Add of existing embedding ID: DevKidCC\n",
            "Add of existing embedding ID: DEWAKSS\n",
            "Add of existing embedding ID: Dhaka\n",
            "Add of existing embedding ID: DIALOGUE\n",
            "Add of existing embedding ID: DICE\n",
            "Add of existing embedding ID: Dictys\n",
            "Add of existing embedding ID: DIEM\n",
            "Add of existing embedding ID: DifferentialRegulation\n",
            "Add of existing embedding ID: DigitalCellSorter\n",
            "Add of existing embedding ID: DIMMSC\n",
            "Add of existing embedding ID: dincta\n",
            "Add of existing embedding ID: Dino\n",
            "Add of existing embedding ID: DIRECT-NET\n",
            "Add of existing embedding ID: DISC\n",
            "Add of existing embedding ID: DIscBIO\n",
            "Add of existing embedding ID: DISCERN\n",
            "Add of existing embedding ID: DISCO\n",
            "Add of existing embedding ID: DiSiR\n",
            "Add of existing embedding ID: distinct\n",
            "Add of existing embedding ID: DistMap\n",
            "Add of existing embedding ID: dittoSeq\n",
            "Add of existing embedding ID: dmatch\n",
            "Add of existing embedding ID: Domino\n",
            "Add of existing embedding ID: DoubletCollection\n",
            "Add of existing embedding ID: DoubletDecon\n",
            "Add of existing embedding ID: DoubletDetection\n",
            "Add of existing embedding ID: DoubletFinder\n",
            "Add of existing embedding ID: Dozer\n",
            "Add of existing embedding ID: dpi\n",
            "Add of existing embedding ID: DPre\n",
            "Add of existing embedding ID: DPT\n",
            "Add of existing embedding ID: DRA\n",
            "Add of existing embedding ID: dragonsc\n",
            "Add of existing embedding ID: DREAM\n",
            "Add of existing embedding ID: dreamlet\n",
            "Add of existing embedding ID: DREEP\n",
            "Add of existing embedding ID: DrImpute\n",
            "Add of existing embedding ID: DrivAER\n",
            "Add of existing embedding ID: dropbead\n",
            "Add of existing embedding ID: dropClust\n",
            "Add of existing embedding ID: dropEst\n",
            "Add of existing embedding ID: dropkick\n",
            "Add of existing embedding ID: DropLasso\n",
            "Add of existing embedding ID: DropletQC\n",
            "Add of existing embedding ID: DropletUtils\n",
            "Add of existing embedding ID: dropSeqPipe\n",
            "Add of existing embedding ID: dropsim\n",
            "Add of existing embedding ID: DrSeq2\n",
            "Add of existing embedding ID: DSAVE-M\n",
            "Add of existing embedding ID: DSAVE-R\n",
            "Add of existing embedding ID: dsb\n",
            "Add of existing embedding ID: dsCellNet\n",
            "Add of existing embedding ID: DSINMF\n",
            "Add of existing embedding ID: DTAE\n",
            "Add of existing embedding ID: DTFLOW\n",
            "Add of existing embedding ID: DTUrtle\n",
            "Add of existing embedding ID: DTWScore\n",
            "Add of existing embedding ID: DUBStepR\n",
            "Add of existing embedding ID: Dune\n",
            "Add of existing embedding ID: DURIAN\n",
            "Add of existing embedding ID: DUSC\n",
            "Add of existing embedding ID: Dynamo\n",
            "Add of existing embedding ID: dynDeepDRIM\n",
            "Add of existing embedding ID: dyngen\n",
            "Add of existing embedding ID: DynGFN\n",
            "Add of existing embedding ID: dynverse\n",
            "Add of existing embedding ID: EasyCellType\n",
            "Add of existing embedding ID: ECLAIR\n",
            "Add of existing embedding ID: EDClust\n",
            "Add of existing embedding ID: EDGE\n",
            "Add of existing embedding ID: EEISP\n",
            "Add of existing embedding ID: elasticC3\n",
            "Add of existing embedding ID: ELeFHAnt\n",
            "Add of existing embedding ID: ELSA\n",
            "Add of existing embedding ID: ELVAR\n",
            "Add of existing embedding ID: embeddr\n",
            "Add of existing embedding ID: EMBEDR\n",
            "Add of existing embedding ID: EMEP\n",
            "Add of existing embedding ID: EmptyNN\n",
            "Add of existing embedding ID: enclone\n",
            "Add of existing embedding ID: ENCORE\n",
            "Add of existing embedding ID: ENHANCE\n",
            "Add of existing embedding ID: EnImpute\n",
            "Add of existing embedding ID: EnsembleKQC\n",
            "Add of existing embedding ID: EnTSSR\n",
            "Add of existing embedding ID: Eoulsan\n",
            "Add of existing embedding ID: Epi-Impute\n",
            "Add of existing embedding ID: EPIC\n",
            "Add of existing embedding ID: eQTLsingle\n",
            "Add of existing embedding ID: ESAT\n",
            "Add of existing embedding ID: escape\n",
            "Add of existing embedding ID: ESCO\n",
            "Add of existing embedding ID: eSPRESSO\n",
            "Add of existing embedding ID: eSVD\n",
            "Add of existing embedding ID: EVI\n",
            "Add of existing embedding ID: exceiver\n",
            "Add of existing embedding ID: exFINDER\n",
            "Add of existing embedding ID: expressyouRcell\n",
            "Add of existing embedding ID: Falco\n",
            "Add of existing embedding ID: FastCount\n",
            "Add of existing embedding ID: fasthplus\n",
            "Add of existing embedding ID: FastIntegration\n",
            "Add of existing embedding ID: FastProject\n",
            "Add of existing embedding ID: FastRNA\n",
            "Add of existing embedding ID: fastTopics\n",
            "Add of existing embedding ID: FateCompass\n",
            "Add of existing embedding ID: FateID\n",
            "Add of existing embedding ID: FAVA\n",
            "Add of existing embedding ID: fba\n",
            "Add of existing embedding ID: fcoex\n",
            "Add of existing embedding ID: FEAST\n",
            "Add of existing embedding ID: FEATS\n",
            "Add of existing embedding ID: FeatureCorr\n",
            "Add of existing embedding ID: FEM\n",
            "Add of existing embedding ID: FIESTA\n",
            "Add of existing embedding ID: FiRE\n",
            "Add of existing embedding ID: FIRM\n",
            "Add of existing embedding ID: fishpond\n",
            "Add of existing embedding ID: FitDevo\n",
            "Add of existing embedding ID: FLAMES\n",
            "Add of existing embedding ID: Flexiplex\n",
            "Add of existing embedding ID: flotilla\n",
            "Add of existing embedding ID: FlowGrid\n",
            "Add of existing embedding ID: FLOWMAPR\n",
            "Add of existing embedding ID: forest-fire-clustering\n",
            "Add of existing embedding ID: FORKS\n",
            "Add of existing embedding ID: Fotomics\n",
            "Add of existing embedding ID: FRmatch\n",
            "Add of existing embedding ID: FSCAM\n",
            "Add of existing embedding ID: Fugue\n",
            "Add of existing embedding ID: FUNKI\n",
            "Add of existing embedding ID: fusca\n",
            "Add of existing embedding ID: G2S3\n",
            "Add of existing embedding ID: GapClust\n",
            "Add of existing embedding ID: Garnett\n",
            "Add of existing embedding ID: gCAnno\n",
            "Add of existing embedding ID: GEARS\n",
            "Add of existing embedding ID: GEDI\n",
            "Add of existing embedding ID: geneBasisR\n",
            "Add of existing embedding ID: GeneClust\n",
            "Add of existing embedding ID: GenePattern\n",
            "Add of existing embedding ID: Genes2Genes\n",
            "Add of existing embedding ID: genesorteR\n",
            "Add of existing embedding ID: GeneSwitches\n",
            "Add of existing embedding ID: GeneVector\n",
            "Add of existing embedding ID: GenKI\n",
            "Add of existing embedding ID: genomicInstability\n",
            "Add of existing embedding ID: geosketch\n",
            "Add of existing embedding ID: gExcite\n",
            "Add of existing embedding ID: gficf\n",
            "Add of existing embedding ID: GFPA\n",
            "Add of existing embedding ID: ggsc\n",
            "Add of existing embedding ID: GIANT\n",
            "Add of existing embedding ID: GiniClust\n",
            "Add of existing embedding ID: GiniClust2\n",
            "Add of existing embedding ID: GiniClust3\n",
            "Add of existing embedding ID: GiRAFR\n",
            "Add of existing embedding ID: GLM-PCA\n",
            "Add of existing embedding ID: glmGamPoi\n",
            "Add of existing embedding ID: GloScope\n",
            "Add of existing embedding ID: GLUE\n",
            "Add of existing embedding ID: GLUER\n",
            "Add of existing embedding ID: GMM-Demux\n",
            "Add of existing embedding ID: GNNImpute\n",
            "Add of existing embedding ID: GPcounts\n",
            "Add of existing embedding ID: GPfates\n",
            "Add of existing embedding ID: GPseudoClust\n",
            "Add of existing embedding ID: GPseudoRank\n",
            "Add of existing embedding ID: Granatum\n",
            "Add of existing embedding ID: GrandPrix\n",
            "Add of existing embedding ID: Graph-Dynamo\n",
            "Add of existing embedding ID: graph-sc\n",
            "Add of existing embedding ID: GraphCS\n",
            "Add of existing embedding ID: GraphEmbed\n",
            "Add of existing embedding ID: GraphSCC\n",
            "Add of existing embedding ID: GraphSCI\n",
            "Add of existing embedding ID: GrID-Net\n",
            "Add of existing embedding ID: GRISLI\n",
            "Add of existing embedding ID: GRM\n",
            "Add of existing embedding ID: GRN-VAE\n",
            "Add of existing embedding ID: GRNUlar\n",
            "Add of existing embedding ID: GRNVBEM\n",
            "Add of existing embedding ID: GRouNdGAN\n",
            "Add of existing embedding ID: GRUFFI\n",
            "Add of existing embedding ID: gsdensity\n",
            "Add of existing embedding ID: GSEApy\n",
            "Add of existing embedding ID: GSFA\n",
            "Add of existing embedding ID: gssnng\n",
            "Add of existing embedding ID: h-tSNE\n",
            "Add of existing embedding ID: hadge\n",
            "Add of existing embedding ID: Haisu\n",
            "Add of existing embedding ID: HAL\n",
            "Add of existing embedding ID: HARISSA\n",
            "Add of existing embedding ID: Harmony\n",
            "Add of existing embedding ID: hbfm\n",
            "Add of existing embedding ID: HD-AE\n",
            "Add of existing embedding ID: HDMC\n",
            "Add of existing embedding ID: HGC\n",
            "Add of existing embedding ID: HiDeF\n",
            "Add of existing embedding ID: hierarchicell\n",
            "Add of existing embedding ID: HieRFIT\n",
            "Add of existing embedding ID: HIPPO\n",
            "Add of existing embedding ID: HocusPocus\n",
            "Add of existing embedding ID: HoneyBADGER\n",
            "Add of existing embedding ID: HopLand\n",
            "Add of existing embedding ID: Hopper\n",
            "Add of existing embedding ID: Hotspot\n",
            "Add of existing embedding ID: HPL\n",
            "Add of existing embedding ID: hscScore\n",
            "Add of existing embedding ID: hsslda\n",
            "Add of existing embedding ID: HTOreader\n",
            "Add of existing embedding ID: HurdleNormal\n",
            "Add of existing embedding ID: hypercluster\n",
            "Add of existing embedding ID: I-Impute\n",
            "Add of existing embedding ID: i2dash\n",
            "Add of existing embedding ID: iasva\n",
            "Add of existing embedding ID: Ibex\n",
            "Add of existing embedding ID: IBMR\n",
            "Add of existing embedding ID: IBRAP\n",
            "Add of existing embedding ID: iBRIDGE\n",
            "Add of existing embedding ID: ICAnet\n",
            "Add of existing embedding ID: ICARUS\n",
            "Add of existing embedding ID: ICAT\n",
            "Add of existing embedding ID: ICELLNET\n",
            "Add of existing embedding ID: iCellR\n",
            "Add of existing embedding ID: iDEA\n",
            "Add of existing embedding ID: IDEAS\n",
            "Add of existing embedding ID: IEntropy\n",
            "Add of existing embedding ID: IKAP\n",
            "Add of existing embedding ID: ikarus\n",
            "Add of existing embedding ID: Ikura\n",
            "Add of existing embedding ID: ILoReg\n",
            "Add of existing embedding ID: iMAP\n",
            "Add of existing embedding ID: ImmClassifier\n",
            "Add of existing embedding ID: immunarch\n",
            "Add of existing embedding ID: ImSpectR\n",
            "Add of existing embedding ID: IndepthPathway\n",
            "Add of existing embedding ID: indeXplorer\n",
            "Add of existing embedding ID: inferCNV\n",
            "Add of existing embedding ID: inferCNVpy\n",
            "Add of existing embedding ID: inferCSN\n",
            "Add of existing embedding ID: Inferelator\n",
            "Add of existing embedding ID: infohet\n",
            "Add of existing embedding ID: InGene\n",
            "Add of existing embedding ID: INSCT\n",
            "Add of existing embedding ID: INSISTC\n",
            "Add of existing embedding ID: IntegratedLearner\n",
            "Add of existing embedding ID: InterCellar\n",
            "Add of existing embedding ID: InterCellDB\n",
            "Add of existing embedding ID: iQcell\n",
            "Add of existing embedding ID: IReNA\n",
            "Add of existing embedding ID: IRIS-FGM\n",
            "Add of existing embedding ID: IronThrone-GoT\n",
            "Add of existing embedding ID: iS-CellR\n",
            "Add of existing embedding ID: iscandar\n",
            "Add of existing embedding ID: iSEE\n",
            "Add of existing embedding ID: iSMNN\n",
            "Add of existing embedding ID: ISnorm\n",
            "Add of existing embedding ID: IsoCell\n",
            "Add of existing embedding ID: ISOP\n",
            "Add of existing embedding ID: iTALK\n",
            "Add of existing embedding ID: ItClust\n",
            "Add of existing embedding ID: iterClust\n",
            "Add of existing embedding ID: ivis\n",
            "Add of existing embedding ID: jackstraw\n",
            "Add of existing embedding ID: JAMIE\n",
            "Add of existing embedding ID: JGNsc\n",
            "Add of existing embedding ID: JIND\n",
            "Add of existing embedding ID: JOINT\n",
            "Add of existing embedding ID: JSNMF_py\n",
            "Add of existing embedding ID: JSOM\n",
            "Add of existing embedding ID: jSRC\n",
            "Add of existing embedding ID: Jvis\n",
            "Add of existing embedding ID: k-branches\n",
            "Add of existing embedding ID: K2Taxonomer\n",
            "Add of existing embedding ID: kallisto\n",
            "Add of existing embedding ID: kana\n",
            "Add of existing embedding ID: kBET\n",
            "Add of existing embedding ID: KMD\n",
            "Add of existing embedding ID: kmermaid\n",
            "Add of existing embedding ID: knn_smooth\n",
            "Add of existing embedding ID: KPNN\n",
            "Add of existing embedding ID: ktest\n",
            "Add of existing embedding ID: LAK\n",
            "Add of existing embedding ID: LAmbDA\n",
            "Add of existing embedding ID: Lamian\n",
            "Add of existing embedding ID: LandSCENT\n",
            "Add of existing embedding ID: LANTSA\n",
            "Add of existing embedding ID: LATE\n",
            "Add of existing embedding ID: LatentVelo\n",
            "Add of existing embedding ID: LDA_scRNAseq\n",
            "Add of existing embedding ID: leader\n",
            "Add of existing embedding ID: LEAP\n",
            "Add of existing embedding ID: learn2count\n",
            "Add of existing embedding ID: lemur\n",
            "Add of existing embedding ID: LIANA\n",
            "Add of existing embedding ID: Libra\n",
            "Add of existing embedding ID: LIGER\n",
            "Add of existing embedding ID: LINEAGE\n",
            "Add of existing embedding ID: LineagePulse\n",
            "Add of existing embedding ID: Linnorm\n",
            "Add of existing embedding ID: LinQView\n",
            "Add of existing embedding ID: LinRace\n",
            "Add of existing embedding ID: locaTE\n",
            "Add of existing embedding ID: locCSN\n",
            "Add of existing embedding ID: Longcell\n",
            "Add of existing embedding ID: LRLoop\n",
            "Add of existing embedding ID: LRT\n",
            "Add of existing embedding ID: LTMGSCA\n",
            "Add of existing embedding ID: M3Drop\n",
            "Add of existing embedding ID: M3S\n",
            "Add of existing embedding ID: MAAPER\n",
            "Add of existing embedding ID: MACA\n",
            "Add of existing embedding ID: MaCroDNA\n",
            "Add of existing embedding ID: MacroPolo\n",
            "Add of existing embedding ID: MAESTRO\n",
            "Add of existing embedding ID: MAGAN\n",
            "Add of existing embedding ID: MAGIC\n",
            "Add of existing embedding ID: MAPiT\n",
            "Add of existing embedding ID: MARGARET\n",
            "Add of existing embedding ID: MarkerCount\n",
            "Add of existing embedding ID: MarkerMap\n",
            "Add of existing embedding ID: MarkovHC\n",
            "Add of existing embedding ID: MARS\n",
            "Add of existing embedding ID: MASI\n",
            "Add of existing embedding ID: MAST\n",
            "Add of existing embedding ID: MAT2\n",
            "Add of existing embedding ID: MatchCLOT\n",
            "Add of existing embedding ID: MATCHER\n",
            "Add of existing embedding ID: matchSCore\n",
            "Add of existing embedding ID: matchSCore2\n",
            "Add of existing embedding ID: Matilda\n",
            "Add of existing embedding ID: MATTE\n",
            "Add of existing embedding ID: MaxFuse\n",
            "Add of existing embedding ID: mbkmeans\n",
            "Add of existing embedding ID: mcap\n",
            "Add of existing embedding ID: McImpute\n",
            "Add of existing embedding ID: MCXpress\n",
            "Add of existing embedding ID: MEBOCOST\n",
            "Add of existing embedding ID: MEDALT\n",
            "Add of existing embedding ID: MELD\n",
            "Add of existing embedding ID: MELT\n",
            "Add of existing embedding ID: memento\n",
            "Add of existing embedding ID: MERLoT\n",
            "Add of existing embedding ID: MetaCell\n",
            "Add of existing embedding ID: metacells\n",
            "Add of existing embedding ID: METAFlux\n",
            "Add of existing embedding ID: MetaNeighbor\n",
            "Add of existing embedding ID: MetaTiME\n",
            "Add of existing embedding ID: MFA\n",
            "Add of existing embedding ID: MGFR\n",
            "Add of existing embedding ID: MicroCellClust\n",
            "Add of existing embedding ID: MICTI\n",
            "Add of existing embedding ID: Millefy\n",
            "Add of existing embedding ID: miloR\n",
            "Add of existing embedding ID: MIMOSCA\n",
            "Add of existing embedding ID: MIND\n",
            "Add of existing embedding ID: minicore\n",
            "Add of existing embedding ID: minnow\n",
            "Add of existing embedding ID: miQC\n",
            "Add of existing embedding ID: MIRA\n",
            "Add of existing embedding ID: miRSCAPE\n",
            "Add of existing embedding ID: Miscell\n",
            "Add of existing embedding ID: mitch\n",
            "Add of existing embedding ID: MitoHEAR\n",
            "Add of existing embedding ID: mitoSplitter\n",
            "Add of existing embedding ID: MitoTrace\n",
            "Add of existing embedding ID: MiXCR\n",
            "Add of existing embedding ID: MLG\n",
            "Add of existing embedding ID: mmutil\n",
            "Add of existing embedding ID: mnnpy\n",
            "Add of existing embedding ID: Moana\n",
            "Add of existing embedding ID: MOFA\n",
            "Add of existing embedding ID: MOFA2\n",
            "Add of existing embedding ID: Monet\n",
            "Add of existing embedding ID: Monocle\n",
            "Add of existing embedding ID: Monod\n",
            "Add of existing embedding ID: Monopogen\n",
            "Add of existing embedding ID: mosaicMPI\n",
            "Add of existing embedding ID: MOSCATO\n",
            "Add of existing embedding ID: moscot\n",
            "Add of existing embedding ID: Mowgli\n",
            "Add of existing embedding ID: Mpath\n",
            "Add of existing embedding ID: mpCMF\n",
            "Add of existing embedding ID: MPSSC\n",
            "Add of existing embedding ID: MQuad\n",
            "Add of existing embedding ID: MRFscRNAseq\n",
            "Add of existing embedding ID: mrtree\n",
            "Add of existing embedding ID: MSIsensor-RNA\n",
            "Add of existing embedding ID: mtANN\n",
            "Add of existing embedding ID: MTD\n",
            "Add of existing embedding ID: MTGO-SC\n",
            "Add of existing embedding ID: mtSC\n",
            "Add of existing embedding ID: MUDAN\n",
            "Add of existing embedding ID: MuDCoD\n",
            "Add of existing embedding ID: multiGroupVI\n",
            "Add of existing embedding ID: MultiK\n",
            "Add of existing embedding ID: MultiMAP\n",
            "Add of existing embedding ID: multinichenetr\n",
            "Add of existing embedding ID: Multiscale_PHATE\n",
            "Add of existing embedding ID: MultiVelo\n",
            "Add of existing embedding ID: MUREN\n",
            "Add of existing embedding ID: MURP\n",
            "Add of existing embedding ID: muscat\n",
            "Add of existing embedding ID: MuSiC\n",
            "Add of existing embedding ID: MUSIC\n",
            "Add of existing embedding ID: MuTrans\n",
            "Add of existing embedding ID: mvTCR\n",
            "Add of existing embedding ID: N-ACT\n",
            "Add of existing embedding ID: NaiveDE\n",
            "Add of existing embedding ID: NanoPARE\n",
            "Add of existing embedding ID: NATMI\n",
            "Add of existing embedding ID: NBID\n",
            "Add of existing embedding ID: NEBULA\n",
            "Add of existing embedding ID: Nebulosa\n",
            "Add of existing embedding ID: Neighbor-seq\n",
            "Add of existing embedding ID: NetImpute\n",
            "Add of existing embedding ID: netNMF-sc\n",
            "Add of existing embedding ID: netSmooth\n",
            "Add of existing embedding ID: netsne\n",
            "Add of existing embedding ID: NetworkInference\n",
            "Add of existing embedding ID: NeuCA\n",
            "Add of existing embedding ID: NEUROeSTIMator\n",
            "Add of existing embedding ID: NeuronChat\n",
            "Add of existing embedding ID: NewWave\n",
            "Add of existing embedding ID: nichenetr\n",
            "Add of existing embedding ID: NICHES\n",
            "Add of existing embedding ID: NIFA\n",
            "Add of existing embedding ID: nimfa\n",
            "Add of existing embedding ID: NLBayes\n",
            "Add of existing embedding ID: NOMAD\n",
            "Add of existing embedding ID: NOMAD2\n",
            "Add of existing embedding ID: Normalisr\n",
            "Add of existing embedding ID: northstar\n",
            "Add of existing embedding ID: novoSpaRc\n",
            "Add of existing embedding ID: NSForest\n",
            "Add of existing embedding ID: OCAT\n",
            "Add of existing embedding ID: Ocelli\n",
            "Add of existing embedding ID: ODEGRfinder\n",
            "Add of existing embedding ID: OEFinder\n",
            "Add of existing embedding ID: OGFSC\n",
            "Add of existing embedding ID: OmicInt\n",
            "Add of existing embedding ID: OnClass\n",
            "Add of existing embedding ID: OnlinePCA\n",
            "Add of existing embedding ID: OpenBioMed\n",
            "Add of existing embedding ID: OpenVisuMap\n",
            "Add of existing embedding ID: ORIGINS\n",
            "Add of existing embedding ID: ORIGINS2\n",
            "Add of existing embedding ID: OscoNet\n",
            "Add of existing embedding ID: Oscope\n",
            "Add of existing embedding ID: Ouija\n",
            "Add of existing embedding ID: ouijaflow\n",
            "Add of existing embedding ID: Outrigger\n",
            "Add of existing embedding ID: PACSI\n",
            "Add of existing embedding ID: pagoda2\n",
            "Add of existing embedding ID: Palantir\n",
            "Add of existing embedding ID: PALMO\n",
            "Add of existing embedding ID: Palo\n",
            "Add of existing embedding ID: Pamona\n",
            "Add of existing embedding ID: Pando\n",
            "Add of existing embedding ID: Panpipes\n",
            "Add of existing embedding ID: ParaDPMM\n",
            "Add of existing embedding ID: PARC\n",
            "Add of existing embedding ID: PAUSE\n",
            "Add of existing embedding ID: PBA\n",
            "Add of existing embedding ID: PBLR\n",
            "Add of existing embedding ID: PCAone\n",
            "Add of existing embedding ID: pcaReduce\n",
            "Add of existing embedding ID: pCMF\n",
            "Add of existing embedding ID: pcqc\n",
            "Add of existing embedding ID: pCreode\n",
            "Add of existing embedding ID: peco\n",
            "Add of existing embedding ID: Pegasus\n",
            "Add of existing embedding ID: PENCIL\n",
            "Add of existing embedding ID: PercentMaxDiff\n",
            "Add of existing embedding ID: PHATE\n",
            "Add of existing embedding ID: PhEMD\n",
            "Add of existing embedding ID: PhenoPath\n",
            "Add of existing embedding ID: phenotype-cover\n",
            "Add of existing embedding ID: phiclust\n",
            "Add of existing embedding ID: PhitestR\n",
            "Add of existing embedding ID: PhylEx\n",
            "Add of existing embedding ID: phyloRNA\n",
            "Add of existing embedding ID: PhyloVelo\n",
            "Add of existing embedding ID: Piccolo\n",
            "Add of existing embedding ID: PieParty\n",
            "Add of existing embedding ID: PILOT\n",
            "Add of existing embedding ID: PIMseq\n",
            "Add of existing embedding ID: pipeComp\n",
            "Add of existing embedding ID: PISCES\n",
            "Add of existing embedding ID: PIVOT\n",
            "Add of existing embedding ID: pLDA\n",
            "Add of existing embedding ID: pmVAE\n",
            "Add of existing embedding ID: PoissonUMIs\n",
            "Add of existing embedding ID: Polaratio\n",
            "Add of existing embedding ID: Pollock\n",
            "Add of existing embedding ID: polyApipe\n",
            "Add of existing embedding ID: PopAlign\n",
            "Add of existing embedding ID: popsicleR\n",
            "Add of existing embedding ID: PopV\n",
            "Add of existing embedding ID: Portal\n",
            "Add of existing embedding ID: powerEQTL\n",
            "Add of existing embedding ID: POWSC\n",
            "Add of existing embedding ID: powsimR\n",
            "Add of existing embedding ID: PPMS\n",
            "Add of existing embedding ID: Pre-Phaser\n",
            "Add of existing embedding ID: PreCanCell\n",
            "Add of existing embedding ID: PRESCIENT\n",
            "Add of existing embedding ID: ProjecTILs\n",
            "Add of existing embedding ID: ProjectSVR\n",
            "Add of existing embedding ID: PROSSTT\n",
            "Add of existing embedding ID: protaccel\n",
            "Add of existing embedding ID: Pseudocell_Tracer\n",
            "Add of existing embedding ID: pseudodynamics\n",
            "Add of existing embedding ID: pseudoga\n",
            "Add of existing embedding ID: pseudogp\n",
            "Add of existing embedding ID: PseudotimeDE\n",
            "Add of existing embedding ID: PsiNorm\n",
            "Add of existing embedding ID: Psix\n",
            "Add of existing embedding ID: psupertime\n",
            "Add of existing embedding ID: pyHSICLasso\n",
            "Add of existing embedding ID: PyMINEr\n",
            "Add of existing embedding ID: pyNVR\n",
            "Add of existing embedding ID: Pyro-Velocity\n",
            "Add of existing embedding ID: PySCUBA\n",
            "Add of existing embedding ID: qSNE\n",
            "Add of existing embedding ID: quasildr\n",
            "Add of existing embedding ID: QUBIC2\n",
            "Add of existing embedding ID: RaceID\n",
            "Add of existing embedding ID: RAFSIL\n",
            "Add of existing embedding ID: RainDrop\n",
            "Add of existing embedding ID: Randomly\n",
            "Add of existing embedding ID: RankCorr\n",
            "Add of existing embedding ID: RayleighSelection\n",
            "Add of existing embedding ID: RCA\n",
            "Add of existing embedding ID: rCASC\n",
            "Add of existing embedding ID: RCAv2\n",
            "Add of existing embedding ID: RCI-NB\n",
            "Add of existing embedding ID: RcppML\n",
            "Add of existing embedding ID: RCSL\n",
            "Add of existing embedding ID: ReadZS\n",
            "Add of existing embedding ID: REBET\n",
            "Add of existing embedding ID: reCAT\n",
            "Add of existing embedding ID: RECODE\n",
            "Add of existing embedding ID: RECOMBINE\n",
            "Add of existing embedding ID: RecursiveConsensusClustering\n",
            "Add of existing embedding ID: Red_Panda\n",
            "Add of existing embedding ID: redPATH\n",
            "Add of existing embedding ID: Regionvelocity\n",
            "Add of existing embedding ID: RESCUE\n",
            "Add of existing embedding ID: ResPAN\n",
            "Add of existing embedding ID: Revelio\n",
            "Add of existing embedding ID: RgCop\n",
            "Add of existing embedding ID: RGGC\n",
            "Add of existing embedding ID: RISC\n",
            "Add of existing embedding ID: RNA-Bloom\n",
            "Add of existing embedding ID: RNA-combine\n",
            "Add of existing embedding ID: RNA-Sieve\n",
            "Add of existing embedding ID: RNA-strain-match\n",
            "Add of existing embedding ID: RNAForecaster\n",
            "Add of existing embedding ID: RobustCCC\n",
            "Add of existing embedding ID: robustSingleCell\n",
            "Add of existing embedding ID: ROGUE\n",
            "Add of existing embedding ID: ROSeq\n",
            "Add of existing embedding ID: ROSeq-R\n",
            "Add of existing embedding ID: rpca\n",
            "Add of existing embedding ID: RSCORE\n",
            "Add of existing embedding ID: Rsubread\n",
            "Add of existing embedding ID: rSuperCT\n",
            "Add of existing embedding ID: rusty_axe\n",
            "Add of existing embedding ID: ruvIIInb\n",
            "Add of existing embedding ID: RVAgene\n",
            "Add of existing embedding ID: RZiMM\n",
            "Add of existing embedding ID: SAFE\n",
            "Add of existing embedding ID: SageNet\n",
            "Add of existing embedding ID: SAIC\n",
            "Add of existing embedding ID: sake\n",
            "Add of existing embedding ID: salmon\n",
            "Add of existing embedding ID: SAM\n",
            "Add of existing embedding ID: SAMap\n",
            "Add of existing embedding ID: SampleQC\n",
            "Add of existing embedding ID: SAMstrt\n",
            "Add of existing embedding ID: Sanity\n",
            "Add of existing embedding ID: Sargent\n",
            "Add of existing embedding ID: SASCRiP\n",
            "Add of existing embedding ID: satuRn\n",
            "Add of existing embedding ID: SATURN\n",
            "Add of existing embedding ID: SAUCIE\n",
            "Add of existing embedding ID: saVAE\n",
            "Add of existing embedding ID: SAVER\n",
            "Add of existing embedding ID: SAVER-X\n",
            "Add of existing embedding ID: SBDN\n",
            "Add of existing embedding ID: sc-compReg\n",
            "Add of existing embedding ID: sc-jnmf\n",
            "Add of existing embedding ID: sc-Libra\n",
            "Add of existing embedding ID: sc-REnF\n",
            "Add of existing embedding ID: sc2marker\n",
            "Add of existing embedding ID: sc2MeNetDrug\n",
            "Add of existing embedding ID: SC2P\n",
            "Add of existing embedding ID: SC2Spa\n",
            "Add of existing embedding ID: SC3\n",
            "Add of existing embedding ID: SC3s\n",
            "Add of existing embedding ID: SCA\n",
            "Add of existing embedding ID: scAAnet\n",
            "Add of existing embedding ID: scAce\n",
            "Add of existing embedding ID: scACE\n",
            "Add of existing embedding ID: scAdapt\n",
            "Add of existing embedding ID: scAEGAN\n",
            "Add of existing embedding ID: scAEspy\n",
            "Add of existing embedding ID: SCAFE\n",
            "Add of existing embedding ID: scaffold\n",
            "Add of existing embedding ID: scAI\n",
            "Add of existing embedding ID: scAIDE\n",
            "Add of existing embedding ID: scAL\n",
            "Add of existing embedding ID: SCALE\n",
            "Add of existing embedding ID: SCALEX\n",
            "Add of existing embedding ID: scAlign\n",
            "Add of existing embedding ID: scAllele\n",
            "Add of existing embedding ID: scAlleleCount\n",
            "Add of existing embedding ID: Scallop2\n",
            "Add of existing embedding ID: scAMACE\n",
            "Add of existing embedding ID: scAMACE_py\n",
            "Add of existing embedding ID: scAmpi\n",
            "Add of existing embedding ID: scAN10\n",
            "Add of existing embedding ID: scanalysis\n",
            "Add of existing embedding ID: scAnCluster\n",
            "Add of existing embedding ID: SCANet\n",
            "Add of existing embedding ID: scANNA\n",
            "Add of existing embedding ID: scAnno\n",
            "Add of existing embedding ID: scAnnotatR\n",
            "Add of existing embedding ID: scanorama\n",
            "Add of existing embedding ID: Scanpro\n",
            "Add of existing embedding ID: Scanpy\n",
            "Add of existing embedding ID: scAPA\n",
            "Add of existing embedding ID: scAPAtrap\n",
            "Add of existing embedding ID: SCAPE\n",
            "Add of existing embedding ID: SCAPTURE\n",
            "Add of existing embedding ID: scAR\n",
            "Add of existing embedding ID: scArches\n",
            "Add of existing embedding ID: scarf\n",
            "Add of existing embedding ID: scasa\n",
            "Add of existing embedding ID: scASfind\n",
            "Add of existing embedding ID: scASK\n",
            "Add of existing embedding ID: Scater\n",
            "Add of existing embedding ID: scATOMIC\n",
            "Add of existing embedding ID: SCATrEx\n",
            "Add of existing embedding ID: SCATS\n",
            "Add of existing embedding ID: scatterHatch\n",
            "Add of existing embedding ID: SCATTome\n",
            "Add of existing embedding ID: scBASE\n",
            "Add of existing embedding ID: scBatch\n",
            "Add of existing embedding ID: scbean\n",
            "Add of existing embedding ID: scBERT\n",
            "Add of existing embedding ID: scBGEDA\n",
            "Add of existing embedding ID: scBio\n",
            "Add of existing embedding ID: scBonita\n",
            "Add of existing embedding ID: scBoolSeq\n",
            "Add of existing embedding ID: scBridge\n",
            "Add of existing embedding ID: scBT\n",
            "Add of existing embedding ID: scBTF\n",
            "Add of existing embedding ID: scBubbletree\n",
            "Add of existing embedding ID: SCC\n",
            "Add of existing embedding ID: scCAEs\n",
            "Add of existing embedding ID: SCCAF\n",
            "Add of existing embedding ID: scCAN\n",
            "Add of existing embedding ID: scCancer\n",
            "Add of existing embedding ID: scCancer2\n",
            "Add of existing embedding ID: scCaps\n",
            "Add of existing embedding ID: scCapsNet-mask\n",
            "Add of existing embedding ID: scCATCH\n",
            "Add of existing embedding ID: scCB2\n",
            "Add of existing embedding ID: scCCESS\n",
            "Add of existing embedding ID: scClassifR\n",
            "Add of existing embedding ID: scClassify\n",
            "Add of existing embedding ID: scCLUE\n",
            "Add of existing embedding ID: scClustBench\n",
            "Add of existing embedding ID: scclusteval\n",
            "Add of existing embedding ID: scClustViz\n",
            "Add of existing embedding ID: scCobra\n",
            "Add of existing embedding ID: scCODA\n",
            "Add of existing embedding ID: scCODE\n",
            "Add of existing embedding ID: sccomp\n",
            "Add of existing embedding ID: scConnect\n",
            "Add of existing embedding ID: scConsensus\n",
            "Add of existing embedding ID: scCorr\n",
            "Add of existing embedding ID: scCustomize\n",
            "Add of existing embedding ID: scDA\n",
            "Add of existing embedding ID: scDALI\n",
            "Add of existing embedding ID: scDAPA\n",
            "Add of existing embedding ID: scDaPars\n",
            "Add of existing embedding ID: scDART\n",
            "Add of existing embedding ID: scDataviz\n",
            "Add of existing embedding ID: scDblFinder\n",
            "Add of existing embedding ID: scDBM\n",
            "Add of existing embedding ID: scDC\n",
            "Add of existing embedding ID: scDCC\n",
            "Add of existing embedding ID: scDD\n",
            "Add of existing embedding ID: scDDboost\n",
            "Add of existing embedding ID: SCDE\n",
            "Add of existing embedding ID: scDEA\n",
            "Add of existing embedding ID: scDEAL\n",
            "Add of existing embedding ID: scDECAF\n",
            "Add of existing embedding ID: scDecouple\n",
            "Add of existing embedding ID: scDEED\n",
            "Add of existing embedding ID: scDeepCluster\n",
            "Add of existing embedding ID: scDeepHash\n",
            "Add of existing embedding ID: scDeepInsight\n",
            "Add of existing embedding ID: scDeepSort\n",
            "Add of existing embedding ID: scDesign\n",
            "Add of existing embedding ID: scDesign2\n",
            "Add of existing embedding ID: scDesign3\n",
            "Add of existing embedding ID: scDetect\n",
            "Add of existing embedding ID: scDGN\n",
            "Add of existing embedding ID: scDHA\n",
            "Add of existing embedding ID: scDHMap\n",
            "Add of existing embedding ID: scdiff\n",
            "Add of existing embedding ID: scDiffCom\n",
            "Add of existing embedding ID: scDisInFact\n",
            "Add of existing embedding ID: scDist\n",
            "Add of existing embedding ID: scdNet\n",
            "Add of existing embedding ID: scdney\n",
            "Add of existing embedding ID: scDoc\n",
            "Add of existing embedding ID: scdrake\n",
            "Add of existing embedding ID: scDREAMER\n",
            "Add of existing embedding ID: scDRS\n",
            "Add of existing embedding ID: scDrug\n",
            "Add of existing embedding ID: scds\n",
            "Add of existing embedding ID: sceb\n",
            "Add of existing embedding ID: scEC\n",
            "Add of existing embedding ID: Scedar\n",
            "Add of existing embedding ID: scEFSC\n",
            "Add of existing embedding ID: SCell\n",
            "Add of existing embedding ID: SCellBOW\n",
            "Add of existing embedding ID: scellpam\n",
            "Add of existing embedding ID: SCelVis\n",
            "Add of existing embedding ID: scEMAIL\n",
            "Add of existing embedding ID: SCENA\n",
            "Add of existing embedding ID: SCENIC\n",
            "Add of existing embedding ID: scenicplus\n",
            "Add of existing embedding ID: SCEnt\n",
            "Add of existing embedding ID: SCENT\n",
            "Add of existing embedding ID: scEpath\n",
            "Add of existing embedding ID: scETM\n",
            "Add of existing embedding ID: SCEVAN\n",
            "Add of existing embedding ID: scEvoNet\n",
            "Add of existing embedding ID: scFates\n",
            "Add of existing embedding ID: scFBA\n",
            "Add of existing embedding ID: scFEA\n",
            "Add of existing embedding ID: scFeatureFilter\n",
            "Add of existing embedding ID: scFeatures\n",
            "Add of existing embedding ID: scfind\n",
            "Add of existing embedding ID: scFlow\n",
            "Add of existing embedding ID: scFormer\n",
            "Add of existing embedding ID: scFseCluster\n",
            "Add of existing embedding ID: scFusion\n",
            "Add of existing embedding ID: scGAE\n",
            "Add of existing embedding ID: scGAN\n",
            "Add of existing embedding ID: scGate\n",
            "Add of existing embedding ID: scGBM\n",
            "Add of existing embedding ID: scGCL\n",
            "Add of existing embedding ID: scGEAToolbox\n",
            "Add of existing embedding ID: scGen\n",
            "Add of existing embedding ID: scGeneFit\n",
            "Add of existing embedding ID: scGeom\n",
            "Add of existing embedding ID: scGESTALT\n",
            "Add of existing embedding ID: scGNN\n",
            "Add of existing embedding ID: scGNN2\n",
            "Add of existing embedding ID: scGPS\n",
            "Add of existing embedding ID: scGPT\n",
            "Add of existing embedding ID: scGraph\n",
            "Add of existing embedding ID: scGraphNE\n",
            "Add of existing embedding ID: scGREAT\n",
            "Add of existing embedding ID: scGTM\n",
            "Add of existing embedding ID: scGWAS\n",
            "Add of existing embedding ID: schema\n",
            "Add of existing embedding ID: schex\n",
            "Add of existing embedding ID: scHinter\n",
            "Add of existing embedding ID: SCHiRM\n",
            "Add of existing embedding ID: schist\n",
            "Add of existing embedding ID: scHLAcount\n",
            "Add of existing embedding ID: SCHNAPPs\n",
            "Add of existing embedding ID: SCHNEL\n",
            "Add of existing embedding ID: scHOT\n",
            "Add of existing embedding ID: scHPF\n",
            "Add of existing embedding ID: scHPL\n",
            "Add of existing embedding ID: scHumanNet\n",
            "Add of existing embedding ID: scIAE\n",
            "Add of existing embedding ID: scIB\n",
            "Add of existing embedding ID: SCIBER\n",
            "Add of existing embedding ID: SciBet\n",
            "Add of existing embedding ID: sciCAN\n",
            "Add of existing embedding ID: scicast\n",
            "Add of existing embedding ID: sciCSR\n",
            "Add of existing embedding ID: scID\n",
            "Add of existing embedding ID: scIGANs\n",
            "Add of existing embedding ID: scIHPF\n",
            "Add of existing embedding ID: SCILEN\n",
            "Add of existing embedding ID: Scillus\n",
            "Add of existing embedding ID: SCIM\n",
            "Add of existing embedding ID: SCIMITAR\n",
            "Add of existing embedding ID: scImpute\n",
            "Add of existing embedding ID: SCINA\n",
            "Add of existing embedding ID: SCINET\n",
            "Add of existing embedding ID: SCING\n",
            "Add of existing embedding ID: SCING-pipeline\n",
            "Add of existing embedding ID: scINSIGHT\n",
            "Add of existing embedding ID: scIntegral\n",
            "Add of existing embedding ID: sciPENN\n",
            "Add of existing embedding ID: scipr\n",
            "Add of existing embedding ID: scirpy\n",
            "Add of existing embedding ID: scisorseqr\n",
            "Add of existing embedding ID: ScisorWiz\n",
            "Add of existing embedding ID: scISR\n",
            "Add of existing embedding ID: SCISSORS\n",
            "Add of existing embedding ID: scITD\n",
            "Add of existing embedding ID: scIVA\n",
            "Add of existing embedding ID: Sciviewer\n",
            "Add of existing embedding ID: scJoint\n",
            "Add of existing embedding ID: scKB\n",
            "Add of existing embedding ID: scKINETICS\n",
            "Add of existing embedding ID: scLAPA\n",
            "Add of existing embedding ID: scLCA\n",
            "Add of existing embedding ID: scLINE\n",
            "Add of existing embedding ID: SClineager\n",
            "Add of existing embedding ID: scLink\n",
            "Add of existing embedding ID: scLRTC\n",
            "Add of existing embedding ID: scLVM\n",
            "Add of existing embedding ID: scMAGeCK\n",
            "Add of existing embedding ID: scMAGIC\n",
            "Add of existing embedding ID: scMAGS\n",
            "Add of existing embedding ID: scmap\n",
            "Add of existing embedding ID: scMAPA\n",
            "Add of existing embedding ID: scMappR\n",
            "Add of existing embedding ID: SCMAPpy\n",
            "Add of existing embedding ID: SCMarker\n",
            "Add of existing embedding ID: scMatch\n",
            "Add of existing embedding ID: scMayoMap\n",
            "Add of existing embedding ID: scMC\n",
            "Add of existing embedding ID: scMCA\n",
            "Add of existing embedding ID: scMDC\n",
            "Add of existing embedding ID: scMEGA\n",
            "Add of existing embedding ID: SCMER\n",
            "Add of existing embedding ID: scMerge\n",
            "Add of existing embedding ID: scMetric\n",
            "Add of existing embedding ID: scMGCA\n",
            "Add of existing embedding ID: scMiko\n",
            "Add of existing embedding ID: scMINER\n",
            "Add of existing embedding ID: scmixture\n",
            "Add of existing embedding ID: scMLnet\n",
            "Add of existing embedding ID: scMM\n",
            "Add of existing embedding ID: scModels\n",
            "Add of existing embedding ID: scMOG\n",
            "Add of existing embedding ID: scMoMaT\n",
            "Add of existing embedding ID: scMomentum\n",
            "Add of existing embedding ID: scMOO\n",
            "Add of existing embedding ID: scMRA\n",
            "Add of existing embedding ID: scMRMA\n",
            "Add of existing embedding ID: scMTD\n",
            "Add of existing embedding ID: scMuffin\n",
            "Add of existing embedding ID: scMultiSim\n",
            "Add of existing embedding ID: SCmut\n",
            "Add of existing embedding ID: scMVAE\n",
            "Add of existing embedding ID: scNanoGPS\n",
            "Add of existing embedding ID: scNBMF\n",
            "Add of existing embedding ID: scNCA\n",
            "Add of existing embedding ID: scNetViz\n",
            "Add of existing embedding ID: SCNorm\n",
            "Add of existing embedding ID: scNPF\n",
            "Add of existing embedding ID: SCNS\n",
            "Add of existing embedding ID: scNym\n",
            "Add of existing embedding ID: SCODE\n",
            "Add of existing embedding ID: SCOIT\n",
            "Add of existing embedding ID: SComatic\n",
            "Add of existing embedding ID: scone\n",
            "Add of existing embedding ID: scOntoMatch\n",
            "Add of existing embedding ID: SCope\n",
            "Add of existing embedding ID: scopeplus\n",
            "Add of existing embedding ID: SCORPION\n",
            "Add of existing embedding ID: SCORPIUS\n",
            "Add of existing embedding ID: SCOUP\n",
            "Add of existing embedding ID: SCOUT\n",
            "Add of existing embedding ID: scover\n",
            "Add of existing embedding ID: SCP\n",
            "Add of existing embedding ID: SCPA\n",
            "Add of existing embedding ID: scPagwas\n",
            "Add of existing embedding ID: scPanoView\n",
            "Add of existing embedding ID: SCPattern\n",
            "Add of existing embedding ID: scPCA\n",
            "Add of existing embedding ID: scphaser\n",
            "Add of existing embedding ID: scPheno\n",
            "Add of existing embedding ID: scPhere\n",
            "Add of existing embedding ID: scPipe\n",
            "Add of existing embedding ID: scPlant\n",
            "Add of existing embedding ID: scploid\n",
            "Add of existing embedding ID: scPlot\n",
            "Add of existing embedding ID: scPNMF\n",
            "Add of existing embedding ID: scpoisson\n",
            "Add of existing embedding ID: scPOP\n",
            "Add of existing embedding ID: scPopCorn\n",
            "Add of existing embedding ID: scPOST\n",
            "Add of existing embedding ID: scPower\n",
            "Add of existing embedding ID: scPred\n",
            "Add of existing embedding ID: scprep\n",
            "Add of existing embedding ID: scPretrain\n",
            "Add of existing embedding ID: scPrisma\n",
            "Add of existing embedding ID: scProject\n",
            "Add of existing embedding ID: scProximitE\n",
            "Add of existing embedding ID: scPSD\n",
            "Add of existing embedding ID: scPSM\n",
            "Add of existing embedding ID: SCpubr\n",
            "Add of existing embedding ID: scQA\n",
            "Add of existing embedding ID: SCQC\n",
            "Add of existing embedding ID: scQCEA\n",
            "Add of existing embedding ID: scQcut\n",
            "Add of existing embedding ID: scQuint\n",
            "Add of existing embedding ID: SCRABBLE\n",
            "Add of existing embedding ID: scRAE\n",
            "Add of existing embedding ID: scran\n",
            "Add of existing embedding ID: scRANK\n",
            "Add of existing embedding ID: scRare\n",
            "Add of existing embedding ID: scRAS\n",
            "Add of existing embedding ID: SCRAT\n",
            "Add of existing embedding ID: scRCMF\n",
            "Add of existing embedding ID: SCReadCounts\n",
            "Add of existing embedding ID: scReadSim\n",
            "Add of existing embedding ID: scReClassify\n",
            "Add of existing embedding ID: scRecover\n",
            "Add of existing embedding ID: scREhurdle\n",
            "Add of existing embedding ID: scriabin\n",
            "Add of existing embedding ID: Scribe\n",
            "Add of existing embedding ID: SCRIP\n",
            "Add of existing embedding ID: SCRL\n",
            "Add of existing embedding ID: scRMD\n",
            "Add of existing embedding ID: scRNA\n",
            "Add of existing embedding ID: scRNABatchQC\n",
            "Add of existing embedding ID: scRNAseqApp\n",
            "Add of existing embedding ID: scRNASequest\n",
            "Add of existing embedding ID: scRNAss\n",
            "Add of existing embedding ID: scROSHI\n",
            "Add of existing embedding ID: Scrublet\n",
            "Add of existing embedding ID: scruff\n",
            "Add of existing embedding ID: scRutiNy\n",
            "Add of existing embedding ID: scry\n",
            "Add of existing embedding ID: SCSA\n",
            "Add of existing embedding ID: scSampler\n",
            "Add of existing embedding ID: SCSC\n",
            "Add of existing embedding ID: scScope\n",
            "Add of existing embedding ID: scSDAE\n",
            "Add of existing embedding ID: scSemiAE\n",
            "Add of existing embedding ID: scSemiCluster\n",
            "Add of existing embedding ID: scSensitiveGeneDefine\n",
            "Add of existing embedding ID: scSeqComm\n",
            "Add of existing embedding ID: scSeqR\n",
            "Add of existing embedding ID: scShaper\n",
            "Add of existing embedding ID: scShapes\n",
            "Add of existing embedding ID: scSHARP\n",
            "Add of existing embedding ID: scsim\n",
            "Add of existing embedding ID: scSimClassify\n",
            "Add of existing embedding ID: SCSit\n",
            "Add of existing embedding ID: scSNPdemux\n",
            "Add of existing embedding ID: scsnv\n",
            "Add of existing embedding ID: scSO\n",
            "Add of existing embedding ID: scSorter\n",
            "Add of existing embedding ID: scSpace\n",
            "Add of existing embedding ID: scSPARKL\n",
            "Add of existing embedding ID: scSplit\n",
            "Add of existing embedding ID: scSTEM\n",
            "Add of existing embedding ID: scSVA\n",
            "Add of existing embedding ID: scTab\n",
            "Add of existing embedding ID: scTCRseq\n",
            "Add of existing embedding ID: scTDA\n",
            "Add of existing embedding ID: scTE\n",
            "Add of existing embedding ID: scTenifoldKnk\n",
            "Add of existing embedding ID: scTenifoldNet\n",
            "Add of existing embedding ID: scTenifoldpy\n",
            "Add of existing embedding ID: scTensor\n",
            "Add of existing embedding ID: scTEP\n",
            "Add of existing embedding ID: scTGIF\n",
            "Add of existing embedding ID: scTHI\n",
            "Add of existing embedding ID: scTIM\n",
            "Add of existing embedding ID: scTITANS\n",
            "Add of existing embedding ID: scTM\n",
            "Add of existing embedding ID: scTOP\n",
            "Add of existing embedding ID: scTopoGAN\n",
            "Add of existing embedding ID: scTour\n",
            "Add of existing embedding ID: sctransform\n",
            "Add of existing embedding ID: scTree\n",
            "Add of existing embedding ID: scTreeViz\n",
            "Add of existing embedding ID: scTriangulate\n",
            "Add of existing embedding ID: scTSSR\n",
            "Add of existing embedding ID: ScType\n",
            "Add of existing embedding ID: scTyper\n",
            "Add of existing embedding ID: SCUBA\n",
            "Add of existing embedding ID: SCUBI\n",
            "Add of existing embedding ID: scumi\n",
            "Add of existing embedding ID: scUnifrac\n",
            "Add of existing embedding ID: SCUREL\n",
            "Add of existing embedding ID: scUTRquant\n",
            "Add of existing embedding ID: scuttle\n",
            "Add of existing embedding ID: scVAE\n",
            "Add of existing embedding ID: scVAEIT\n",
            "Add of existing embedding ID: scVDMC\n",
            "Add of existing embedding ID: scVEGs\n",
            "Add of existing embedding ID: scVelo\n",
            "Add of existing embedding ID: scvi-tools\n",
            "Add of existing embedding ID: scViewer\n",
            "Add of existing embedding ID: scViralQuant\n",
            "Add of existing embedding ID: scVirtualCyto\n",
            "Add of existing embedding ID: scvis\n",
            "Add of existing embedding ID: scvr\n",
            "Add of existing embedding ID: scWizard\n",
            "Add of existing embedding ID: scziDesk\n",
            "Add of existing embedding ID: SDAMS\n",
            "Add of existing embedding ID: SDImpute\n",
            "Add of existing embedding ID: SEACells\n",
            "Add of existing embedding ID: SEAT\n",
            "Add of existing embedding ID: secDrug\n",
            "Add of existing embedding ID: Secuer\n",
            "Add of existing embedding ID: SEDIM\n",
            "Add of existing embedding ID: SEGtool\n",
            "Add of existing embedding ID: SelfE\n",
            "Add of existing embedding ID: SELINA\n",
            "Add of existing embedding ID: SEMITONES\n",
            "Add of existing embedding ID: SenePy\n",
            "Add of existing embedding ID: SEPA\n",
            "Add of existing embedding ID: SEPIRA\n",
            "Add of existing embedding ID: seqgendiff\n",
            "Add of existing embedding ID: SEQUIN\n",
            "Add of existing embedding ID: SERGIO\n",
            "Add of existing embedding ID: SERM\n",
            "Add of existing embedding ID: Seurat\n",
            "Add of existing embedding ID: SeuratV3Wizard\n",
            "Add of existing embedding ID: SeuratWizard\n",
            "Add of existing embedding ID: sfaira\n",
            "Add of existing embedding ID: ShareNet\n",
            "Add of existing embedding ID: SHARP\n",
            "Add of existing embedding ID: Sharq\n",
            "Add of existing embedding ID: ShinyCell\n",
            "Add of existing embedding ID: ShinyMultiome-UiO\n",
            "Add of existing embedding ID: SiCeLoRe\n",
            "Add of existing embedding ID: SICILIAN\n",
            "Add of existing embedding ID: SiCR\n",
            "Add of existing embedding ID: Sierra\n",
            "Add of existing embedding ID: SIEVE\n",
            "Add of existing embedding ID: SiFINeT\n",
            "Add of existing embedding ID: SiFT\n",
            "Add of existing embedding ID: SigEMD\n",
            "Add of existing embedding ID: sigGCN\n",
            "Add of existing embedding ID: SigHotSpotter\n",
            "Add of existing embedding ID: SIGMA\n",
            "Add of existing embedding ID: signac\n",
            "Add of existing embedding ID: SignacX\n",
            "Add of existing embedding ID: signifinder\n",
            "Add of existing embedding ID: SILGGM\n",
            "Add of existing embedding ID: SIMBA\n",
            "Add of existing embedding ID: SimCD\n",
            "Add of existing embedding ID: SimiC\n",
            "Add of existing embedding ID: SIMLR\n",
            "Add of existing embedding ID: simpleaf\n",
            "Add of existing embedding ID: SIMPLEs\n",
            "Add of existing embedding ID: SIMS\n",
            "Add of existing embedding ID: simspec\n",
            "Add of existing embedding ID: SINC\n",
            "Add of existing embedding ID: Sincast\n",
            "Add of existing embedding ID: sincell\n",
            "Add of existing embedding ID: SINCERA\n",
            "Add of existing embedding ID: SINCERITIES\n",
            "Add of existing embedding ID: SinCHet\n",
            "Add of existing embedding ID: SingCellaR\n",
            "Add of existing embedding ID: SINGE\n",
            "Add of existing embedding ID: SingleCellExplorer\n",
            "Add of existing embedding ID: SingleCellGGM\n",
            "Add of existing embedding ID: singleCellHaystack\n",
            "Add of existing embedding ID: singleCellHaystack-py\n",
            "Add of existing embedding ID: SingleCellNet\n",
            "Add of existing embedding ID: SingleCellSignalR\n",
            "Add of existing embedding ID: SingleCellSignatureExplorer\n",
            "Add of existing embedding ID: singleCellTK\n",
            "Add of existing embedding ID: SinglePointRNA\n",
            "Add of existing embedding ID: SingleR\n",
            "Add of existing embedding ID: singleseqgset\n",
            "Add of existing embedding ID: SingleSplice\n",
            "Add of existing embedding ID: singlet\n",
            "Add of existing embedding ID: SinNLRR\n",
            "Add of existing embedding ID: SinQC\n",
            "Add of existing embedding ID: Sinto\n",
            "Add of existing embedding ID: SIP\n",
            "Add of existing embedding ID: SiPSiC\n",
            "Add of existing embedding ID: sircel\n",
            "Add of existing embedding ID: SISUA\n",
            "Add of existing embedding ID: siVAE\n",
            "Add of existing embedding ID: SJD\n",
            "Add of existing embedding ID: SkewC\n",
            "Add of existing embedding ID: slalom\n",
            "Add of existing embedding ID: sleepwalk\n",
            "Add of existing embedding ID: SLICE\n",
            "Add of existing embedding ID: SLICER\n",
            "Add of existing embedding ID: SLIDE\n",
            "Add of existing embedding ID: slingshot\n",
            "Add of existing embedding ID: SMAI\n",
            "Add of existing embedding ID: SMaSH\n",
            "Add of existing embedding ID: SMGR\n",
            "Add of existing embedding ID: SMILE\n",
            "Add of existing embedding ID: SMNN\n",
            "Add of existing embedding ID: SMSC\n",
            "Add of existing embedding ID: SMURF\n",
            "Add of existing embedding ID: SnapATAC2\n",
            "Add of existing embedding ID: SnapCCESS\n",
            "Add of existing embedding ID: SNN-Cliq\n",
            "Add of existing embedding ID: SoCube\n",
            "Add of existing embedding ID: solo\n",
            "Add of existing embedding ID: SoloTE\n",
            "Add of existing embedding ID: SoptSC\n",
            "Add of existing embedding ID: SOUP\n",
            "Add of existing embedding ID: souporcell\n",
            "Add of existing embedding ID: SoupX\n",
            "Add of existing embedding ID: SPADE\n",
            "Add of existing embedding ID: SpaOTsc\n",
            "Add of existing embedding ID: sparca\n",
            "Add of existing embedding ID: SparseDC\n",
            "Add of existing embedding ID: SPARSim\n",
            "Add of existing embedding ID: SPaRTAN\n",
            "Add of existing embedding ID: SpatialScope\n",
            "Add of existing embedding ID: SPECK\n",
            "Add of existing embedding ID: speckle\n",
            "Add of existing embedding ID: Specter\n",
            "Add of existing embedding ID: SPECTRA\n",
            "Add of existing embedding ID: Spectre\n",
            "Add of existing embedding ID: Spectrum\n",
            "Add of existing embedding ID: Sphetcher\n",
            "Add of existing embedding ID: splattdr\n",
            "Add of existing embedding ID: Splatter\n",
            "Add of existing embedding ID: spliceJAC\n",
            "Add of existing embedding ID: SPmarker\n",
            "Add of existing embedding ID: SPRING\n",
            "Add of existing embedding ID: SPsimSeq\n",
            "Add of existing embedding ID: SR2\n",
            "Add of existing embedding ID: SSBER\n",
            "Add of existing embedding ID: sscClust\n",
            "Add of existing embedding ID: SSrGE\n",
            "Add of existing embedding ID: StabMap\n",
            "Add of existing embedding ID: STACAS\n",
            "Add of existing embedding ID: STAR\n",
            "Add of existing embedding ID: STARE\n",
            "Add of existing embedding ID: starmapVis\n",
            "Add of existing embedding ID: StationaryOT\n",
            "Add of existing embedding ID: STEGO-R\n",
            "Add of existing embedding ID: STEM\n",
            "Add of existing embedding ID: STEMNET\n",
            "Add of existing embedding ID: STREAK\n",
            "Add of existing embedding ID: STREAM\n",
            "Add of existing embedding ID: supCPM\n",
            "Add of existing embedding ID: SuperCell\n",
            "Add of existing embedding ID: supercells\n",
            "Add of existing embedding ID: Superscan\n",
            "Add of existing embedding ID: SwarnSeq\n",
            "Add of existing embedding ID: switchde\n",
            "Add of existing embedding ID: SWNE\n",
            "Add of existing embedding ID: Symphony\n",
            "Add of existing embedding ID: SymSim\n",
            "Add of existing embedding ID: SynergisticCore\n",
            "Add of existing embedding ID: TACTiCS\n",
            "Add of existing embedding ID: talklr\n",
            "Add of existing embedding ID: Tangram\n",
            "Add of existing embedding ID: TAR-scRNA-seq\n",
            "Add of existing embedding ID: TASC\n",
            "Add of existing embedding ID: tascCODA\n",
            "Add of existing embedding ID: TASIC\n",
            "Add of existing embedding ID: tauFisher\n",
            "Add of existing embedding ID: TBSP\n",
            "Add of existing embedding ID: TCM\n",
            "Add of existing embedding ID: TDEseq\n",
            "Add of existing embedding ID: TDL\n",
            "Add of existing embedding ID: TedSim\n",
            "Add of existing embedding ID: Tempo\n",
            "Add of existing embedding ID: Tempora\n",
            "Add of existing embedding ID: TENET\n",
            "Add of existing embedding ID: TFvelo\n",
            "Add of existing embedding ID: tGPT\n",
            "Add of existing embedding ID: TILPRED\n",
            "Add of existing embedding ID: TimeTalk\n",
            "Add of existing embedding ID: TimiGP\n",
            "Add of existing embedding ID: TinGa\n",
            "Add of existing embedding ID: TITAN\n",
            "Add of existing embedding ID: tn_test\n",
            "Add of existing embedding ID: Tobiko\n",
            "Add of existing embedding ID: TOMAS\n",
            "Add of existing embedding ID: TooManyCells\n",
            "Add of existing embedding ID: TooManyCellsInteractive\n",
            "Add of existing embedding ID: TopOMetry\n",
            "Add of existing embedding ID: topslam\n",
            "Add of existing embedding ID: TOSICA\n",
            "Add of existing embedding ID: Totem\n",
            "Add of existing embedding ID: TraCeR\n",
            "Add of existing embedding ID: trackplot\n",
            "Add of existing embedding ID: tradeSeq\n",
            "Add of existing embedding ID: TrAGEDy\n",
            "Add of existing embedding ID: Trajan\n",
            "Add of existing embedding ID: TrajectoryNet\n",
            "Add of existing embedding ID: transformGamPoi\n",
            "Add of existing embedding ID: Transmorph\n",
            "Add of existing embedding ID: TRAPeS\n",
            "Add of existing embedding ID: TraSig\n",
            "Add of existing embedding ID: traviz\n",
            "Add of existing embedding ID: Tree-SNE\n",
            "Add of existing embedding ID: TreeCorTreat\n",
            "Add of existing embedding ID: TREG\n",
            "Add of existing embedding ID: trendsceek\n",
            "Add of existing embedding ID: Triangulate\n",
            "Add of existing embedding ID: tricycle\n",
            "Add of existing embedding ID: Triku\n",
            "Add of existing embedding ID: TripletCell\n",
            "Add of existing embedding ID: TRIPOD\n",
            "Add of existing embedding ID: TRUmiCount\n",
            "Add of existing embedding ID: TRUST4\n",
            "Add of existing embedding ID: trVAE\n",
            "Add of existing embedding ID: TSCAN\n",
            "Add of existing embedding ID: TSEE\n",
            "Add of existing embedding ID: Tweedieverse\n",
            "Add of existing embedding ID: TWO-SIGMA\n",
            "Add of existing embedding ID: txburst\n",
            "Add of existing embedding ID: UCell\n",
            "Add of existing embedding ID: ucie\n",
            "Add of existing embedding ID: UICPC\n",
            "Add of existing embedding ID: UMI_analysis\n",
            "Add of existing embedding ID: UMI-tools\n",
            "Add of existing embedding ID: UMICollapse\n",
            "Add of existing embedding ID: umis\n",
            "Add of existing embedding ID: UNCURL\n",
            "Add of existing embedding ID: UNIFAN\n",
            "Add of existing embedding ID: UnionCom\n",
            "Add of existing embedding ID: UniPath\n",
            "Add of existing embedding ID: uniPort\n",
            "Add of existing embedding ID: UniTVelo\n",
            "Add of existing embedding ID: UniverSC\n",
            "Add of existing embedding ID: UPhyloplot2\n",
            "Add of existing embedding ID: URD\n",
            "Add of existing embedding ID: uSORT\n",
            "Add of existing embedding ID: VAE-SNE\n",
            "Add of existing embedding ID: vaeda\n",
            "Add of existing embedding ID: VALERIE\n",
            "Add of existing embedding ID: valiDrops\n",
            "Add of existing embedding ID: VAM\n",
            "Add of existing embedding ID: VAMF\n",
            "Add of existing embedding ID: VarTrix\n",
            "Add of existing embedding ID: VASC\n",
            "Add of existing embedding ID: VDJdive\n",
            "Add of existing embedding ID: VDJPuzzle\n",
            "Add of existing embedding ID: VECTOR\n",
            "Add of existing embedding ID: VEGA\n",
            "Add of existing embedding ID: velociraptor\n",
            "Add of existing embedding ID: velocyto\n",
            "Add of existing embedding ID: Velorama\n",
            "Add of existing embedding ID: VeloRep\n",
            "Add of existing embedding ID: VeloSim\n",
            "Add of existing embedding ID: VeloVAE\n",
            "Add of existing embedding ID: VeloViz\n",
            "Add of existing embedding ID: velvet\n",
            "Add of existing embedding ID: Venus\n",
            "Add of existing embedding ID: VeTra\n",
            "Add of existing embedding ID: VIA\n",
            "Add of existing embedding ID: VIPCCA\n",
            "Add of existing embedding ID: VIPER\n",
            "Add of existing embedding ID: Vireo\n",
            "Add of existing embedding ID: VisCello\n",
            "Add of existing embedding ID: VISION\n",
            "Add of existing embedding ID: visnormsc\n",
            "Add of existing embedding ID: vissE\n",
            "Add of existing embedding ID: VITAE\n",
            "Add of existing embedding ID: Vitessce\n",
            "Add of existing embedding ID: VoxHunt\n",
            "Add of existing embedding ID: VPAC\n",
            "Add of existing embedding ID: Vulture\n",
            "Add of existing embedding ID: waddR\n",
            "Add of existing embedding ID: WASP\n",
            "Add of existing embedding ID: WAT3R\n",
            "Add of existing embedding ID: WaveCrest\n",
            "Add of existing embedding ID: WEDGE\n",
            "Add of existing embedding ID: wf-single-cell\n",
            "Add of existing embedding ID: Wishbone\n",
            "Add of existing embedding ID: wot\n",
            "Add of existing embedding ID: Wx\n",
            "Add of existing embedding ID: XClone\n",
            "Add of existing embedding ID: XenoCell\n",
            "Add of existing embedding ID: ZetaSuite\n",
            "Add of existing embedding ID: ZIAQ\n",
            "Add of existing embedding ID: ZIFA\n",
            "Add of existing embedding ID: ZINB-WaVE\n",
            "Add of existing embedding ID: ZINBAE\n",
            "Add of existing embedding ID: ZINBMM\n",
            "Add of existing embedding ID: zingeR\n",
            "Add of existing embedding ID: zUMIs\n",
            "Insert of existing embedding ID: acorde\n",
            "Insert of existing embedding ID: ACTINN\n",
            "Insert of existing embedding ID: ACTION\n",
            "Insert of existing embedding ID: ACTIONet\n",
            "Insert of existing embedding ID: ACTIVA\n",
            "Insert of existing embedding ID: activeSVC\n",
            "Insert of existing embedding ID: ADC\n",
            "Insert of existing embedding ID: ADClust\n",
            "Insert of existing embedding ID: ADImpute\n",
            "Insert of existing embedding ID: adobo\n",
            "Insert of existing embedding ID: adverSCarial\n",
            "Insert of existing embedding ID: AE-TPGG\n",
            "Insert of existing embedding ID: aggregateBioVar\n",
            "Insert of existing embedding ID: airpart\n",
            "Insert of existing embedding ID: aLDG\n",
            "Insert of existing embedding ID: alevin-fry\n",
            "Insert of existing embedding ID: alevinQC\n",
            "Insert of existing embedding ID: ALRA\n",
            "Insert of existing embedding ID: AltAnalyze\n",
            "Insert of existing embedding ID: AlzScPred\n",
            "Insert of existing embedding ID: AmbientContaminationMetrics\n",
            "Insert of existing embedding ID: anchor\n",
            "Insert of existing embedding ID: ANS\n",
            "Insert of existing embedding ID: anticor_features\n",
            "Insert of existing embedding ID: APL\n",
            "Insert of existing embedding ID: ASAP\n",
            "Insert of existing embedding ID: Asc-Seurat\n",
            "Insert of existing embedding ID: ascend\n",
            "Insert of existing embedding ID: ASTK\n",
            "Insert of existing embedding ID: ASURAT\n",
            "Insert of existing embedding ID: ATAXIC\n",
            "Insert of existing embedding ID: AUCell\n",
            "Insert of existing embedding ID: Augur\n",
            "Insert of existing embedding ID: AutoClass\n",
            "Insert of existing embedding ID: AutoImpute\n",
            "Insert of existing embedding ID: AutomaticCellTypeIdentification\n",
            "Insert of existing embedding ID: AWGAN\n",
            "Insert of existing embedding ID: BABEL\n",
            "Insert of existing embedding ID: BackSPIN\n",
            "Insert of existing embedding ID: badger\n",
            "Insert of existing embedding ID: BALDR\n",
            "Insert of existing embedding ID: BAMMSC\n",
            "Insert of existing embedding ID: BarCounter\n",
            "Insert of existing embedding ID: baredSC\n",
            "Insert of existing embedding ID: BarMixer\n",
            "Insert of existing embedding ID: BASIC\n",
            "Insert of existing embedding ID: BASiCS\n",
            "Insert of existing embedding ID: BASiCStan\n",
            "Insert of existing embedding ID: BasisVAE\n",
            "Insert of existing embedding ID: BatchBench\n",
            "Insert of existing embedding ID: batchelor\n",
            "Insert of existing embedding ID: BATMAN\n",
            "Insert of existing embedding ID: BayesEATS\n",
            "Insert of existing embedding ID: BayesPrism\n",
            "Insert of existing embedding ID: bayNorm\n",
            "Insert of existing embedding ID: bbknn\n",
            "Insert of existing embedding ID: BC_tSNE\n",
            "Insert of existing embedding ID: bcbio\n",
            "Insert of existing embedding ID: bcbioSingleCell\n",
            "Insert of existing embedding ID: BCseq\n",
            "Insert of existing embedding ID: BDA\n",
            "Insert of existing embedding ID: BEANIE\n",
            "Insert of existing embedding ID: BEARscc\n",
            "Insert of existing embedding ID: BEELINE\n",
            "Insert of existing embedding ID: BEENE\n",
            "Insert of existing embedding ID: BEER\n",
            "Insert of existing embedding ID: BENISSE\n",
            "Insert of existing embedding ID: BERMUDA\n",
            "Insert of existing embedding ID: BESCA\n",
            "Insert of existing embedding ID: Beyondcell\n",
            "Insert of existing embedding ID: Bfimpute\n",
            "Insert of existing embedding ID: bigSCale\n",
            "Insert of existing embedding ID: bigSCale2\n",
            "Insert of existing embedding ID: bindSC\n",
            "Insert of existing embedding ID: BingleSeq\n",
            "Insert of existing embedding ID: biolord\n",
            "Insert of existing embedding ID: Biomodelling\n",
            "Insert of existing embedding ID: BioTIP\n",
            "Insert of existing embedding ID: BIRD\n",
            "Insert of existing embedding ID: BISCUIT\n",
            "Insert of existing embedding ID: Bisque\n",
            "Insert of existing embedding ID: BITFAM\n",
            "Insert of existing embedding ID: BLGGM\n",
            "Insert of existing embedding ID: bollito\n",
            "Insert of existing embedding ID: bonvoyage\n",
            "Insert of existing embedding ID: BPSC\n",
            "Insert of existing embedding ID: BraCeR\n",
            "Insert of existing embedding ID: BranchedGP\n",
            "Insert of existing embedding ID: BRAPeS\n",
            "Insert of existing embedding ID: BREM-SC\n",
            "Insert of existing embedding ID: BRIE\n",
            "Insert of existing embedding ID: BSDE\n",
            "Insert of existing embedding ID: bseqsc\n",
            "Insert of existing embedding ID: BTR\n",
            "Insert of existing embedding ID: BUSseq\n",
            "Insert of existing embedding ID: bustools\n",
            "Insert of existing embedding ID: BZINB\n",
            "Insert of existing embedding ID: CAbiNet\n",
            "Insert of existing embedding ID: CACTUS\n",
            "Insert of existing embedding ID: CAISC\n",
            "Insert of existing embedding ID: CALISTA\n",
            "Insert of existing embedding ID: CAMLU\n",
            "Insert of existing embedding ID: CancerInSilico\n",
            "Insert of existing embedding ID: Canek\n",
            "Insert of existing embedding ID: CanSig\n",
            "Insert of existing embedding ID: CAPITAL\n",
            "Insert of existing embedding ID: Capybara\n",
            "Insert of existing embedding ID: CARDAMOM\n",
            "Insert of existing embedding ID: CarDEC\n",
            "Insert of existing embedding ID: cardelino\n",
            "Insert of existing embedding ID: CaSee\n",
            "Insert of existing embedding ID: CASi\n",
            "Insert of existing embedding ID: CaSpER\n",
            "Insert of existing embedding ID: CaSTLe\n",
            "Insert of existing embedding ID: CausalDAG\n",
            "Insert of existing embedding ID: cb_sniffer\n",
            "Insert of existing embedding ID: ccdf\n",
            "Insert of existing embedding ID: ccfindR\n",
            "Insert of existing embedding ID: ccImpute\n",
            "Insert of existing embedding ID: ccNet\n",
            "Insert of existing embedding ID: CCNMF\n",
            "Insert of existing embedding ID: CCPE\n",
            "Insert of existing embedding ID: CCPlotR\n",
            "Insert of existing embedding ID: ccRemover\n",
            "Insert of existing embedding ID: CCSF\n",
            "Insert of existing embedding ID: CCSN\n",
            "Insert of existing embedding ID: CDI\n",
            "Insert of existing embedding ID: CDR-g\n",
            "Insert of existing embedding ID: celaref\n",
            "Insert of existing embedding ID: celda\n",
            "Insert of existing embedding ID: CeLEry\n",
            "Insert of existing embedding ID: Cell-BLAST\n",
            "Insert of existing embedding ID: cell2cell\n",
            "Insert of existing embedding ID: cell2fate\n",
            "Insert of existing embedding ID: cell2location\n",
            "Insert of existing embedding ID: cell2sentence\n",
            "Insert of existing embedding ID: cell2sentence-ft\n",
            "Insert of existing embedding ID: cellAlign\n",
            "Insert of existing embedding ID: CellANOVA\n",
            "Insert of existing embedding ID: Cellar\n",
            "Insert of existing embedding ID: CellaRepertorium\n",
            "Insert of existing embedding ID: cellassign\n",
            "Insert of existing embedding ID: cellatlas\n",
            "Insert of existing embedding ID: CellBench\n",
            "Insert of existing embedding ID: CellBender\n",
            "Insert of existing embedding ID: CellBIC\n",
            "Insert of existing embedding ID: cellBrowser\n",
            "Insert of existing embedding ID: CellCall\n",
            "Insert of existing embedding ID: CellChat\n",
            "Insert of existing embedding ID: CellContrast\n",
            "Insert of existing embedding ID: cellDancer\n",
            "Insert of existing embedding ID: CellDrift\n",
            "Insert of existing embedding ID: cellenium\n",
            "Insert of existing embedding ID: CellexalVR\n",
            "Insert of existing embedding ID: CellFishing\n",
            "Insert of existing embedding ID: CellGEP\n",
            "Insert of existing embedding ID: cellgrid\n",
            "Insert of existing embedding ID: CellHint\n",
            "Insert of existing embedding ID: CellID\n",
            "Insert of existing embedding ID: Cellity\n",
            "Insert of existing embedding ID: CellLayers\n",
            "Insert of existing embedding ID: CellMap\n",
            "Insert of existing embedding ID: CellMeSH\n",
            "Insert of existing embedding ID: CellMixS\n",
            "Insert of existing embedding ID: CellO\n",
            "Insert of existing embedding ID: cellograph\n",
            "Insert of existing embedding ID: CellOracle\n",
            "Insert of existing embedding ID: CellOT\n",
            "Insert of existing embedding ID: CellPath\n",
            "Insert of existing embedding ID: CellPhoneDB\n",
            "Insert of existing embedding ID: CellPLM\n",
            "Insert of existing embedding ID: cellpypes\n",
            "Insert of existing embedding ID: CellRanger\n",
            "Insert of existing embedding ID: CellRank\n",
            "Insert of existing embedding ID: CellRegMap\n",
            "Insert of existing embedding ID: CellRouter\n",
            "Insert of existing embedding ID: CellSIUS\n",
            "Insert of existing embedding ID: cellsnake\n",
            "Insert of existing embedding ID: cellSNP\n",
            "Insert of existing embedding ID: cellsnp-lite\n",
            "Insert of existing embedding ID: CellST\n",
            "Insert of existing embedding ID: CellTagR\n",
            "Insert of existing embedding ID: celltalker\n",
            "Insert of existing embedding ID: CellTools\n",
            "Insert of existing embedding ID: CellTrails\n",
            "Insert of existing embedding ID: CellTree\n",
            "Insert of existing embedding ID: CellTrek\n",
            "Insert of existing embedding ID: CellTypist\n",
            "Insert of existing embedding ID: CellView\n",
            "Insert of existing embedding ID: cellxgene\n",
            "Insert of existing embedding ID: cellxgene_VIP\n",
            "Insert of existing embedding ID: cellXY\n",
            "Insert of existing embedding ID: Census\n",
            "Insert of existing embedding ID: Cepo\n",
            "Insert of existing embedding ID: Cerebro\n",
            "Insert of existing embedding ID: CeSpGRN\n",
            "Insert of existing embedding ID: CeTF\n",
            "Insert of existing embedding ID: cFIT\n",
            "Insert of existing embedding ID: CHETAH\n",
            "Insert of existing embedding ID: Chloris\n",
            "Insert of existing embedding ID: chooseR\n",
            "Insert of existing embedding ID: Chord\n",
            "Insert of existing embedding ID: CIARA\n",
            "Insert of existing embedding ID: CIDER\n",
            "Insert of existing embedding ID: CIDR\n",
            "Insert of existing embedding ID: CIForm\n",
            "Insert of existing embedding ID: CINEMA-OT\n",
            "Insert of existing embedding ID: CIPR\n",
            "Insert of existing embedding ID: Cirrocumulus\n",
            "Insert of existing embedding ID: CiteFuse\n",
            "Insert of existing embedding ID: CITEseqCount\n",
            "Insert of existing embedding ID: Citrus\n",
            "Insert of existing embedding ID: CL-Impute\n",
            "Insert of existing embedding ID: CLEAR\n",
            "Insert of existing embedding ID: clonealign\n",
            "Insert of existing embedding ID: ClonoCluster\n",
            "Insert of existing embedding ID: ClustAssess\n",
            "Insert of existing embedding ID: cluster_diffex\n",
            "Insert of existing embedding ID: ClusterDE\n",
            "Insert of existing embedding ID: clusterExperiment\n",
            "Insert of existing embedding ID: ClusterFoldSimilarity\n",
            "Insert of existing embedding ID: Clustergrammer2\n",
            "Insert of existing embedding ID: ClusterMap\n",
            "Insert of existing embedding ID: ClusterMine\n",
            "Insert of existing embedding ID: clustermole\n",
            "Insert of existing embedding ID: clusterpval\n",
            "Insert of existing embedding ID: clustifyr\n",
            "Insert of existing embedding ID: CMFImpute\n",
            "Insert of existing embedding ID: cna\n",
            "Insert of existing embedding ID: cNMF\n",
            "Insert of existing embedding ID: CNNC\n",
            "Insert of existing embedding ID: COAC\n",
            "Insert of existing embedding ID: Cobolt\n",
            "Insert of existing embedding ID: cocoa\n",
            "Insert of existing embedding ID: CoGAPs\n",
            "Insert of existing embedding ID: cola\n",
            "Insert of existing embedding ID: CombiROC\n",
            "Insert of existing embedding ID: COMET\n",
            "Insert of existing embedding ID: CommPath\n",
            "Insert of existing embedding ID: Compartmap\n",
            "Insert of existing embedding ID: compass\n",
            "Insert of existing embedding ID: CONCLUS\n",
            "Insert of existing embedding ID: condiments\n",
            "Insert of existing embedding ID: CoNGA\n",
            "Insert of existing embedding ID: CONGAS\n",
            "Insert of existing embedding ID: CONGASp\n",
            "Insert of existing embedding ID: CONICS\n",
            "Insert of existing embedding ID: Connectome\n",
            "Insert of existing embedding ID: Conos\n",
            "Insert of existing embedding ID: constclust\n",
            "Insert of existing embedding ID: contrastiveVI\n",
            "Insert of existing embedding ID: COPILOT\n",
            "Insert of existing embedding ID: CopulaGCN\n",
            "Insert of existing embedding ID: CopyKAT\n",
            "Insert of existing embedding ID: corral\n",
            "Insert of existing embedding ID: Corvo\n",
            "Insert of existing embedding ID: COSG\n",
            "Insert of existing embedding ID: COSGR\n",
            "Insert of existing embedding ID: cosin\n",
            "Insert of existing embedding ID: COSLIR\n",
            "Insert of existing embedding ID: CoSpar\n",
            "Insert of existing embedding ID: CosTaL\n",
            "Insert of existing embedding ID: COT\n",
            "Insert of existing embedding ID: COTAN\n",
            "Insert of existing embedding ID: countClust\n",
            "Insert of existing embedding ID: countland\n",
            "Insert of existing embedding ID: countsplit\n",
            "Insert of existing embedding ID: coupleCoC\n",
            "Insert of existing embedding ID: coupleCoC_plus\n",
            "Insert of existing embedding ID: coupledAE-patchseq\n",
            "Insert of existing embedding ID: CPA\n",
            "Insert of existing embedding ID: CPLVM\n",
            "Insert of existing embedding ID: CRMetrics\n",
            "Insert of existing embedding ID: CrossTalkeR\n",
            "Insert of existing embedding ID: CS-CORE\n",
            "Insert of existing embedding ID: CSHMM\n",
            "Insert of existing embedding ID: CSHMM-TF\n",
            "Insert of existing embedding ID: CSMF\n",
            "Insert of existing embedding ID: CSN\n",
            "Insert of existing embedding ID: CSOmap\n",
            "Insert of existing embedding ID: CStreet\n",
            "Insert of existing embedding ID: ctgGEM\n",
            "Insert of existing embedding ID: CTMM\n",
            "Insert of existing embedding ID: CTpathway\n",
            "Insert of existing embedding ID: cTPnet\n",
            "Insert of existing embedding ID: cvCovEst\n",
            "Insert of existing embedding ID: Cyclum\n",
            "Insert of existing embedding ID: cytocipher\n",
            "Insert of existing embedding ID: Cytopath\n",
            "Insert of existing embedding ID: CytoSPACE\n",
            "Insert of existing embedding ID: CytoTalk\n",
            "Insert of existing embedding ID: D-EE\n",
            "Insert of existing embedding ID: D-SPIN\n",
            "Insert of existing embedding ID: D3E\n",
            "Insert of existing embedding ID: DALI\n",
            "Insert of existing embedding ID: DANCE\n",
            "Insert of existing embedding ID: dandelion\n",
            "Insert of existing embedding ID: DAPCA\n",
            "Insert of existing embedding ID: DAZZLE\n",
            "Insert of existing embedding ID: dbcti\n",
            "Insert of existing embedding ID: DC3\n",
            "Insert of existing embedding ID: DCA\n",
            "Insert of existing embedding ID: DCA_R\n",
            "Insert of existing embedding ID: DCATS\n",
            "Insert of existing embedding ID: DCLEAR\n",
            "Insert of existing embedding ID: ddqc\n",
            "Insert of existing embedding ID: ddSeeker\n",
            "Insert of existing embedding ID: de-goulash\n",
            "Insert of existing embedding ID: DECENT\n",
            "Insert of existing embedding ID: DECODE\n",
            "Insert of existing embedding ID: decontX\n",
            "Insert of existing embedding ID: decoupler\n",
            "Insert of existing embedding ID: deCS\n",
            "Insert of existing embedding ID: DEEF\n",
            "Insert of existing embedding ID: DeepCCI\n",
            "Insert of existing embedding ID: DeepCycle\n",
            "Insert of existing embedding ID: DeepDRIM\n",
            "Insert of existing embedding ID: DeepImpute\n",
            "Insert of existing embedding ID: DeepRIG\n",
            "Insert of existing embedding ID: DeepSEM\n",
            "Insert of existing embedding ID: DeepSort\n",
            "Insert of existing embedding ID: DeepVelo\n",
            "Insert of existing embedding ID: DEGAS\n",
            "Insert of existing embedding ID: DElegate\n",
            "Insert of existing embedding ID: DeLorean\n",
            "Insert of existing embedding ID: DELVE\n",
            "Insert of existing embedding ID: deMULTIplex2\n",
            "Insert of existing embedding ID: Demuxafy\n",
            "Insert of existing embedding ID: Demuxalot\n",
            "Insert of existing embedding ID: demuxlet\n",
            "Insert of existing embedding ID: demuxmix\n",
            "Insert of existing embedding ID: DEN\n",
            "Insert of existing embedding ID: DENDRO\n",
            "Insert of existing embedding ID: DendroSplit\n",
            "Insert of existing embedding ID: densityCut\n",
            "Insert of existing embedding ID: DensityPath\n",
            "Insert of existing embedding ID: densvis\n",
            "Insert of existing embedding ID: densVis\n",
            "Insert of existing embedding ID: DepecheR\n",
            "Insert of existing embedding ID: DESC\n",
            "Insert of existing embedding ID: DESCEND\n",
            "Insert of existing embedding ID: DEsingle\n",
            "Insert of existing embedding ID: destiny\n",
            "Insert of existing embedding ID: DevKidCC\n",
            "Insert of existing embedding ID: DEWAKSS\n",
            "Insert of existing embedding ID: Dhaka\n",
            "Insert of existing embedding ID: DIALOGUE\n",
            "Insert of existing embedding ID: DICE\n",
            "Insert of existing embedding ID: Dictys\n",
            "Insert of existing embedding ID: DIEM\n",
            "Insert of existing embedding ID: DifferentialRegulation\n",
            "Insert of existing embedding ID: DigitalCellSorter\n",
            "Insert of existing embedding ID: DIMMSC\n",
            "Insert of existing embedding ID: dincta\n",
            "Insert of existing embedding ID: Dino\n",
            "Insert of existing embedding ID: DIRECT-NET\n",
            "Insert of existing embedding ID: DISC\n",
            "Insert of existing embedding ID: DIscBIO\n",
            "Insert of existing embedding ID: DISCERN\n",
            "Insert of existing embedding ID: DISCO\n",
            "Insert of existing embedding ID: DiSiR\n",
            "Insert of existing embedding ID: distinct\n",
            "Insert of existing embedding ID: DistMap\n",
            "Insert of existing embedding ID: dittoSeq\n",
            "Insert of existing embedding ID: dmatch\n",
            "Insert of existing embedding ID: Domino\n",
            "Insert of existing embedding ID: DoubletCollection\n",
            "Insert of existing embedding ID: DoubletDecon\n",
            "Insert of existing embedding ID: DoubletDetection\n",
            "Insert of existing embedding ID: DoubletFinder\n",
            "Insert of existing embedding ID: Dozer\n",
            "Insert of existing embedding ID: dpi\n",
            "Insert of existing embedding ID: DPre\n",
            "Insert of existing embedding ID: DPT\n",
            "Insert of existing embedding ID: DRA\n",
            "Insert of existing embedding ID: dragonsc\n",
            "Insert of existing embedding ID: DREAM\n",
            "Insert of existing embedding ID: dreamlet\n",
            "Insert of existing embedding ID: DREEP\n",
            "Insert of existing embedding ID: DrImpute\n",
            "Insert of existing embedding ID: DrivAER\n",
            "Insert of existing embedding ID: dropbead\n",
            "Insert of existing embedding ID: dropClust\n",
            "Insert of existing embedding ID: dropEst\n",
            "Insert of existing embedding ID: dropkick\n",
            "Insert of existing embedding ID: DropLasso\n",
            "Insert of existing embedding ID: DropletQC\n",
            "Insert of existing embedding ID: DropletUtils\n",
            "Insert of existing embedding ID: dropSeqPipe\n",
            "Insert of existing embedding ID: dropsim\n",
            "Insert of existing embedding ID: DrSeq2\n",
            "Insert of existing embedding ID: DSAVE-M\n",
            "Insert of existing embedding ID: DSAVE-R\n",
            "Insert of existing embedding ID: dsb\n",
            "Insert of existing embedding ID: dsCellNet\n",
            "Insert of existing embedding ID: DSINMF\n",
            "Insert of existing embedding ID: DTAE\n",
            "Insert of existing embedding ID: DTFLOW\n",
            "Insert of existing embedding ID: DTUrtle\n",
            "Insert of existing embedding ID: DTWScore\n",
            "Insert of existing embedding ID: DUBStepR\n",
            "Insert of existing embedding ID: Dune\n",
            "Insert of existing embedding ID: DURIAN\n",
            "Insert of existing embedding ID: DUSC\n",
            "Insert of existing embedding ID: Dynamo\n",
            "Insert of existing embedding ID: dynDeepDRIM\n",
            "Insert of existing embedding ID: dyngen\n",
            "Insert of existing embedding ID: DynGFN\n",
            "Insert of existing embedding ID: dynverse\n",
            "Insert of existing embedding ID: EasyCellType\n",
            "Insert of existing embedding ID: ECLAIR\n",
            "Insert of existing embedding ID: EDClust\n",
            "Insert of existing embedding ID: EDGE\n",
            "Insert of existing embedding ID: EEISP\n",
            "Insert of existing embedding ID: elasticC3\n",
            "Insert of existing embedding ID: ELeFHAnt\n",
            "Insert of existing embedding ID: ELSA\n",
            "Insert of existing embedding ID: ELVAR\n",
            "Insert of existing embedding ID: embeddr\n",
            "Insert of existing embedding ID: EMBEDR\n",
            "Insert of existing embedding ID: EMEP\n",
            "Insert of existing embedding ID: EmptyNN\n",
            "Insert of existing embedding ID: enclone\n",
            "Insert of existing embedding ID: ENCORE\n",
            "Insert of existing embedding ID: ENHANCE\n",
            "Insert of existing embedding ID: EnImpute\n",
            "Insert of existing embedding ID: EnsembleKQC\n",
            "Insert of existing embedding ID: EnTSSR\n",
            "Insert of existing embedding ID: Eoulsan\n",
            "Insert of existing embedding ID: Epi-Impute\n",
            "Insert of existing embedding ID: EPIC\n",
            "Insert of existing embedding ID: eQTLsingle\n",
            "Insert of existing embedding ID: ESAT\n",
            "Insert of existing embedding ID: escape\n",
            "Insert of existing embedding ID: ESCO\n",
            "Insert of existing embedding ID: eSPRESSO\n",
            "Insert of existing embedding ID: eSVD\n",
            "Insert of existing embedding ID: EVI\n",
            "Insert of existing embedding ID: exceiver\n",
            "Insert of existing embedding ID: exFINDER\n",
            "Insert of existing embedding ID: expressyouRcell\n",
            "Insert of existing embedding ID: Falco\n",
            "Insert of existing embedding ID: FastCount\n",
            "Insert of existing embedding ID: fasthplus\n",
            "Insert of existing embedding ID: FastIntegration\n",
            "Insert of existing embedding ID: FastProject\n",
            "Insert of existing embedding ID: FastRNA\n",
            "Insert of existing embedding ID: fastTopics\n",
            "Insert of existing embedding ID: FateCompass\n",
            "Insert of existing embedding ID: FateID\n",
            "Insert of existing embedding ID: FAVA\n",
            "Insert of existing embedding ID: fba\n",
            "Insert of existing embedding ID: fcoex\n",
            "Insert of existing embedding ID: FEAST\n",
            "Insert of existing embedding ID: FEATS\n",
            "Insert of existing embedding ID: FeatureCorr\n",
            "Insert of existing embedding ID: FEM\n",
            "Insert of existing embedding ID: FIESTA\n",
            "Insert of existing embedding ID: FiRE\n",
            "Insert of existing embedding ID: FIRM\n",
            "Insert of existing embedding ID: fishpond\n",
            "Insert of existing embedding ID: FitDevo\n",
            "Insert of existing embedding ID: FLAMES\n",
            "Insert of existing embedding ID: Flexiplex\n",
            "Insert of existing embedding ID: flotilla\n",
            "Insert of existing embedding ID: FlowGrid\n",
            "Insert of existing embedding ID: FLOWMAPR\n",
            "Insert of existing embedding ID: forest-fire-clustering\n",
            "Insert of existing embedding ID: FORKS\n",
            "Insert of existing embedding ID: Fotomics\n",
            "Insert of existing embedding ID: FRmatch\n",
            "Insert of existing embedding ID: FSCAM\n",
            "Insert of existing embedding ID: Fugue\n",
            "Insert of existing embedding ID: FUNKI\n",
            "Insert of existing embedding ID: fusca\n",
            "Insert of existing embedding ID: G2S3\n",
            "Insert of existing embedding ID: GapClust\n",
            "Insert of existing embedding ID: Garnett\n",
            "Insert of existing embedding ID: gCAnno\n",
            "Insert of existing embedding ID: GEARS\n",
            "Insert of existing embedding ID: GEDI\n",
            "Insert of existing embedding ID: geneBasisR\n",
            "Insert of existing embedding ID: GeneClust\n",
            "Insert of existing embedding ID: GenePattern\n",
            "Insert of existing embedding ID: Genes2Genes\n",
            "Insert of existing embedding ID: genesorteR\n",
            "Insert of existing embedding ID: GeneSwitches\n",
            "Insert of existing embedding ID: GeneVector\n",
            "Insert of existing embedding ID: GenKI\n",
            "Insert of existing embedding ID: genomicInstability\n",
            "Insert of existing embedding ID: geosketch\n",
            "Insert of existing embedding ID: gExcite\n",
            "Insert of existing embedding ID: gficf\n",
            "Insert of existing embedding ID: GFPA\n",
            "Insert of existing embedding ID: ggsc\n",
            "Insert of existing embedding ID: GIANT\n",
            "Insert of existing embedding ID: GiniClust\n",
            "Insert of existing embedding ID: GiniClust2\n",
            "Insert of existing embedding ID: GiniClust3\n",
            "Insert of existing embedding ID: GiRAFR\n",
            "Insert of existing embedding ID: GLM-PCA\n",
            "Insert of existing embedding ID: glmGamPoi\n",
            "Insert of existing embedding ID: GloScope\n",
            "Insert of existing embedding ID: GLUE\n",
            "Insert of existing embedding ID: GLUER\n",
            "Insert of existing embedding ID: GMM-Demux\n",
            "Insert of existing embedding ID: GNNImpute\n",
            "Insert of existing embedding ID: GPcounts\n",
            "Insert of existing embedding ID: GPfates\n",
            "Insert of existing embedding ID: GPseudoClust\n",
            "Insert of existing embedding ID: GPseudoRank\n",
            "Insert of existing embedding ID: Granatum\n",
            "Insert of existing embedding ID: GrandPrix\n",
            "Insert of existing embedding ID: Graph-Dynamo\n",
            "Insert of existing embedding ID: graph-sc\n",
            "Insert of existing embedding ID: GraphCS\n",
            "Insert of existing embedding ID: GraphEmbed\n",
            "Insert of existing embedding ID: GraphSCC\n",
            "Insert of existing embedding ID: GraphSCI\n",
            "Insert of existing embedding ID: GrID-Net\n",
            "Insert of existing embedding ID: GRISLI\n",
            "Insert of existing embedding ID: GRM\n",
            "Insert of existing embedding ID: GRN-VAE\n",
            "Insert of existing embedding ID: GRNUlar\n",
            "Insert of existing embedding ID: GRNVBEM\n",
            "Insert of existing embedding ID: GRouNdGAN\n",
            "Insert of existing embedding ID: GRUFFI\n",
            "Insert of existing embedding ID: gsdensity\n",
            "Insert of existing embedding ID: GSEApy\n",
            "Insert of existing embedding ID: GSFA\n",
            "Insert of existing embedding ID: gssnng\n",
            "Insert of existing embedding ID: h-tSNE\n",
            "Insert of existing embedding ID: hadge\n",
            "Insert of existing embedding ID: Haisu\n",
            "Insert of existing embedding ID: HAL\n",
            "Insert of existing embedding ID: HARISSA\n",
            "Insert of existing embedding ID: Harmony\n",
            "Insert of existing embedding ID: hbfm\n",
            "Insert of existing embedding ID: HD-AE\n",
            "Insert of existing embedding ID: HDMC\n",
            "Insert of existing embedding ID: HGC\n",
            "Insert of existing embedding ID: HiDeF\n",
            "Insert of existing embedding ID: hierarchicell\n",
            "Insert of existing embedding ID: HieRFIT\n",
            "Insert of existing embedding ID: HIPPO\n",
            "Insert of existing embedding ID: HocusPocus\n",
            "Insert of existing embedding ID: HoneyBADGER\n",
            "Insert of existing embedding ID: HopLand\n",
            "Insert of existing embedding ID: Hopper\n",
            "Insert of existing embedding ID: Hotspot\n",
            "Insert of existing embedding ID: HPL\n",
            "Insert of existing embedding ID: hscScore\n",
            "Insert of existing embedding ID: hsslda\n",
            "Insert of existing embedding ID: HTOreader\n",
            "Insert of existing embedding ID: HurdleNormal\n",
            "Insert of existing embedding ID: hypercluster\n",
            "Insert of existing embedding ID: I-Impute\n",
            "Insert of existing embedding ID: i2dash\n",
            "Insert of existing embedding ID: iasva\n",
            "Insert of existing embedding ID: Ibex\n",
            "Insert of existing embedding ID: IBMR\n",
            "Insert of existing embedding ID: IBRAP\n",
            "Insert of existing embedding ID: iBRIDGE\n",
            "Insert of existing embedding ID: ICAnet\n",
            "Insert of existing embedding ID: ICARUS\n",
            "Insert of existing embedding ID: ICAT\n",
            "Insert of existing embedding ID: ICELLNET\n",
            "Insert of existing embedding ID: iCellR\n",
            "Insert of existing embedding ID: iDEA\n",
            "Insert of existing embedding ID: IDEAS\n",
            "Insert of existing embedding ID: IEntropy\n",
            "Insert of existing embedding ID: IKAP\n",
            "Insert of existing embedding ID: ikarus\n",
            "Insert of existing embedding ID: Ikura\n",
            "Insert of existing embedding ID: ILoReg\n",
            "Insert of existing embedding ID: iMAP\n",
            "Insert of existing embedding ID: ImmClassifier\n",
            "Insert of existing embedding ID: immunarch\n",
            "Insert of existing embedding ID: ImSpectR\n",
            "Insert of existing embedding ID: IndepthPathway\n",
            "Insert of existing embedding ID: indeXplorer\n",
            "Insert of existing embedding ID: inferCNV\n",
            "Insert of existing embedding ID: inferCNVpy\n",
            "Insert of existing embedding ID: inferCSN\n",
            "Insert of existing embedding ID: Inferelator\n",
            "Insert of existing embedding ID: infohet\n",
            "Insert of existing embedding ID: InGene\n",
            "Insert of existing embedding ID: INSCT\n",
            "Insert of existing embedding ID: INSISTC\n",
            "Insert of existing embedding ID: IntegratedLearner\n",
            "Insert of existing embedding ID: InterCellar\n",
            "Insert of existing embedding ID: InterCellDB\n",
            "Insert of existing embedding ID: iQcell\n",
            "Insert of existing embedding ID: IReNA\n",
            "Insert of existing embedding ID: IRIS-FGM\n",
            "Insert of existing embedding ID: IronThrone-GoT\n",
            "Insert of existing embedding ID: iS-CellR\n",
            "Insert of existing embedding ID: iscandar\n",
            "Insert of existing embedding ID: iSEE\n",
            "Insert of existing embedding ID: iSMNN\n",
            "Insert of existing embedding ID: ISnorm\n",
            "Insert of existing embedding ID: IsoCell\n",
            "Insert of existing embedding ID: ISOP\n",
            "Insert of existing embedding ID: iTALK\n",
            "Insert of existing embedding ID: ItClust\n",
            "Insert of existing embedding ID: iterClust\n",
            "Insert of existing embedding ID: ivis\n",
            "Insert of existing embedding ID: jackstraw\n",
            "Insert of existing embedding ID: JAMIE\n",
            "Insert of existing embedding ID: JGNsc\n",
            "Insert of existing embedding ID: JIND\n",
            "Insert of existing embedding ID: JOINT\n",
            "Insert of existing embedding ID: JSNMF_py\n",
            "Insert of existing embedding ID: JSOM\n",
            "Insert of existing embedding ID: jSRC\n",
            "Insert of existing embedding ID: Jvis\n",
            "Insert of existing embedding ID: k-branches\n",
            "Insert of existing embedding ID: K2Taxonomer\n",
            "Insert of existing embedding ID: kallisto\n",
            "Insert of existing embedding ID: kana\n",
            "Insert of existing embedding ID: kBET\n",
            "Insert of existing embedding ID: KMD\n",
            "Insert of existing embedding ID: kmermaid\n",
            "Insert of existing embedding ID: knn_smooth\n",
            "Insert of existing embedding ID: KPNN\n",
            "Insert of existing embedding ID: ktest\n",
            "Insert of existing embedding ID: LAK\n",
            "Insert of existing embedding ID: LAmbDA\n",
            "Insert of existing embedding ID: Lamian\n",
            "Insert of existing embedding ID: LandSCENT\n",
            "Insert of existing embedding ID: LANTSA\n",
            "Insert of existing embedding ID: LATE\n",
            "Insert of existing embedding ID: LatentVelo\n",
            "Insert of existing embedding ID: LDA_scRNAseq\n",
            "Insert of existing embedding ID: leader\n",
            "Insert of existing embedding ID: LEAP\n",
            "Insert of existing embedding ID: learn2count\n",
            "Insert of existing embedding ID: lemur\n",
            "Insert of existing embedding ID: LIANA\n",
            "Insert of existing embedding ID: Libra\n",
            "Insert of existing embedding ID: LIGER\n",
            "Insert of existing embedding ID: LINEAGE\n",
            "Insert of existing embedding ID: LineagePulse\n",
            "Insert of existing embedding ID: Linnorm\n",
            "Insert of existing embedding ID: LinQView\n",
            "Insert of existing embedding ID: LinRace\n",
            "Insert of existing embedding ID: locaTE\n",
            "Insert of existing embedding ID: locCSN\n",
            "Insert of existing embedding ID: Longcell\n",
            "Insert of existing embedding ID: LRLoop\n",
            "Insert of existing embedding ID: LRT\n",
            "Insert of existing embedding ID: LTMGSCA\n",
            "Insert of existing embedding ID: M3Drop\n",
            "Insert of existing embedding ID: M3S\n",
            "Insert of existing embedding ID: MAAPER\n",
            "Insert of existing embedding ID: MACA\n",
            "Insert of existing embedding ID: MaCroDNA\n",
            "Insert of existing embedding ID: MacroPolo\n",
            "Insert of existing embedding ID: MAESTRO\n",
            "Insert of existing embedding ID: MAGAN\n",
            "Insert of existing embedding ID: MAGIC\n",
            "Insert of existing embedding ID: MAPiT\n",
            "Insert of existing embedding ID: MARGARET\n",
            "Insert of existing embedding ID: MarkerCount\n",
            "Insert of existing embedding ID: MarkerMap\n",
            "Insert of existing embedding ID: MarkovHC\n",
            "Insert of existing embedding ID: MARS\n",
            "Insert of existing embedding ID: MASI\n",
            "Insert of existing embedding ID: MAST\n",
            "Insert of existing embedding ID: MAT2\n",
            "Insert of existing embedding ID: MatchCLOT\n",
            "Insert of existing embedding ID: MATCHER\n",
            "Insert of existing embedding ID: matchSCore\n",
            "Insert of existing embedding ID: matchSCore2\n",
            "Insert of existing embedding ID: Matilda\n",
            "Insert of existing embedding ID: MATTE\n",
            "Insert of existing embedding ID: MaxFuse\n",
            "Insert of existing embedding ID: mbkmeans\n",
            "Insert of existing embedding ID: mcap\n",
            "Insert of existing embedding ID: McImpute\n",
            "Insert of existing embedding ID: MCXpress\n",
            "Insert of existing embedding ID: MEBOCOST\n",
            "Insert of existing embedding ID: MEDALT\n",
            "Insert of existing embedding ID: MELD\n",
            "Insert of existing embedding ID: MELT\n",
            "Insert of existing embedding ID: memento\n",
            "Insert of existing embedding ID: MERLoT\n",
            "Insert of existing embedding ID: MetaCell\n",
            "Insert of existing embedding ID: metacells\n",
            "Insert of existing embedding ID: METAFlux\n",
            "Insert of existing embedding ID: MetaNeighbor\n",
            "Insert of existing embedding ID: MetaTiME\n",
            "Insert of existing embedding ID: MFA\n",
            "Insert of existing embedding ID: MGFR\n",
            "Insert of existing embedding ID: MicroCellClust\n",
            "Insert of existing embedding ID: MICTI\n",
            "Insert of existing embedding ID: Millefy\n",
            "Insert of existing embedding ID: miloR\n",
            "Insert of existing embedding ID: MIMOSCA\n",
            "Insert of existing embedding ID: MIND\n",
            "Insert of existing embedding ID: minicore\n",
            "Insert of existing embedding ID: minnow\n",
            "Insert of existing embedding ID: miQC\n",
            "Insert of existing embedding ID: MIRA\n",
            "Insert of existing embedding ID: miRSCAPE\n",
            "Insert of existing embedding ID: Miscell\n",
            "Insert of existing embedding ID: mitch\n",
            "Insert of existing embedding ID: MitoHEAR\n",
            "Insert of existing embedding ID: mitoSplitter\n",
            "Insert of existing embedding ID: MitoTrace\n",
            "Insert of existing embedding ID: MiXCR\n",
            "Insert of existing embedding ID: MLG\n",
            "Insert of existing embedding ID: mmutil\n",
            "Insert of existing embedding ID: mnnpy\n",
            "Insert of existing embedding ID: Moana\n",
            "Insert of existing embedding ID: MOFA\n",
            "Insert of existing embedding ID: MOFA2\n",
            "Insert of existing embedding ID: Monet\n",
            "Insert of existing embedding ID: Monocle\n",
            "Insert of existing embedding ID: Monod\n",
            "Insert of existing embedding ID: Monopogen\n",
            "Insert of existing embedding ID: mosaicMPI\n",
            "Insert of existing embedding ID: MOSCATO\n",
            "Insert of existing embedding ID: moscot\n",
            "Insert of existing embedding ID: Mowgli\n",
            "Insert of existing embedding ID: Mpath\n",
            "Insert of existing embedding ID: mpCMF\n",
            "Insert of existing embedding ID: MPSSC\n",
            "Insert of existing embedding ID: MQuad\n",
            "Insert of existing embedding ID: MRFscRNAseq\n",
            "Insert of existing embedding ID: mrtree\n",
            "Insert of existing embedding ID: MSIsensor-RNA\n",
            "Insert of existing embedding ID: mtANN\n",
            "Insert of existing embedding ID: MTD\n",
            "Insert of existing embedding ID: MTGO-SC\n",
            "Insert of existing embedding ID: mtSC\n",
            "Insert of existing embedding ID: MUDAN\n",
            "Insert of existing embedding ID: MuDCoD\n",
            "Insert of existing embedding ID: multiGroupVI\n",
            "Insert of existing embedding ID: MultiK\n",
            "Insert of existing embedding ID: MultiMAP\n",
            "Insert of existing embedding ID: multinichenetr\n",
            "Insert of existing embedding ID: Multiscale_PHATE\n",
            "Insert of existing embedding ID: MultiVelo\n",
            "Insert of existing embedding ID: MUREN\n",
            "Insert of existing embedding ID: MURP\n",
            "Insert of existing embedding ID: muscat\n",
            "Insert of existing embedding ID: MuSiC\n",
            "Insert of existing embedding ID: MUSIC\n",
            "Insert of existing embedding ID: MuTrans\n",
            "Insert of existing embedding ID: mvTCR\n",
            "Insert of existing embedding ID: N-ACT\n",
            "Insert of existing embedding ID: NaiveDE\n",
            "Insert of existing embedding ID: NanoPARE\n",
            "Insert of existing embedding ID: NATMI\n",
            "Insert of existing embedding ID: NBID\n",
            "Insert of existing embedding ID: NEBULA\n",
            "Insert of existing embedding ID: Nebulosa\n",
            "Insert of existing embedding ID: Neighbor-seq\n",
            "Insert of existing embedding ID: NetImpute\n",
            "Insert of existing embedding ID: netNMF-sc\n",
            "Insert of existing embedding ID: netSmooth\n",
            "Insert of existing embedding ID: netsne\n",
            "Insert of existing embedding ID: NetworkInference\n",
            "Insert of existing embedding ID: NeuCA\n",
            "Insert of existing embedding ID: NEUROeSTIMator\n",
            "Insert of existing embedding ID: NeuronChat\n",
            "Insert of existing embedding ID: NewWave\n",
            "Insert of existing embedding ID: nichenetr\n",
            "Insert of existing embedding ID: NICHES\n",
            "Insert of existing embedding ID: NIFA\n",
            "Insert of existing embedding ID: nimfa\n",
            "Insert of existing embedding ID: NLBayes\n",
            "Insert of existing embedding ID: NOMAD\n",
            "Insert of existing embedding ID: NOMAD2\n",
            "Insert of existing embedding ID: Normalisr\n",
            "Insert of existing embedding ID: northstar\n",
            "Insert of existing embedding ID: novoSpaRc\n",
            "Insert of existing embedding ID: NSForest\n",
            "Insert of existing embedding ID: OCAT\n",
            "Insert of existing embedding ID: Ocelli\n",
            "Insert of existing embedding ID: ODEGRfinder\n",
            "Insert of existing embedding ID: OEFinder\n",
            "Insert of existing embedding ID: OGFSC\n",
            "Insert of existing embedding ID: OmicInt\n",
            "Insert of existing embedding ID: OnClass\n",
            "Insert of existing embedding ID: OnlinePCA\n",
            "Insert of existing embedding ID: OpenBioMed\n",
            "Insert of existing embedding ID: OpenVisuMap\n",
            "Insert of existing embedding ID: ORIGINS\n",
            "Insert of existing embedding ID: ORIGINS2\n",
            "Insert of existing embedding ID: OscoNet\n",
            "Insert of existing embedding ID: Oscope\n",
            "Insert of existing embedding ID: Ouija\n",
            "Insert of existing embedding ID: ouijaflow\n",
            "Insert of existing embedding ID: Outrigger\n",
            "Insert of existing embedding ID: PACSI\n",
            "Insert of existing embedding ID: pagoda2\n",
            "Insert of existing embedding ID: Palantir\n",
            "Insert of existing embedding ID: PALMO\n",
            "Insert of existing embedding ID: Palo\n",
            "Insert of existing embedding ID: Pamona\n",
            "Insert of existing embedding ID: Pando\n",
            "Insert of existing embedding ID: Panpipes\n",
            "Insert of existing embedding ID: ParaDPMM\n",
            "Insert of existing embedding ID: PARC\n",
            "Insert of existing embedding ID: PAUSE\n",
            "Insert of existing embedding ID: PBA\n",
            "Insert of existing embedding ID: PBLR\n",
            "Insert of existing embedding ID: PCAone\n",
            "Insert of existing embedding ID: pcaReduce\n",
            "Insert of existing embedding ID: pCMF\n",
            "Insert of existing embedding ID: pcqc\n",
            "Insert of existing embedding ID: pCreode\n",
            "Insert of existing embedding ID: peco\n",
            "Insert of existing embedding ID: Pegasus\n",
            "Insert of existing embedding ID: PENCIL\n",
            "Insert of existing embedding ID: PercentMaxDiff\n",
            "Insert of existing embedding ID: PHATE\n",
            "Insert of existing embedding ID: PhEMD\n",
            "Insert of existing embedding ID: PhenoPath\n",
            "Insert of existing embedding ID: phenotype-cover\n",
            "Insert of existing embedding ID: phiclust\n",
            "Insert of existing embedding ID: PhitestR\n",
            "Insert of existing embedding ID: PhylEx\n",
            "Insert of existing embedding ID: phyloRNA\n",
            "Insert of existing embedding ID: PhyloVelo\n",
            "Insert of existing embedding ID: Piccolo\n",
            "Insert of existing embedding ID: PieParty\n",
            "Insert of existing embedding ID: PILOT\n",
            "Insert of existing embedding ID: PIMseq\n",
            "Insert of existing embedding ID: pipeComp\n",
            "Insert of existing embedding ID: PISCES\n",
            "Insert of existing embedding ID: PIVOT\n",
            "Insert of existing embedding ID: pLDA\n",
            "Insert of existing embedding ID: pmVAE\n",
            "Insert of existing embedding ID: PoissonUMIs\n",
            "Insert of existing embedding ID: Polaratio\n",
            "Insert of existing embedding ID: Pollock\n",
            "Insert of existing embedding ID: polyApipe\n",
            "Insert of existing embedding ID: PopAlign\n",
            "Insert of existing embedding ID: popsicleR\n",
            "Insert of existing embedding ID: PopV\n",
            "Insert of existing embedding ID: Portal\n",
            "Insert of existing embedding ID: powerEQTL\n",
            "Insert of existing embedding ID: POWSC\n",
            "Insert of existing embedding ID: powsimR\n",
            "Insert of existing embedding ID: PPMS\n",
            "Insert of existing embedding ID: Pre-Phaser\n",
            "Insert of existing embedding ID: PreCanCell\n",
            "Insert of existing embedding ID: PRESCIENT\n",
            "Insert of existing embedding ID: ProjecTILs\n",
            "Insert of existing embedding ID: ProjectSVR\n",
            "Insert of existing embedding ID: PROSSTT\n",
            "Insert of existing embedding ID: protaccel\n",
            "Insert of existing embedding ID: Pseudocell_Tracer\n",
            "Insert of existing embedding ID: pseudodynamics\n",
            "Insert of existing embedding ID: pseudoga\n",
            "Insert of existing embedding ID: pseudogp\n",
            "Insert of existing embedding ID: PseudotimeDE\n",
            "Insert of existing embedding ID: PsiNorm\n",
            "Insert of existing embedding ID: Psix\n",
            "Insert of existing embedding ID: psupertime\n",
            "Insert of existing embedding ID: pyHSICLasso\n",
            "Insert of existing embedding ID: PyMINEr\n",
            "Insert of existing embedding ID: pyNVR\n",
            "Insert of existing embedding ID: Pyro-Velocity\n",
            "Insert of existing embedding ID: PySCUBA\n",
            "Insert of existing embedding ID: qSNE\n",
            "Insert of existing embedding ID: quasildr\n",
            "Insert of existing embedding ID: QUBIC2\n",
            "Insert of existing embedding ID: RaceID\n",
            "Insert of existing embedding ID: RAFSIL\n",
            "Insert of existing embedding ID: RainDrop\n",
            "Insert of existing embedding ID: Randomly\n",
            "Insert of existing embedding ID: RankCorr\n",
            "Insert of existing embedding ID: RayleighSelection\n",
            "Insert of existing embedding ID: RCA\n",
            "Insert of existing embedding ID: rCASC\n",
            "Insert of existing embedding ID: RCAv2\n",
            "Insert of existing embedding ID: RCI-NB\n",
            "Insert of existing embedding ID: RcppML\n",
            "Insert of existing embedding ID: RCSL\n",
            "Insert of existing embedding ID: ReadZS\n",
            "Insert of existing embedding ID: REBET\n",
            "Insert of existing embedding ID: reCAT\n",
            "Insert of existing embedding ID: RECODE\n",
            "Insert of existing embedding ID: RECOMBINE\n",
            "Insert of existing embedding ID: RecursiveConsensusClustering\n",
            "Insert of existing embedding ID: Red_Panda\n",
            "Insert of existing embedding ID: redPATH\n",
            "Insert of existing embedding ID: Regionvelocity\n",
            "Insert of existing embedding ID: RESCUE\n",
            "Insert of existing embedding ID: ResPAN\n",
            "Insert of existing embedding ID: Revelio\n",
            "Insert of existing embedding ID: RgCop\n",
            "Insert of existing embedding ID: RGGC\n",
            "Insert of existing embedding ID: RISC\n",
            "Insert of existing embedding ID: RNA-Bloom\n",
            "Insert of existing embedding ID: RNA-combine\n",
            "Insert of existing embedding ID: RNA-Sieve\n",
            "Insert of existing embedding ID: RNA-strain-match\n",
            "Insert of existing embedding ID: RNAForecaster\n",
            "Insert of existing embedding ID: RobustCCC\n",
            "Insert of existing embedding ID: robustSingleCell\n",
            "Insert of existing embedding ID: ROGUE\n",
            "Insert of existing embedding ID: ROSeq\n",
            "Insert of existing embedding ID: ROSeq-R\n",
            "Insert of existing embedding ID: rpca\n",
            "Insert of existing embedding ID: RSCORE\n",
            "Insert of existing embedding ID: Rsubread\n",
            "Insert of existing embedding ID: rSuperCT\n",
            "Insert of existing embedding ID: rusty_axe\n",
            "Insert of existing embedding ID: ruvIIInb\n",
            "Insert of existing embedding ID: RVAgene\n",
            "Insert of existing embedding ID: RZiMM\n",
            "Insert of existing embedding ID: SAFE\n",
            "Insert of existing embedding ID: SageNet\n",
            "Insert of existing embedding ID: SAIC\n",
            "Insert of existing embedding ID: sake\n",
            "Insert of existing embedding ID: salmon\n",
            "Insert of existing embedding ID: SAM\n",
            "Insert of existing embedding ID: SAMap\n",
            "Insert of existing embedding ID: SampleQC\n",
            "Insert of existing embedding ID: SAMstrt\n",
            "Insert of existing embedding ID: Sanity\n",
            "Insert of existing embedding ID: Sargent\n",
            "Insert of existing embedding ID: SASCRiP\n",
            "Insert of existing embedding ID: satuRn\n",
            "Insert of existing embedding ID: SATURN\n",
            "Insert of existing embedding ID: SAUCIE\n",
            "Insert of existing embedding ID: saVAE\n",
            "Insert of existing embedding ID: SAVER\n",
            "Insert of existing embedding ID: SAVER-X\n",
            "Insert of existing embedding ID: SBDN\n",
            "Insert of existing embedding ID: sc-compReg\n",
            "Insert of existing embedding ID: sc-jnmf\n",
            "Insert of existing embedding ID: sc-Libra\n",
            "Insert of existing embedding ID: sc-REnF\n",
            "Insert of existing embedding ID: sc2marker\n",
            "Insert of existing embedding ID: sc2MeNetDrug\n",
            "Insert of existing embedding ID: SC2P\n",
            "Insert of existing embedding ID: SC2Spa\n",
            "Insert of existing embedding ID: SC3\n",
            "Insert of existing embedding ID: SC3s\n",
            "Insert of existing embedding ID: SCA\n",
            "Insert of existing embedding ID: scAAnet\n",
            "Insert of existing embedding ID: scAce\n",
            "Insert of existing embedding ID: scACE\n",
            "Insert of existing embedding ID: scAdapt\n",
            "Insert of existing embedding ID: scAEGAN\n",
            "Insert of existing embedding ID: scAEspy\n",
            "Insert of existing embedding ID: SCAFE\n",
            "Insert of existing embedding ID: scaffold\n",
            "Insert of existing embedding ID: scAI\n",
            "Insert of existing embedding ID: scAIDE\n",
            "Insert of existing embedding ID: scAL\n",
            "Insert of existing embedding ID: SCALE\n",
            "Insert of existing embedding ID: SCALEX\n",
            "Insert of existing embedding ID: scAlign\n",
            "Insert of existing embedding ID: scAllele\n",
            "Insert of existing embedding ID: scAlleleCount\n",
            "Insert of existing embedding ID: Scallop2\n",
            "Insert of existing embedding ID: scAMACE\n",
            "Insert of existing embedding ID: scAMACE_py\n",
            "Insert of existing embedding ID: scAmpi\n",
            "Insert of existing embedding ID: scAN10\n",
            "Insert of existing embedding ID: scanalysis\n",
            "Insert of existing embedding ID: scAnCluster\n",
            "Insert of existing embedding ID: SCANet\n",
            "Insert of existing embedding ID: scANNA\n",
            "Insert of existing embedding ID: scAnno\n",
            "Insert of existing embedding ID: scAnnotatR\n",
            "Insert of existing embedding ID: scanorama\n",
            "Insert of existing embedding ID: Scanpro\n",
            "Insert of existing embedding ID: Scanpy\n",
            "Insert of existing embedding ID: scAPA\n",
            "Insert of existing embedding ID: scAPAtrap\n",
            "Insert of existing embedding ID: SCAPE\n",
            "Insert of existing embedding ID: SCAPTURE\n",
            "Insert of existing embedding ID: scAR\n",
            "Insert of existing embedding ID: scArches\n",
            "Insert of existing embedding ID: scarf\n",
            "Insert of existing embedding ID: scasa\n",
            "Insert of existing embedding ID: scASfind\n",
            "Insert of existing embedding ID: scASK\n",
            "Insert of existing embedding ID: Scater\n",
            "Insert of existing embedding ID: scATOMIC\n",
            "Insert of existing embedding ID: SCATrEx\n",
            "Insert of existing embedding ID: SCATS\n",
            "Insert of existing embedding ID: scatterHatch\n",
            "Insert of existing embedding ID: SCATTome\n",
            "Insert of existing embedding ID: scBASE\n",
            "Insert of existing embedding ID: scBatch\n",
            "Insert of existing embedding ID: scbean\n",
            "Insert of existing embedding ID: scBERT\n",
            "Insert of existing embedding ID: scBGEDA\n",
            "Insert of existing embedding ID: scBio\n",
            "Insert of existing embedding ID: scBonita\n",
            "Insert of existing embedding ID: scBoolSeq\n",
            "Insert of existing embedding ID: scBridge\n",
            "Insert of existing embedding ID: scBT\n",
            "Insert of existing embedding ID: scBTF\n",
            "Insert of existing embedding ID: scBubbletree\n",
            "Insert of existing embedding ID: SCC\n",
            "Insert of existing embedding ID: scCAEs\n",
            "Insert of existing embedding ID: SCCAF\n",
            "Insert of existing embedding ID: scCAN\n",
            "Insert of existing embedding ID: scCancer\n",
            "Insert of existing embedding ID: scCancer2\n",
            "Insert of existing embedding ID: scCaps\n",
            "Insert of existing embedding ID: scCapsNet-mask\n",
            "Insert of existing embedding ID: scCATCH\n",
            "Insert of existing embedding ID: scCB2\n",
            "Insert of existing embedding ID: scCCESS\n",
            "Insert of existing embedding ID: scClassifR\n",
            "Insert of existing embedding ID: scClassify\n",
            "Insert of existing embedding ID: scCLUE\n",
            "Insert of existing embedding ID: scClustBench\n",
            "Insert of existing embedding ID: scclusteval\n",
            "Insert of existing embedding ID: scClustViz\n",
            "Insert of existing embedding ID: scCobra\n",
            "Insert of existing embedding ID: scCODA\n",
            "Insert of existing embedding ID: scCODE\n",
            "Insert of existing embedding ID: sccomp\n",
            "Insert of existing embedding ID: scConnect\n",
            "Insert of existing embedding ID: scConsensus\n",
            "Insert of existing embedding ID: scCorr\n",
            "Insert of existing embedding ID: scCustomize\n",
            "Insert of existing embedding ID: scDA\n",
            "Insert of existing embedding ID: scDALI\n",
            "Insert of existing embedding ID: scDAPA\n",
            "Insert of existing embedding ID: scDaPars\n",
            "Insert of existing embedding ID: scDART\n",
            "Insert of existing embedding ID: scDataviz\n",
            "Insert of existing embedding ID: scDblFinder\n",
            "Insert of existing embedding ID: scDBM\n",
            "Insert of existing embedding ID: scDC\n",
            "Insert of existing embedding ID: scDCC\n",
            "Insert of existing embedding ID: scDD\n",
            "Insert of existing embedding ID: scDDboost\n",
            "Insert of existing embedding ID: SCDE\n",
            "Insert of existing embedding ID: scDEA\n",
            "Insert of existing embedding ID: scDEAL\n",
            "Insert of existing embedding ID: scDECAF\n",
            "Insert of existing embedding ID: scDecouple\n",
            "Insert of existing embedding ID: scDEED\n",
            "Insert of existing embedding ID: scDeepCluster\n",
            "Insert of existing embedding ID: scDeepHash\n",
            "Insert of existing embedding ID: scDeepInsight\n",
            "Insert of existing embedding ID: scDeepSort\n",
            "Insert of existing embedding ID: scDesign\n",
            "Insert of existing embedding ID: scDesign2\n",
            "Insert of existing embedding ID: scDesign3\n",
            "Insert of existing embedding ID: scDetect\n",
            "Insert of existing embedding ID: scDGN\n",
            "Insert of existing embedding ID: scDHA\n",
            "Insert of existing embedding ID: scDHMap\n",
            "Insert of existing embedding ID: scdiff\n",
            "Insert of existing embedding ID: scDiffCom\n",
            "Insert of existing embedding ID: scDisInFact\n",
            "Insert of existing embedding ID: scDist\n",
            "Insert of existing embedding ID: scdNet\n",
            "Insert of existing embedding ID: scdney\n",
            "Insert of existing embedding ID: scDoc\n",
            "Insert of existing embedding ID: scdrake\n",
            "Insert of existing embedding ID: scDREAMER\n",
            "Insert of existing embedding ID: scDRS\n",
            "Insert of existing embedding ID: scDrug\n",
            "Insert of existing embedding ID: scds\n",
            "Insert of existing embedding ID: sceb\n",
            "Insert of existing embedding ID: scEC\n",
            "Insert of existing embedding ID: Scedar\n",
            "Insert of existing embedding ID: scEFSC\n",
            "Insert of existing embedding ID: SCell\n",
            "Insert of existing embedding ID: SCellBOW\n",
            "Insert of existing embedding ID: scellpam\n",
            "Insert of existing embedding ID: SCelVis\n",
            "Insert of existing embedding ID: scEMAIL\n",
            "Insert of existing embedding ID: SCENA\n",
            "Insert of existing embedding ID: SCENIC\n",
            "Insert of existing embedding ID: scenicplus\n",
            "Insert of existing embedding ID: SCEnt\n",
            "Insert of existing embedding ID: SCENT\n",
            "Insert of existing embedding ID: scEpath\n",
            "Insert of existing embedding ID: scETM\n",
            "Insert of existing embedding ID: SCEVAN\n",
            "Insert of existing embedding ID: scEvoNet\n",
            "Insert of existing embedding ID: scFates\n",
            "Insert of existing embedding ID: scFBA\n",
            "Insert of existing embedding ID: scFEA\n",
            "Insert of existing embedding ID: scFeatureFilter\n",
            "Insert of existing embedding ID: scFeatures\n",
            "Insert of existing embedding ID: scfind\n",
            "Insert of existing embedding ID: scFlow\n",
            "Insert of existing embedding ID: scFormer\n",
            "Insert of existing embedding ID: scFseCluster\n",
            "Insert of existing embedding ID: scFusion\n",
            "Insert of existing embedding ID: scGAE\n",
            "Insert of existing embedding ID: scGAN\n",
            "Insert of existing embedding ID: scGate\n",
            "Insert of existing embedding ID: scGBM\n",
            "Insert of existing embedding ID: scGCL\n",
            "Insert of existing embedding ID: scGEAToolbox\n",
            "Insert of existing embedding ID: scGen\n",
            "Insert of existing embedding ID: scGeneFit\n",
            "Insert of existing embedding ID: scGeom\n",
            "Insert of existing embedding ID: scGESTALT\n",
            "Insert of existing embedding ID: scGNN\n",
            "Insert of existing embedding ID: scGNN2\n",
            "Insert of existing embedding ID: scGPS\n",
            "Insert of existing embedding ID: scGPT\n",
            "Insert of existing embedding ID: scGraph\n",
            "Insert of existing embedding ID: scGraphNE\n",
            "Insert of existing embedding ID: scGREAT\n",
            "Insert of existing embedding ID: scGTM\n",
            "Insert of existing embedding ID: scGWAS\n",
            "Insert of existing embedding ID: schema\n",
            "Insert of existing embedding ID: schex\n",
            "Insert of existing embedding ID: scHinter\n",
            "Insert of existing embedding ID: SCHiRM\n",
            "Insert of existing embedding ID: schist\n",
            "Insert of existing embedding ID: scHLAcount\n",
            "Insert of existing embedding ID: SCHNAPPs\n",
            "Insert of existing embedding ID: SCHNEL\n",
            "Insert of existing embedding ID: scHOT\n",
            "Insert of existing embedding ID: scHPF\n",
            "Insert of existing embedding ID: scHPL\n",
            "Insert of existing embedding ID: scHumanNet\n",
            "Insert of existing embedding ID: scIAE\n",
            "Insert of existing embedding ID: scIB\n",
            "Insert of existing embedding ID: SCIBER\n",
            "Insert of existing embedding ID: SciBet\n",
            "Insert of existing embedding ID: sciCAN\n",
            "Insert of existing embedding ID: scicast\n",
            "Insert of existing embedding ID: sciCSR\n",
            "Insert of existing embedding ID: scID\n",
            "Insert of existing embedding ID: scIGANs\n",
            "Insert of existing embedding ID: scIHPF\n",
            "Insert of existing embedding ID: SCILEN\n",
            "Insert of existing embedding ID: Scillus\n",
            "Insert of existing embedding ID: SCIM\n",
            "Insert of existing embedding ID: SCIMITAR\n",
            "Insert of existing embedding ID: scImpute\n",
            "Insert of existing embedding ID: SCINA\n",
            "Insert of existing embedding ID: SCINET\n",
            "Insert of existing embedding ID: SCING\n",
            "Insert of existing embedding ID: SCING-pipeline\n",
            "Insert of existing embedding ID: scINSIGHT\n",
            "Insert of existing embedding ID: scIntegral\n",
            "Insert of existing embedding ID: sciPENN\n",
            "Insert of existing embedding ID: scipr\n",
            "Insert of existing embedding ID: scirpy\n",
            "Insert of existing embedding ID: scisorseqr\n",
            "Insert of existing embedding ID: ScisorWiz\n",
            "Insert of existing embedding ID: scISR\n",
            "Insert of existing embedding ID: SCISSORS\n",
            "Insert of existing embedding ID: scITD\n",
            "Insert of existing embedding ID: scIVA\n",
            "Insert of existing embedding ID: Sciviewer\n",
            "Insert of existing embedding ID: scJoint\n",
            "Insert of existing embedding ID: scKB\n",
            "Insert of existing embedding ID: scKINETICS\n",
            "Insert of existing embedding ID: scLAPA\n",
            "Insert of existing embedding ID: scLCA\n",
            "Insert of existing embedding ID: scLINE\n",
            "Insert of existing embedding ID: SClineager\n",
            "Insert of existing embedding ID: scLink\n",
            "Insert of existing embedding ID: scLRTC\n",
            "Insert of existing embedding ID: scLVM\n",
            "Insert of existing embedding ID: scMAGeCK\n",
            "Insert of existing embedding ID: scMAGIC\n",
            "Insert of existing embedding ID: scMAGS\n",
            "Insert of existing embedding ID: scmap\n",
            "Insert of existing embedding ID: scMAPA\n",
            "Insert of existing embedding ID: scMappR\n",
            "Insert of existing embedding ID: SCMAPpy\n",
            "Insert of existing embedding ID: SCMarker\n",
            "Insert of existing embedding ID: scMatch\n",
            "Insert of existing embedding ID: scMayoMap\n",
            "Insert of existing embedding ID: scMC\n",
            "Insert of existing embedding ID: scMCA\n",
            "Insert of existing embedding ID: scMDC\n",
            "Insert of existing embedding ID: scMEGA\n",
            "Insert of existing embedding ID: SCMER\n",
            "Insert of existing embedding ID: scMerge\n",
            "Insert of existing embedding ID: scMetric\n",
            "Insert of existing embedding ID: scMGCA\n",
            "Insert of existing embedding ID: scMiko\n",
            "Insert of existing embedding ID: scMINER\n",
            "Insert of existing embedding ID: scmixture\n",
            "Insert of existing embedding ID: scMLnet\n",
            "Insert of existing embedding ID: scMM\n",
            "Insert of existing embedding ID: scModels\n",
            "Insert of existing embedding ID: scMOG\n",
            "Insert of existing embedding ID: scMoMaT\n",
            "Insert of existing embedding ID: scMomentum\n",
            "Insert of existing embedding ID: scMOO\n",
            "Insert of existing embedding ID: scMRA\n",
            "Insert of existing embedding ID: scMRMA\n",
            "Insert of existing embedding ID: scMTD\n",
            "Insert of existing embedding ID: scMuffin\n",
            "Insert of existing embedding ID: scMultiSim\n",
            "Insert of existing embedding ID: SCmut\n",
            "Insert of existing embedding ID: scMVAE\n",
            "Insert of existing embedding ID: scNanoGPS\n",
            "Insert of existing embedding ID: scNBMF\n",
            "Insert of existing embedding ID: scNCA\n",
            "Insert of existing embedding ID: scNetViz\n",
            "Insert of existing embedding ID: SCNorm\n",
            "Insert of existing embedding ID: scNPF\n",
            "Insert of existing embedding ID: SCNS\n",
            "Insert of existing embedding ID: scNym\n",
            "Insert of existing embedding ID: SCODE\n",
            "Insert of existing embedding ID: SCOIT\n",
            "Insert of existing embedding ID: SComatic\n",
            "Insert of existing embedding ID: scone\n",
            "Insert of existing embedding ID: scOntoMatch\n",
            "Insert of existing embedding ID: SCope\n",
            "Insert of existing embedding ID: scopeplus\n",
            "Insert of existing embedding ID: SCORPION\n",
            "Insert of existing embedding ID: SCORPIUS\n",
            "Insert of existing embedding ID: SCOUP\n",
            "Insert of existing embedding ID: SCOUT\n",
            "Insert of existing embedding ID: scover\n",
            "Insert of existing embedding ID: SCP\n",
            "Insert of existing embedding ID: SCPA\n",
            "Insert of existing embedding ID: scPagwas\n",
            "Insert of existing embedding ID: scPanoView\n",
            "Insert of existing embedding ID: SCPattern\n",
            "Insert of existing embedding ID: scPCA\n",
            "Insert of existing embedding ID: scphaser\n",
            "Insert of existing embedding ID: scPheno\n",
            "Insert of existing embedding ID: scPhere\n",
            "Insert of existing embedding ID: scPipe\n",
            "Insert of existing embedding ID: scPlant\n",
            "Insert of existing embedding ID: scploid\n",
            "Insert of existing embedding ID: scPlot\n",
            "Insert of existing embedding ID: scPNMF\n",
            "Insert of existing embedding ID: scpoisson\n",
            "Insert of existing embedding ID: scPOP\n",
            "Insert of existing embedding ID: scPopCorn\n",
            "Insert of existing embedding ID: scPOST\n",
            "Insert of existing embedding ID: scPower\n",
            "Insert of existing embedding ID: scPred\n",
            "Insert of existing embedding ID: scprep\n",
            "Insert of existing embedding ID: scPretrain\n",
            "Insert of existing embedding ID: scPrisma\n",
            "Insert of existing embedding ID: scProject\n",
            "Insert of existing embedding ID: scProximitE\n",
            "Insert of existing embedding ID: scPSD\n",
            "Insert of existing embedding ID: scPSM\n",
            "Insert of existing embedding ID: SCpubr\n",
            "Insert of existing embedding ID: scQA\n",
            "Insert of existing embedding ID: SCQC\n",
            "Insert of existing embedding ID: scQCEA\n",
            "Insert of existing embedding ID: scQcut\n",
            "Insert of existing embedding ID: scQuint\n",
            "Insert of existing embedding ID: SCRABBLE\n",
            "Insert of existing embedding ID: scRAE\n",
            "Insert of existing embedding ID: scran\n",
            "Insert of existing embedding ID: scRANK\n",
            "Insert of existing embedding ID: scRare\n",
            "Insert of existing embedding ID: scRAS\n",
            "Insert of existing embedding ID: SCRAT\n",
            "Insert of existing embedding ID: scRCMF\n",
            "Insert of existing embedding ID: SCReadCounts\n",
            "Insert of existing embedding ID: scReadSim\n",
            "Insert of existing embedding ID: scReClassify\n",
            "Insert of existing embedding ID: scRecover\n",
            "Insert of existing embedding ID: scREhurdle\n",
            "Insert of existing embedding ID: scriabin\n",
            "Insert of existing embedding ID: Scribe\n",
            "Insert of existing embedding ID: SCRIP\n",
            "Insert of existing embedding ID: SCRL\n",
            "Insert of existing embedding ID: scRMD\n",
            "Insert of existing embedding ID: scRNA\n",
            "Insert of existing embedding ID: scRNABatchQC\n",
            "Insert of existing embedding ID: scRNAseqApp\n",
            "Insert of existing embedding ID: scRNASequest\n",
            "Insert of existing embedding ID: scRNAss\n",
            "Insert of existing embedding ID: scROSHI\n",
            "Insert of existing embedding ID: Scrublet\n",
            "Insert of existing embedding ID: scruff\n",
            "Insert of existing embedding ID: scRutiNy\n",
            "Insert of existing embedding ID: scry\n",
            "Insert of existing embedding ID: SCSA\n",
            "Insert of existing embedding ID: scSampler\n",
            "Insert of existing embedding ID: SCSC\n",
            "Insert of existing embedding ID: scScope\n",
            "Insert of existing embedding ID: scSDAE\n",
            "Insert of existing embedding ID: scSemiAE\n",
            "Insert of existing embedding ID: scSemiCluster\n",
            "Insert of existing embedding ID: scSensitiveGeneDefine\n",
            "Insert of existing embedding ID: scSeqComm\n",
            "Insert of existing embedding ID: scSeqR\n",
            "Insert of existing embedding ID: scShaper\n",
            "Insert of existing embedding ID: scShapes\n",
            "Insert of existing embedding ID: scSHARP\n",
            "Insert of existing embedding ID: scsim\n",
            "Insert of existing embedding ID: scSimClassify\n",
            "Insert of existing embedding ID: SCSit\n",
            "Insert of existing embedding ID: scSNPdemux\n",
            "Insert of existing embedding ID: scsnv\n",
            "Insert of existing embedding ID: scSO\n",
            "Insert of existing embedding ID: scSorter\n",
            "Insert of existing embedding ID: scSpace\n",
            "Insert of existing embedding ID: scSPARKL\n",
            "Insert of existing embedding ID: scSplit\n",
            "Insert of existing embedding ID: scSTEM\n",
            "Insert of existing embedding ID: scSVA\n",
            "Insert of existing embedding ID: scTab\n",
            "Insert of existing embedding ID: scTCRseq\n",
            "Insert of existing embedding ID: scTDA\n",
            "Insert of existing embedding ID: scTE\n",
            "Insert of existing embedding ID: scTenifoldKnk\n",
            "Insert of existing embedding ID: scTenifoldNet\n",
            "Insert of existing embedding ID: scTenifoldpy\n",
            "Insert of existing embedding ID: scTensor\n",
            "Insert of existing embedding ID: scTEP\n",
            "Insert of existing embedding ID: scTGIF\n",
            "Insert of existing embedding ID: scTHI\n",
            "Insert of existing embedding ID: scTIM\n",
            "Insert of existing embedding ID: scTITANS\n",
            "Insert of existing embedding ID: scTM\n",
            "Insert of existing embedding ID: scTOP\n",
            "Insert of existing embedding ID: scTopoGAN\n",
            "Insert of existing embedding ID: scTour\n",
            "Insert of existing embedding ID: sctransform\n",
            "Insert of existing embedding ID: scTree\n",
            "Insert of existing embedding ID: scTreeViz\n",
            "Insert of existing embedding ID: scTriangulate\n",
            "Insert of existing embedding ID: scTSSR\n",
            "Insert of existing embedding ID: ScType\n",
            "Insert of existing embedding ID: scTyper\n",
            "Insert of existing embedding ID: SCUBA\n",
            "Insert of existing embedding ID: SCUBI\n",
            "Insert of existing embedding ID: scumi\n",
            "Insert of existing embedding ID: scUnifrac\n",
            "Insert of existing embedding ID: SCUREL\n",
            "Insert of existing embedding ID: scUTRquant\n",
            "Insert of existing embedding ID: scuttle\n",
            "Insert of existing embedding ID: scVAE\n",
            "Insert of existing embedding ID: scVAEIT\n",
            "Insert of existing embedding ID: scVDMC\n",
            "Insert of existing embedding ID: scVEGs\n",
            "Insert of existing embedding ID: scVelo\n",
            "Insert of existing embedding ID: scvi-tools\n",
            "Insert of existing embedding ID: scViewer\n",
            "Insert of existing embedding ID: scViralQuant\n",
            "Insert of existing embedding ID: scVirtualCyto\n",
            "Insert of existing embedding ID: scvis\n",
            "Insert of existing embedding ID: scvr\n",
            "Insert of existing embedding ID: scWizard\n",
            "Insert of existing embedding ID: scziDesk\n",
            "Insert of existing embedding ID: SDAMS\n",
            "Insert of existing embedding ID: SDImpute\n",
            "Insert of existing embedding ID: SEACells\n",
            "Insert of existing embedding ID: SEAT\n",
            "Insert of existing embedding ID: secDrug\n",
            "Insert of existing embedding ID: Secuer\n",
            "Insert of existing embedding ID: SEDIM\n",
            "Insert of existing embedding ID: SEGtool\n",
            "Insert of existing embedding ID: SelfE\n",
            "Insert of existing embedding ID: SELINA\n",
            "Insert of existing embedding ID: SEMITONES\n",
            "Insert of existing embedding ID: SenePy\n",
            "Insert of existing embedding ID: SEPA\n",
            "Insert of existing embedding ID: SEPIRA\n",
            "Insert of existing embedding ID: seqgendiff\n",
            "Insert of existing embedding ID: SEQUIN\n",
            "Insert of existing embedding ID: SERGIO\n",
            "Insert of existing embedding ID: SERM\n",
            "Insert of existing embedding ID: Seurat\n",
            "Insert of existing embedding ID: SeuratV3Wizard\n",
            "Insert of existing embedding ID: SeuratWizard\n",
            "Insert of existing embedding ID: sfaira\n",
            "Insert of existing embedding ID: ShareNet\n",
            "Insert of existing embedding ID: SHARP\n",
            "Insert of existing embedding ID: Sharq\n",
            "Insert of existing embedding ID: ShinyCell\n",
            "Insert of existing embedding ID: ShinyMultiome-UiO\n",
            "Insert of existing embedding ID: SiCeLoRe\n",
            "Insert of existing embedding ID: SICILIAN\n",
            "Insert of existing embedding ID: SiCR\n",
            "Insert of existing embedding ID: Sierra\n",
            "Insert of existing embedding ID: SIEVE\n",
            "Insert of existing embedding ID: SiFINeT\n",
            "Insert of existing embedding ID: SiFT\n",
            "Insert of existing embedding ID: SigEMD\n",
            "Insert of existing embedding ID: sigGCN\n",
            "Insert of existing embedding ID: SigHotSpotter\n",
            "Insert of existing embedding ID: SIGMA\n",
            "Insert of existing embedding ID: signac\n",
            "Insert of existing embedding ID: SignacX\n",
            "Insert of existing embedding ID: signifinder\n",
            "Insert of existing embedding ID: SILGGM\n",
            "Insert of existing embedding ID: SIMBA\n",
            "Insert of existing embedding ID: SimCD\n",
            "Insert of existing embedding ID: SimiC\n",
            "Insert of existing embedding ID: SIMLR\n",
            "Insert of existing embedding ID: simpleaf\n",
            "Insert of existing embedding ID: SIMPLEs\n",
            "Insert of existing embedding ID: SIMS\n",
            "Insert of existing embedding ID: simspec\n",
            "Insert of existing embedding ID: SINC\n",
            "Insert of existing embedding ID: Sincast\n",
            "Insert of existing embedding ID: sincell\n",
            "Insert of existing embedding ID: SINCERA\n",
            "Insert of existing embedding ID: SINCERITIES\n",
            "Insert of existing embedding ID: SinCHet\n",
            "Insert of existing embedding ID: SingCellaR\n",
            "Insert of existing embedding ID: SINGE\n",
            "Insert of existing embedding ID: SingleCellExplorer\n",
            "Insert of existing embedding ID: SingleCellGGM\n",
            "Insert of existing embedding ID: singleCellHaystack\n",
            "Insert of existing embedding ID: singleCellHaystack-py\n",
            "Insert of existing embedding ID: SingleCellNet\n",
            "Insert of existing embedding ID: SingleCellSignalR\n",
            "Insert of existing embedding ID: SingleCellSignatureExplorer\n",
            "Insert of existing embedding ID: singleCellTK\n",
            "Insert of existing embedding ID: SinglePointRNA\n",
            "Insert of existing embedding ID: SingleR\n",
            "Insert of existing embedding ID: singleseqgset\n",
            "Insert of existing embedding ID: SingleSplice\n",
            "Insert of existing embedding ID: singlet\n",
            "Insert of existing embedding ID: SinNLRR\n",
            "Insert of existing embedding ID: SinQC\n",
            "Insert of existing embedding ID: Sinto\n",
            "Insert of existing embedding ID: SIP\n",
            "Insert of existing embedding ID: SiPSiC\n",
            "Insert of existing embedding ID: sircel\n",
            "Insert of existing embedding ID: SISUA\n",
            "Insert of existing embedding ID: siVAE\n",
            "Insert of existing embedding ID: SJD\n",
            "Insert of existing embedding ID: SkewC\n",
            "Insert of existing embedding ID: slalom\n",
            "Insert of existing embedding ID: sleepwalk\n",
            "Insert of existing embedding ID: SLICE\n",
            "Insert of existing embedding ID: SLICER\n",
            "Insert of existing embedding ID: SLIDE\n",
            "Insert of existing embedding ID: slingshot\n",
            "Insert of existing embedding ID: SMAI\n",
            "Insert of existing embedding ID: SMaSH\n",
            "Insert of existing embedding ID: SMGR\n",
            "Insert of existing embedding ID: SMILE\n",
            "Insert of existing embedding ID: SMNN\n",
            "Insert of existing embedding ID: SMSC\n",
            "Insert of existing embedding ID: SMURF\n",
            "Insert of existing embedding ID: SnapATAC2\n",
            "Insert of existing embedding ID: SnapCCESS\n",
            "Insert of existing embedding ID: SNN-Cliq\n",
            "Insert of existing embedding ID: SoCube\n",
            "Insert of existing embedding ID: solo\n",
            "Insert of existing embedding ID: SoloTE\n",
            "Insert of existing embedding ID: SoptSC\n",
            "Insert of existing embedding ID: SOUP\n",
            "Insert of existing embedding ID: souporcell\n",
            "Insert of existing embedding ID: SoupX\n",
            "Insert of existing embedding ID: SPADE\n",
            "Insert of existing embedding ID: SpaOTsc\n",
            "Insert of existing embedding ID: sparca\n",
            "Insert of existing embedding ID: SparseDC\n",
            "Insert of existing embedding ID: SPARSim\n",
            "Insert of existing embedding ID: SPaRTAN\n",
            "Insert of existing embedding ID: SpatialScope\n",
            "Insert of existing embedding ID: SPECK\n",
            "Insert of existing embedding ID: speckle\n",
            "Insert of existing embedding ID: Specter\n",
            "Insert of existing embedding ID: SPECTRA\n",
            "Insert of existing embedding ID: Spectre\n",
            "Insert of existing embedding ID: Spectrum\n",
            "Insert of existing embedding ID: Sphetcher\n",
            "Insert of existing embedding ID: splattdr\n",
            "Insert of existing embedding ID: Splatter\n",
            "Insert of existing embedding ID: spliceJAC\n",
            "Insert of existing embedding ID: SPmarker\n",
            "Insert of existing embedding ID: SPRING\n",
            "Insert of existing embedding ID: SPsimSeq\n",
            "Insert of existing embedding ID: SR2\n",
            "Insert of existing embedding ID: SSBER\n",
            "Insert of existing embedding ID: sscClust\n",
            "Insert of existing embedding ID: SSrGE\n",
            "Insert of existing embedding ID: StabMap\n",
            "Insert of existing embedding ID: STACAS\n",
            "Insert of existing embedding ID: STAR\n",
            "Insert of existing embedding ID: STARE\n",
            "Insert of existing embedding ID: starmapVis\n",
            "Insert of existing embedding ID: StationaryOT\n",
            "Insert of existing embedding ID: STEGO-R\n",
            "Insert of existing embedding ID: STEM\n",
            "Insert of existing embedding ID: STEMNET\n",
            "Insert of existing embedding ID: STREAK\n",
            "Insert of existing embedding ID: STREAM\n",
            "Insert of existing embedding ID: supCPM\n",
            "Insert of existing embedding ID: SuperCell\n",
            "Insert of existing embedding ID: supercells\n",
            "Insert of existing embedding ID: Superscan\n",
            "Insert of existing embedding ID: SwarnSeq\n",
            "Insert of existing embedding ID: switchde\n",
            "Insert of existing embedding ID: SWNE\n",
            "Insert of existing embedding ID: Symphony\n",
            "Insert of existing embedding ID: SymSim\n",
            "Insert of existing embedding ID: SynergisticCore\n",
            "Insert of existing embedding ID: TACTiCS\n",
            "Insert of existing embedding ID: talklr\n",
            "Insert of existing embedding ID: Tangram\n",
            "Insert of existing embedding ID: TAR-scRNA-seq\n",
            "Insert of existing embedding ID: TASC\n",
            "Insert of existing embedding ID: tascCODA\n",
            "Insert of existing embedding ID: TASIC\n",
            "Insert of existing embedding ID: tauFisher\n",
            "Insert of existing embedding ID: TBSP\n",
            "Insert of existing embedding ID: TCM\n",
            "Insert of existing embedding ID: TDEseq\n",
            "Insert of existing embedding ID: TDL\n",
            "Insert of existing embedding ID: TedSim\n",
            "Insert of existing embedding ID: Tempo\n",
            "Insert of existing embedding ID: Tempora\n",
            "Insert of existing embedding ID: TENET\n",
            "Insert of existing embedding ID: TFvelo\n",
            "Insert of existing embedding ID: tGPT\n",
            "Insert of existing embedding ID: TILPRED\n",
            "Insert of existing embedding ID: TimeTalk\n",
            "Insert of existing embedding ID: TimiGP\n",
            "Insert of existing embedding ID: TinGa\n",
            "Insert of existing embedding ID: TITAN\n",
            "Insert of existing embedding ID: tn_test\n",
            "Insert of existing embedding ID: Tobiko\n",
            "Insert of existing embedding ID: TOMAS\n",
            "Insert of existing embedding ID: TooManyCells\n",
            "Insert of existing embedding ID: TooManyCellsInteractive\n",
            "Insert of existing embedding ID: TopOMetry\n",
            "Insert of existing embedding ID: topslam\n",
            "Insert of existing embedding ID: TOSICA\n",
            "Insert of existing embedding ID: Totem\n",
            "Insert of existing embedding ID: TraCeR\n",
            "Insert of existing embedding ID: trackplot\n",
            "Insert of existing embedding ID: tradeSeq\n",
            "Insert of existing embedding ID: TrAGEDy\n",
            "Insert of existing embedding ID: Trajan\n",
            "Insert of existing embedding ID: TrajectoryNet\n",
            "Insert of existing embedding ID: transformGamPoi\n",
            "Insert of existing embedding ID: Transmorph\n",
            "Insert of existing embedding ID: TRAPeS\n",
            "Insert of existing embedding ID: TraSig\n",
            "Insert of existing embedding ID: traviz\n",
            "Insert of existing embedding ID: Tree-SNE\n",
            "Insert of existing embedding ID: TreeCorTreat\n",
            "Insert of existing embedding ID: TREG\n",
            "Insert of existing embedding ID: trendsceek\n",
            "Insert of existing embedding ID: Triangulate\n",
            "Insert of existing embedding ID: tricycle\n",
            "Insert of existing embedding ID: Triku\n",
            "Insert of existing embedding ID: TripletCell\n",
            "Insert of existing embedding ID: TRIPOD\n",
            "Insert of existing embedding ID: TRUmiCount\n",
            "Insert of existing embedding ID: TRUST4\n",
            "Insert of existing embedding ID: trVAE\n",
            "Insert of existing embedding ID: TSCAN\n",
            "Insert of existing embedding ID: TSEE\n",
            "Insert of existing embedding ID: Tweedieverse\n",
            "Insert of existing embedding ID: TWO-SIGMA\n",
            "Insert of existing embedding ID: txburst\n",
            "Insert of existing embedding ID: UCell\n",
            "Insert of existing embedding ID: ucie\n",
            "Insert of existing embedding ID: UICPC\n",
            "Insert of existing embedding ID: UMI_analysis\n",
            "Insert of existing embedding ID: UMI-tools\n",
            "Insert of existing embedding ID: UMICollapse\n",
            "Insert of existing embedding ID: umis\n",
            "Insert of existing embedding ID: UNCURL\n",
            "Insert of existing embedding ID: UNIFAN\n",
            "Insert of existing embedding ID: UnionCom\n",
            "Insert of existing embedding ID: UniPath\n",
            "Insert of existing embedding ID: uniPort\n",
            "Insert of existing embedding ID: UniTVelo\n",
            "Insert of existing embedding ID: UniverSC\n",
            "Insert of existing embedding ID: UPhyloplot2\n",
            "Insert of existing embedding ID: URD\n",
            "Insert of existing embedding ID: uSORT\n",
            "Insert of existing embedding ID: VAE-SNE\n",
            "Insert of existing embedding ID: vaeda\n",
            "Insert of existing embedding ID: VALERIE\n",
            "Insert of existing embedding ID: valiDrops\n",
            "Insert of existing embedding ID: VAM\n",
            "Insert of existing embedding ID: VAMF\n",
            "Insert of existing embedding ID: VarTrix\n",
            "Insert of existing embedding ID: VASC\n",
            "Insert of existing embedding ID: VDJdive\n",
            "Insert of existing embedding ID: VDJPuzzle\n",
            "Insert of existing embedding ID: VECTOR\n",
            "Insert of existing embedding ID: VEGA\n",
            "Insert of existing embedding ID: velociraptor\n",
            "Insert of existing embedding ID: velocyto\n",
            "Insert of existing embedding ID: Velorama\n",
            "Insert of existing embedding ID: VeloRep\n",
            "Insert of existing embedding ID: VeloSim\n",
            "Insert of existing embedding ID: VeloVAE\n",
            "Insert of existing embedding ID: VeloViz\n",
            "Insert of existing embedding ID: velvet\n",
            "Insert of existing embedding ID: Venus\n",
            "Insert of existing embedding ID: VeTra\n",
            "Insert of existing embedding ID: VIA\n",
            "Insert of existing embedding ID: VIPCCA\n",
            "Insert of existing embedding ID: VIPER\n",
            "Insert of existing embedding ID: Vireo\n",
            "Insert of existing embedding ID: VisCello\n",
            "Insert of existing embedding ID: VISION\n",
            "Insert of existing embedding ID: visnormsc\n",
            "Insert of existing embedding ID: vissE\n",
            "Insert of existing embedding ID: VITAE\n",
            "Insert of existing embedding ID: Vitessce\n",
            "Insert of existing embedding ID: VoxHunt\n",
            "Insert of existing embedding ID: VPAC\n",
            "Insert of existing embedding ID: Vulture\n",
            "Insert of existing embedding ID: waddR\n",
            "Insert of existing embedding ID: WASP\n",
            "Insert of existing embedding ID: WAT3R\n",
            "Insert of existing embedding ID: WaveCrest\n",
            "Insert of existing embedding ID: WEDGE\n",
            "Insert of existing embedding ID: wf-single-cell\n",
            "Insert of existing embedding ID: Wishbone\n",
            "Insert of existing embedding ID: wot\n",
            "Insert of existing embedding ID: Wx\n",
            "Insert of existing embedding ID: XClone\n",
            "Insert of existing embedding ID: XenoCell\n",
            "Insert of existing embedding ID: ZetaSuite\n",
            "Insert of existing embedding ID: ZIAQ\n",
            "Insert of existing embedding ID: ZIFA\n",
            "Insert of existing embedding ID: ZINB-WaVE\n",
            "Insert of existing embedding ID: ZINBAE\n",
            "Insert of existing embedding ID: ZINBMM\n",
            "Insert of existing embedding ID: zingeR\n",
            "Insert of existing embedding ID: zUMIs\n"
          ]
        }
      ],
      "source": [
        "scrnatools_description_collection.add(\n",
        "    documents = list(df['extented_desc_readme_trim']),\n",
        "    metadatas = df.drop(['extented_desc_readme_trim'], axis = 1).to_dict(orient='records'),\n",
        "    ids = list(df.Name))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Query DB"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "metadata": {
        "id": "LrxrKjXxfWEb"
      },
      "outputs": [],
      "source": [
        "def query_collection(collection, query, max_results, dataframe):\n",
        "    results = collection.query(query_texts=query, n_results=max_results, include=['distances'])\n",
        "    df = pd.DataFrame({\n",
        "                'id':results['ids'][0],\n",
        "                'score':results['distances'][0],\n",
        "                'content': dataframe[dataframe.Name.isin(results['ids'][0])]['extented_desc'],\n",
        "                'platform': dataframe[dataframe.Name.isin(results['ids'][0])]['Platform'],\n",
        "                })\n",
        "\n",
        "    return df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 30,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 204
        },
        "id": "vzw1RN-6fpGe",
        "outputId": "becb7dce-af30-4a5d-8418-b54c7637bcd4"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>id</th>\n",
              "      <th>score</th>\n",
              "      <th>content</th>\n",
              "      <th>platform</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>AmbientContaminationMetrics</td>\n",
              "      <td>0.365000</td>\n",
              "      <td>Platform: Python\\n Description: Quantification...</td>\n",
              "      <td>Python</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>305</th>\n",
              "      <td>ddqc</td>\n",
              "      <td>0.395694</td>\n",
              "      <td>Platform: Python/R\\n Description: Biology-cent...</td>\n",
              "      <td>Python/R</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>390</th>\n",
              "      <td>SinQC</td>\n",
              "      <td>0.401830</td>\n",
              "      <td>Platform: Python\\n Description: Quality contro...</td>\n",
              "      <td>Python</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>451</th>\n",
              "      <td>fba</td>\n",
              "      <td>0.412007</td>\n",
              "      <td>Platform: Python\\n Description: A flexible and...</td>\n",
              "      <td>Python</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1427</th>\n",
              "      <td>DrSeq2</td>\n",
              "      <td>0.416470</td>\n",
              "      <td>Platform: R/Python\\n Description: A Method and...</td>\n",
              "      <td>R/Python</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                               id     score  \\\n",
              "20    AmbientContaminationMetrics  0.365000   \n",
              "305                          ddqc  0.395694   \n",
              "390                         SinQC  0.401830   \n",
              "451                           fba  0.412007   \n",
              "1427                       DrSeq2  0.416470   \n",
              "\n",
              "                                                content  platform  \n",
              "20    Platform: Python\\n Description: Quantification...    Python  \n",
              "305   Platform: Python/R\\n Description: Biology-cent...  Python/R  \n",
              "390   Platform: Python\\n Description: Quality contro...    Python  \n",
              "451   Platform: Python\\n Description: A flexible and...    Python  \n",
              "1427  Platform: R/Python\\n Description: A Method and...  R/Python  "
            ]
          },
          "execution_count": 30,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "query_collection(scrnatools_description_collection, 'quality controll python', 5, df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "nlVX-L4DqccT"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "hackathon",
      "language": "python",
      "name": "hackathon"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.12"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
